{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "gpuType": "T4"
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    },
    "accelerator": "GPU"
  },
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "### download files"
      ],
      "metadata": {
        "id": "u_l0yJZFUQnI"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Aureococcus anophagefferens: https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_000186865.1/\n",
        "\n",
        "NCBI path for my files: https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/186/865/GCA_000186865.1_v_1.0/"
      ],
      "metadata": {
        "id": "t1fwTMRHPDym"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "base_path = 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/151/665/GCF_000151665.1_ASM15166v1/'"
      ],
      "metadata": {
        "id": "mK-rrH0Qh7xi"
      },
      "execution_count": 2,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "from_genomic_file = 'GCF_000151665.1_ASM15166v1_cds_from_genomic.fna.gz'\n",
        "\n",
        "!wget {base_path}/{from_genomic_file} -q && gunzip -d {from_genomic_file} -q && mv {from_genomic_file[:-3]} cds.fna 2> /dev/null\n",
        "!head cds.fna"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "AHnV-F_fP20R",
        "outputId": "f46aa6ef-5d11-486a-8f01-38bf51ab55dc"
      },
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            ">lcl|NW_013171815.1_cds_XP_012893584.1_1 [locus_tag=GSBLH_T00000003001] [db_xref=UniProtKB/TrEMBL:D8LUP7,GeneID:24917328] [protein=uncharacterized protein] [protein_id=XP_012893584.1] [location=complement(join(28397..28559,28593..28678,28708..28801,28831..28969,29003..29226,29259..29364,29398..29875))] [gbkey=CDS]\n",
            "ATGCAAAACAAAGAATTGGATGGCCTTTTCCCGGAAGATCACATTTCGGTATACTATTCTGTTTTACACTTTCGTGCAAT\n",
            "TCACAAAACAGAAGGCGATCTTGTATCAAACCTGACATTACAATGGAGTCAAGAAGAATGCTTCCGTCCAATTCTAGAAA\n",
            "AATATGACTTAGTTCATTTGTTACTCGAGTTTAAGGAAAATGATTGGATACAAGATGTTCGTGCAGCAATCAGTAACCAT\n",
            "ACTGCCACTTCGGCTCTGATTGAACAATACCATGATTTTCAAAGTGTACTCATAACGAGAAAGTTACAATTGCGCTCTGA\n",
            "GTTCTATAATGTGAGTCATTACTCGGATAGTCCAATCCATTGTGATGGATGTGATATTCCAAGAAACCAAAAGAAGGTCA\n",
            "AAGCAATATTGAACTATCAGAACGGCTTGCTCATCCGTAAATTTGTGGGGAATGAGGAAGTCGAATATACAGATACGCCA\n",
            "TGGTTTCCTTCCAACGATCAGAAATACGATGTAACCGCCATTGTCACCGCTTTTGGATACAATCGTTTGTTTGCACTACG\n",
            "TCAATTCATGTATCGATATCAAGGACCGATTGTTCTTGTGATTTATGCCACATCGACCCAGGAAGTCCACCTTGTGCGTT\n",
            "ACATTTCAACTCACTTCATTCCCAAACGTGTTACAATTCTTTTTTATCTAGTCAGTCGTTATCTCAAATCCAGTACTGTG\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "protein_file = 'GCF_000151665.1_ASM15166v1_protein.faa.gz'\n",
        "\n",
        "!wget {base_path}/{protein_file} -q && gunzip -d {protein_file} -q && mv {protein_file[:-3]} protein.faa 2> /dev/null\n",
        "!head protein.faa"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "bauoSRdkTYyR",
        "outputId": "1808c093-ce1f-40a4-f10a-1024ec1c9907"
      },
      "execution_count": 4,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            ">XP_012893584.1 uncharacterized protein [Blastocystis hominis]\n",
            "MQNKELDGLFPEDHISVYYSVLHFRAIHKTEGDLVSNLTLQWSQEECFRPILEKYDLVHLLLEFKENDWIQDVRAAISNH\n",
            "TATSALIEQYHDFQSVLITRKLQLRSEFYNVSHYSDSPIHCDGCDIPRNQKKVKAILNYQNGLLIRKFVGNEEVEYTDTP\n",
            "WFPSNDQKYDVTAIVTAFGYNRLFALRQFMYRYQGPIVLVIYATSTQEVHLVRYISTHFIPKRVTILFYLVSRYLKSSTV\n",
            "FPINRLRNLAIRNIRTTHFLILDMDLRLSLNTYKEVMSLPQFLYHSNRSAVILPVFFYKGKQILAHCSTTESCSYLSNWI\n",
            "QPENKLELIECINRKICISNKNNIRTHMYVMPEWFTTSQESHVSRVRCFITNFMEPYLMVKYSPTLPLFEERFVNYGYNK\n",
            "VQYFEHLRQAGFQFYILNHAFAMDFPHPE\n",
            ">XP_012893585.1 uncharacterized protein [Blastocystis hominis]\n",
            "MNKTNEKAYQVSTVKTSIPTHGNDLLTPIQLKDFEYDYDDMDYPVPFNKLVSDISNRYVNYGVKIESVPFVLTGDIPKFE\n",
            "KQTSEVLMMTGASDNHALGSINCLFSMVVADPYASYMYIDLGLSTRFKNILFTHFNTIHQIQKKMNSTGFIAYRKFKFNS\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### download hmmer"
      ],
      "metadata": {
        "id": "SlonSnxk3egY"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!wget http://eddylab.org/software/hmmer/hmmer-3.3.2.tar.gz -q && tar xf hmmer-3.3.2.tar.gz\n",
        "!cd hmmer-3.3.2 && ./configure > /dev/null && make 2> /dev/null > /dev/null && make install 2> /dev/null > /dev/null"
      ],
      "metadata": {
        "id": "Fbp7lj7e3eQ4"
      },
      "execution_count": 5,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!wget http://ftp.ebi.ac.uk/pub/databases/Pfam/releases/Pfam35.0/Pfam-A.hmm.gz -q && gunzip Pfam-A.hmm.gz -q && hmmpress Pfam-A.hmm > /dev/null\n",
        "!hmmfetch --index Pfam-A.hmm.h3m >> /dev/null"
      ],
      "metadata": {
        "id": "urjRtJHx3roH"
      },
      "execution_count": 6,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### search for sequences with hmmer"
      ],
      "metadata": {
        "id": "EuqLPHhi3yNq"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!hmmstat Pfam-A.hmm.h3m > pfam.txt\n",
        "!head pfam.txt -n20"
      ],
      "metadata": {
        "id": "YXuIlBTrLWq9",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "1484c333-c80d-4ff8-9492-49c5f84fcbd4"
      },
      "execution_count": 14,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "# hmmstat :: display summary statistics for a profile file\n",
            "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
            "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
            "# Freely distributed under the BSD open source license.\n",
            "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
            "#\n",
            "# idx  name                 accession        nseq eff_nseq      M relent   info p relE compKL\n",
            "# ---- -------------------- ------------ -------- -------- ------ ------ ------ ------ ------\n",
            "1      1-cysPrx_C           PF10417.12         40    17.43     40   1.37   1.32   1.29   0.09\n",
            "2      120_Rick_ant         PF12574.11          2     0.45    238   0.59   0.60   0.52   0.02\n",
            "3      12TM_1               PF09847.12          7     2.28    449   0.59   0.62   0.51   0.09\n",
            "4      14-3-3               PF00244.23        149     2.50    222   0.59   0.59   0.52   0.05\n",
            "5      17kDa_Anti_2         PF16998.8           6     0.99    116   0.59   0.57   0.52   0.03\n",
            "6      2-Hacid_dh           PF00389.33         78     5.69    134   0.59   0.62   0.44   0.03\n",
            "7      2-Hacid_dh_C         PF02826.22        135     3.55    178   0.59   0.60   0.53   0.01\n",
            "8      2-oxoacid_dh         PF00198.26         70     3.33    233   0.59   0.61   0.53   0.02\n",
            "9      2-oxogl_dehyd_N      PF16078.8         208    11.82     41   1.34   1.23   1.33   0.13\n",
            "10     2-ph_phosp           PF04029.17        125     3.02    226   0.59   0.62   0.51   0.02\n",
            "11     2-thiour_desulf      PF04463.15        338     2.11    137   0.59   0.57   0.46   0.02\n",
            "12     23ISL                PF16620.8          16     1.48    162   0.59   0.61   0.54   0.07\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "with open('pfam.txt', 'r') as f:\n",
        "    lst = f.readlines()[8:]\n",
        "\n",
        "accessions = {}\n",
        "for line in lst:\n",
        "    k = line.split()[2].split('.')[0]\n",
        "    new = accessions.get(k, []) + [line.split()[2]]\n",
        "    accessions[k] = new\n",
        "\n",
        "names = [\n",
        "    name\n",
        "    for i in [\n",
        "        'PF09103',\n",
        "        'PF00888',\n",
        "        'PF12934',\n",
        "        'PF10513',\n",
        "        'PF12796',\n",
        "        'PF00149',\n",
        "        'PF01585',\n",
        "        'PF00505',\n",
        "        'PF00443',\n",
        "        'PF05764',\n",
        "    ]\n",
        "    for name in accessions[i]\n",
        "]\n",
        "with open('selected.txt', 'w') as f:\n",
        "    [print(name, file=f) for name in names]"
      ],
      "metadata": {
        "id": "HtQPqLG3rnK9"
      },
      "execution_count": 15,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!hmmfetch -f Pfam-A.hmm.h3m selected.txt > selected.hmm\n",
        "!hmmsearch selected.hmm protein.faa > result_hmm.txt\n",
        "!head -n50 result_hmm.txt"
      ],
      "metadata": {
        "id": "QLF5u8xN5kTT",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "0e17d722-c35d-410c-a948-b487512bb301"
      },
      "execution_count": 16,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "# hmmsearch :: search profile(s) against a sequence database\n",
            "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
            "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
            "# Freely distributed under the BSD open source license.\n",
            "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
            "# query HMM file:                  selected.hmm\n",
            "# target sequence database:        protein.faa\n",
            "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
            "\n",
            "Query:       BRCA-2_OB1  [M=124]\n",
            "Accession:   PF09103.13\n",
            "Description: BRCA2, oligonucleotide/oligosaccharide-binding, domain 1\n",
            "Scores for complete sequences (score includes all domains):\n",
            "   --- full sequence ---   --- best 1 domain ---    -#dom-\n",
            "    E-value  score  bias    E-value  score  bias    exp  N  Sequence       Description\n",
            "    ------- ------ -----    ------- ------ -----   ---- --  --------       -----------\n",
            "    3.2e-25   86.9   0.0    1.3e-24   84.9   0.0    2.2  1  XP_012893636.1  uncharacterized protein [Blastocystis hominis\n",
            "  ------ inclusion threshold ------\n",
            "      0.048   12.0   0.0      0.083   11.3   0.0    1.4  1  XP_012895637.1  uncharacterized protein [Blastocystis hominis\n",
            "\n",
            "\n",
            "Domain annotation for each sequence (and alignments):\n",
            ">> XP_012893636.1  uncharacterized protein [Blastocystis hominis]\n",
            "   #    score  bias  c-Evalue  i-Evalue hmmfrom  hmm to    alifrom  ali to    envfrom  env to     acc\n",
            " ---   ------ ----- --------- --------- ------- -------    ------- -------    ------- -------    ----\n",
            "   1 !   84.9   0.0   4.4e-28   1.3e-24       1     123 [.     869    1040 ..     869    1041 .. 0.95\n",
            "\n",
            "  Alignments for each domain:\n",
            "  == domain 1  score: 84.9 bits;  conditional E-value: 4.4e-28\n",
            "                      -HHHHHHTTSSTT--SEEEEEEE------------.....................----------EEEE-SS-EEEE---HHHHHHHHTT-S CS\n",
            "      BRCA-2_OB1    1 salkkilerDdsaskpmvLcVsaiekkesekeeke.....................kekeksekaeleltDGWYsikavlDakltqllkrgki 72  \n",
            "                      ++lk+i++rD+  ++ +vL+Vs+i k++   + +                      +++e+  + ++eltDGWY i+a lDa+lt ll++++i\n",
            "  XP_012893636.1  869 PPLKRIVQRDEAETRFIVLLVSDILKSPPLPDFDYlsdpesnsaadtnndskpdstSKSEAPPAGVIELTDGWYFIEAILDASLTILLQTHQI 961 \n",
            "                      78************************9988766558999**************998777788899**************************** PP\n",
            "\n",
            "                      -TT-EEEES--EEES--S---GGG.............................--SS-EEE--TTTEEE--SSS-SEE- CS\n",
            "      BRCA-2_OB1   73 kvGqKlvvagaelvgseeavsple............................naseslllkisaNstrrarwnakLGfq 123 \n",
            "                      + G Kl++a++++ ++e+++ ple                            +  ++  l +++N+trra wn+ LGf+\n",
            "  XP_012893636.1  962 YPGLKLRIACSHMRNNENGCAPLEittnqprrpcsaffnsfseevpirfyttDGLSTPQLMLQYNGTRRAAWNSYLGFT 1040\n",
            "                      ***************************************************9999**********************96 PP\n",
            "\n",
            ">> XP_012895637.1  uncharacterized protein [Blastocystis hominis]\n",
            "   #    score  bias  c-Evalue  i-Evalue hmmfrom  hmm to    alifrom  ali to    envfrom  env to     acc\n",
            " ---   ------ ----- --------- --------- ------- -------    ------- -------    ------- -------    ----\n",
            "   1 ?   11.3   0.0   2.7e-05     0.083      43      69 ..     198     221 ..     168     245 .. 0.81\n",
            "\n",
            "  Alignments for each domain:\n",
            "  == domain 1  score: 11.3 bits;  conditional E-value: 2.7e-05\n",
            "                     ---EEEE-SS-EEEE---HHHHHHHHT CS\n",
            "      BRCA-2_OB1  43 kaeleltDGWYsikavlDakltqllkr 69 \n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "\n",
        "\n",
        "with open('result_hmm.txt', 'r') as f:\n",
        "    lst = f.readlines()\n",
        "\n",
        "positions = []\n",
        "cnt = 0\n",
        "for line in lst:\n",
        "    if line[:6] == 'Query:':\n",
        "        start = cnt\n",
        "    elif line[:6] == 'Domain':\n",
        "        if start != -1:\n",
        "            positions.append([start, cnt])\n",
        "        start = -1\n",
        "    cnt += 1\n",
        "\n",
        "light_lines = []\n",
        "df = pd.DataFrame(columns=['query', 'accession', 'description', 'gene'])\n",
        "for start, end in positions:\n",
        "    light_lines += lst[start:end]\n",
        "    genes = [\n",
        "        x[8]\n",
        "        for i in range(start + 7, end)\n",
        "        if len(x := lst[i].split()) >= 8\n",
        "    ] or ['None']\n",
        "    for i in genes:\n",
        "        df.loc[len(df.index)] = [\n",
        "            lst[start].split()[1],\n",
        "            lst[start + 1].split()[1],\n",
        "            lst[start + 2].split()[1],\n",
        "            i,\n",
        "        ]\n",
        "\n",
        "with open('epigenetics_genes.txt', 'w') as f:\n",
        "    f.write(''.join(light_lines))\n",
        "df.to_csv('family_gene.csv')\n",
        "df.sample(4)"
      ],
      "metadata": {
        "id": "iB2jNFql4Vmy",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 175
        },
        "outputId": "92407357-c018-49f4-c912-a8a9d4b9ff7b"
      },
      "execution_count": 17,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "           query   accession       description            gene\n",
              "6        FTO_CTD  PF12934.10               FTO  XP_012895156.1\n",
              "34   Metallophos  PF00149.31  Calcineurin-like  XP_012895513.1\n",
              "100          UCH  PF00443.32         Ubiquitin  XP_012898317.1\n",
              "86           UCH  PF00443.32         Ubiquitin  XP_012897478.1"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-b7055f52-5774-4e04-b548-b72cdc8f9904\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>query</th>\n",
              "      <th>accession</th>\n",
              "      <th>description</th>\n",
              "      <th>gene</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>FTO_CTD</td>\n",
              "      <td>PF12934.10</td>\n",
              "      <td>FTO</td>\n",
              "      <td>XP_012895156.1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>34</th>\n",
              "      <td>Metallophos</td>\n",
              "      <td>PF00149.31</td>\n",
              "      <td>Calcineurin-like</td>\n",
              "      <td>XP_012895513.1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>100</th>\n",
              "      <td>UCH</td>\n",
              "      <td>PF00443.32</td>\n",
              "      <td>Ubiquitin</td>\n",
              "      <td>XP_012898317.1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>86</th>\n",
              "      <td>UCH</td>\n",
              "      <td>PF00443.32</td>\n",
              "      <td>Ubiquitin</td>\n",
              "      <td>XP_012897478.1</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-b7055f52-5774-4e04-b548-b72cdc8f9904')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-b7055f52-5774-4e04-b548-b72cdc8f9904 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-b7055f52-5774-4e04-b548-b72cdc8f9904');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-319cec3b-5499-4b57-a402-dd6dcc1f40b9\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-319cec3b-5499-4b57-a402-dd6dcc1f40b9')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-319cec3b-5499-4b57-a402-dd6dcc1f40b9 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "summary": "{\n  \"name\": \"df\",\n  \"rows\": 4,\n  \"fields\": [\n    {\n      \"column\": \"query\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 3,\n        \"samples\": [\n          \"FTO_CTD\",\n          \"Metallophos\",\n          \"UCH\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"accession\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 3,\n        \"samples\": [\n          \"PF12934.10\",\n          \"PF00149.31\",\n          \"PF00443.32\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"description\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 3,\n        \"samples\": [\n          \"FTO\",\n          \"Calcineurin-like\",\n          \"Ubiquitin\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"gene\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 4,\n        \"samples\": [\n          \"XP_012895513.1\",\n          \"XP_012897478.1\",\n          \"XP_012895156.1\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 17
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### G-квадруплексы"
      ],
      "metadata": {
        "id": "Q0_DQjXCrsQF"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "genomic_file = 'GCF_000151665.1_ASM15166v1_genomic.fna.gz'\n",
        "\n",
        "!wget {base_path}/{genomic_file} -q && gunzip -d {genomic_file} -q && mv {genomic_file[:-3]} genome.fna\n",
        "!head genome.fna"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Kv4zzfIXrqho",
        "outputId": "28c28c83-6d4c-4a00-a444-b26fc211bd5a"
      },
      "execution_count": 18,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            ">NW_013171815.1 Blastocystis hominis, Singapore isolate B (sub-type 7) whole genome shotgun sequence assembly, scaffold_0\n",
            "CTGACTAAGCATCCCTCTTAAAAGTCGAGgctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccct\n",
            "aaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaa\n",
            "ccctacccatgaaccctgaccctagccctaaccctggccctcaactcgaaccctagccctaaccctggccctcaactcga\n",
            "accctagccctaaccctgaccctaacgGGTCAGGATAGGGTAATGAGAACCTGAAccgaagcatgaacccgaaccctagc\n",
            "cctaaccctggccctcgACTCGGATCcgaaccctgaccctagccctaaccctggccctcaactcgaatccgaaccctgac\n",
            "cctagcccgaacccgaaccctcgTGGGTTAGGAATggattagggttagggttaggaagaaacaagggttacgaaaagaat\n",
            "tagggttagggttaagaagaaagaagggttacgaaaagaattagggttagggttaggaagaagggttacgaaaagaatca\n",
            "gggttagggttaagaagaaagaagggttatACTAACCTGAAccgaagcatgaacccgaaccctagccctaaccctggccc\n",
            "tcaactcgaacccgaACCTGAAGCACGgaccctcaactcgaatccgaaccctgaccctagcCCTAGCCCGAACCCTCGAG\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import re\n",
        "\n",
        "\n",
        "def gq_search(name, seq):\n",
        "    re_pos = r\"(?:G{3,}[ATGC]{1,7}){3,}G{3,}\"\n",
        "    re_neg = r\"(?:C{3,}[ATGC]{1,7}){3,}C{3,}\"\n",
        "    PQS_pos = [[m.start(), m.end(), m.group(0)] for m in re.finditer(re_pos, seq, re.IGNORECASE)]\n",
        "    PQS_neg = [[m.start(), m.end(), m.group(0)] for m in re.finditer(re_neg, seq, re.IGNORECASE)]\n",
        "    print(f\"{'scaffold':>10}{'length':>10}{'plus':>10}{'minus':>10}\")\n",
        "    print(f\"{name:>10}{len(seq):>10}{len(PQS_pos):>10}{len(PQS_neg):>10}\")\n",
        "    PQS = PQS_pos + PQS_neg\n",
        "    res = ''\n",
        "    for p0, p1, _ in PQS_pos:\n",
        "        res += '\\t'.join([name, str(p0), str(p1), '+\\n'])\n",
        "    for n0, n1, _ in PQS_neg:\n",
        "        res += '\\t'.join([name, str(n0), str(n1), '-\\n'])\n",
        "    return res\n",
        "\n",
        "\n",
        "seq = []\n",
        "PQS = ''\n",
        "cnt = 0\n",
        "with open('genome.fna', 'r') as f:\n",
        "    for line in f.readlines():\n",
        "        if line.startswith('>'):\n",
        "            if seq != []:\n",
        "                seq_up = ''.join(seq).upper()\n",
        "                if cnt <= 1:\n",
        "                    PQS += gq_search('scaffold_' + str(cnt), seq_up)\n",
        "            seq = []\n",
        "            cnt += 1\n",
        "        else:\n",
        "            seq += line.split('\\n')[0]\n",
        "\n",
        "with open('GQ.bed', 'w') as f:\n",
        "    f.write(PQS)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "dz1M2TB8xxGc",
        "outputId": "a2ebf148-5415-4b0c-a735-26f286a13699"
      },
      "execution_count": 25,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "  scaffold    length      plus     minus\n",
            "scaffold_1   1966360        43        62\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "! head GQ.bed"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "K5SfJJN4URe7",
        "outputId": "8ad1fd52-b9c3-4ad6-ddad-74b052c08a84"
      },
      "execution_count": 26,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "scaffold_1\t153838\t153870\t+\n",
            "scaffold_1\t184130\t184155\t+\n",
            "scaffold_1\t258005\t258034\t+\n",
            "scaffold_1\t266501\t266542\t+\n",
            "scaffold_1\t275175\t275198\t+\n",
            "scaffold_1\t303176\t303206\t+\n",
            "scaffold_1\t306648\t306675\t+\n",
            "scaffold_1\t313069\t313092\t+\n",
            "scaffold_1\t339457\t339486\t+\n",
            "scaffold_1\t514445\t514477\t+\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### z-hunt"
      ],
      "metadata": {
        "id": "zJGS-0ToqCpx"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!chmod a+x zhunt3-alan.c\n",
        "!gcc zhunt3-alan.c -lm -o zhunt3 2> /dev/null"
      ],
      "metadata": {
        "id": "ls7JVIXEsFD-"
      },
      "execution_count": 27,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "seq = []\n",
        "cnt = 0\n",
        "\n",
        "with open('genome.fna', 'r') as f:\n",
        "    for line in f.readlines():\n",
        "        if line.startswith('>'):\n",
        "            if seq != []:\n",
        "                sequense = ''.join(seq).upper()\n",
        "                if cnt <= 1:\n",
        "                    with open('scaffold{}.txt'.format(cnt), 'w') as f:\n",
        "                        f.write(sequense)\n",
        "            seq = []\n",
        "            cnt += 1\n",
        "        else:\n",
        "            seq += line.split('\\n')[0]"
      ],
      "metadata": {
        "id": "ebtOMdUB8mTA"
      },
      "execution_count": 28,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!./zhunt3 12 8 12 scaffold1.txt"
      ],
      "metadata": {
        "id": "pcUWCgaisLHF",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "bc09b0ac-add7-4cd1-8225-dff9120d4ca9"
      },
      "execution_count": 24,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "dinucleotides 12\n",
            "min/max 8 12\n",
            "min/max 8 12\n",
            "operating on scaffold1.txt\n",
            "calculating zscore\n",
            "opening scaffold1.txt\n",
            "inputting sequence\n",
            "opening scaffold1.txt.Z-SCORE\n",
            "\n",
            " run time=519 sec\n",
            "use min/max 8 12\n",
            "analyzing_zscore\n",
            "opening scaffold1.txt.Z-SCORE\n",
            "opening scaffold1.txt\n",
            "inputting sequence\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!head scaffold1.txt.Z-SCORE -n5"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "FxxswoC6-PpP",
        "outputId": "c52f2574-b50a-4816-8ca2-dd1b069aff9b"
      },
      "execution_count": 29,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "scaffold1.txt 1966360 8 12\n",
            "1 17 16  29.183  20.075 2.318285e+00 ctgactaagcatccct   SASASASASASASASA\n",
            "2 18 16  30.211  23.440 4.188918e-01 tgactaagcatccctc   ASASASASASASASAS\n",
            "3 19 16  28.978  19.239 2.488467e+00 gactaagcatccctct   SASASASASASASASA\n",
            "4 20 16  31.199  19.872 2.849321e-01 actaagcatccctctt   SAASASASASASASAS\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "selected_lines=[]\n",
        "for i in range(1, 2):\n",
        "    with open(f'scaffold{i}.txt.Z-SCORE', 'r') as f:\n",
        "        for line in f.readlines()[1:]:\n",
        "            if float(line.split()[5]) >= 300.:\n",
        "                s0, s1, _, __, ___, s5, *____ = line.split()\n",
        "                selected_lines.append('\\t'.join([f'scaffold_{i}', s0, s1, s5, '\\n']))\n",
        "\n",
        "with open('zhunt.bed', 'w') as f:\n",
        "    f.write(''.join(selected_lines))\n",
        "\n",
        "!head zhunt.bed\n",
        "print()\n",
        "print(f\"Length: {len(selected_lines)}\")"
      ],
      "metadata": {
        "id": "U3ZjWiGHKdM3",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "da9352a3-fa01-4b12-9a00-758d61731d71"
      },
      "execution_count": 34,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "scaffold_1\t2310\t2334\t3.329764e+02\t\n",
            "scaffold_1\t2312\t2334\t4.394377e+02\t\n",
            "scaffold_1\t2314\t2334\t5.648193e+02\t\n",
            "scaffold_1\t2316\t2334\t7.163835e+02\t\n",
            "scaffold_1\t2317\t2335\t3.284535e+02\t\n",
            "scaffold_1\t2319\t2335\t4.198546e+02\t\n",
            "scaffold_1\t2321\t2337\t4.892471e+02\t\n",
            "scaffold_1\t2322\t2338\t9.815517e+02\t\n",
            "scaffold_1\t2323\t2339\t3.864327e+02\t\n",
            "scaffold_1\t2594\t2616\t3.541938e+02\t\n",
            "\n",
            "Length: 13855\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### ZDNABERT\n",
        "(copied from https://colab.research.google.com/github/mitiau/Z-DNABERT/blob/main/ZDNA-prediction.ipynb#scrollTo=QLn6_9F0TFcJ )"
      ],
      "metadata": {
        "id": "J2dbrpYVB9By"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "#### installation"
      ],
      "metadata": {
        "id": "51ZUlu6yTKIU"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install transformers -q\n",
        "!pip install biopython -q"
      ],
      "metadata": {
        "id": "KS8TudVAB8d0"
      },
      "execution_count": 36,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import torch\n",
        "from torch import nn\n",
        "import transformers\n",
        "from transformers import BertTokenizer, BertForTokenClassification\n",
        "import numpy as np\n",
        "from Bio import SeqIO\n",
        "from io import StringIO, BytesIO\n",
        "from google.colab import drive, files\n",
        "from tqdm import tqdm\n",
        "import pickle\n",
        "import scipy\n",
        "from scipy import ndimage"
      ],
      "metadata": {
        "id": "vjwGHJwWTPxz"
      },
      "execution_count": 37,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def seq2kmer(seq, k):\n",
        "    kmer = [seq[x:x + k] for x in range(len(seq) + 1 - k)]\n",
        "    return kmer\n",
        "\n",
        "def split_seq(seq, length=512, pad=16):\n",
        "    return [\n",
        "        seq[start:min(start + 512, len(seq))]\n",
        "        for start in range(0, len(seq), length - pad)\n",
        "    ]\n",
        "\n",
        "def stitch_np_seq(np_seqs, pad=16):\n",
        "    res = np.array([])\n",
        "    for seq in np_seqs:\n",
        "        res = res[:-pad]\n",
        "        res = np.concatenate([res,seq])\n",
        "    return res"
      ],
      "metadata": {
        "id": "nK_aXuHbTQt_"
      },
      "execution_count": 39,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "#### model and parameters"
      ],
      "metadata": {
        "id": "G4umSpnRThGG"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "model = 'HG kouzine'\n",
        "model_confidence_threshold = 0.5\n",
        "minimum_sequence_length = 10"
      ],
      "metadata": {
        "id": "eGmakBBaTeAQ"
      },
      "execution_count": 40,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "if model == 'HG chipseq':\n",
        "    model_id = '1VAsp8I904y_J0PUhAQqpSlCn1IqfG0FB'\n",
        "elif model == 'HG kouzine':\n",
        "    model_id = '1dAeAt5Gu2cadwDhbc7OnenUgDLHlUvkx'\n",
        "elif model == 'MM curax':\n",
        "    model_id = '1W6GEgHNoitlB-xXJbLJ_jDW4BF35W1Sd'\n",
        "elif model == 'MM kouzine':\n",
        "    model_id = '1dXpQFmheClKXIEoqcZ7kgCwx6hzVCv3H'"
      ],
      "metadata": {
        "id": "WDbn5wjUTQiF"
      },
      "execution_count": 41,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!gdown $model_id\n",
        "!gdown 10sF8Ywktd96HqAL0CwvlZZUUGj05CGk5\n",
        "!gdown 16bT7HDv71aRwyh3gBUbKwign1mtyLD2d\n",
        "!gdown 1EE9goZ2JRSD8UTx501q71lGCk-CK3kqG\n",
        "!gdown 1gZZdtAoDnDiLQqjQfGyuwt268Pe5sXW0\n",
        "\n",
        "\n",
        "!mkdir 6-new-12w-0\n",
        "!mv pytorch_model.bin 6-new-12w-0/\n",
        "!mv config.json 6-new-12w-0/\n",
        "!mv special_tokens_map.json 6-new-12w-0/\n",
        "!mv tokenizer_config.json 6-new-12w-0/\n",
        "!mv vocab.txt 6-new-12w-0/"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "5WqGII1hTvBl",
        "outputId": "faaf7740-6e89-4491-87c3-55fdb4c46e9a"
      },
      "execution_count": 42,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Downloading...\n",
            "From (original): https://drive.google.com/uc?id=1dAeAt5Gu2cadwDhbc7OnenUgDLHlUvkx\n",
            "From (redirected): https://drive.google.com/uc?id=1dAeAt5Gu2cadwDhbc7OnenUgDLHlUvkx&confirm=t&uuid=6bdd3e09-ed73-416f-a6c7-df6514105713\n",
            "To: /content/pytorch_model.bin\n",
            "100% 354M/354M [00:05<00:00, 68.3MB/s]\n",
            "Downloading...\n",
            "From: https://drive.google.com/uc?id=10sF8Ywktd96HqAL0CwvlZZUUGj05CGk5\n",
            "To: /content/config.json\n",
            "100% 634/634 [00:00<00:00, 4.31MB/s]\n",
            "Downloading...\n",
            "From: https://drive.google.com/uc?id=16bT7HDv71aRwyh3gBUbKwign1mtyLD2d\n",
            "To: /content/special_tokens_map.json\n",
            "100% 112/112 [00:00<00:00, 743kB/s]\n",
            "Downloading...\n",
            "From: https://drive.google.com/uc?id=1EE9goZ2JRSD8UTx501q71lGCk-CK3kqG\n",
            "To: /content/tokenizer_config.json\n",
            "100% 40.0/40.0 [00:00<00:00, 269kB/s]\n",
            "Downloading...\n",
            "From: https://drive.google.com/uc?id=1gZZdtAoDnDiLQqjQfGyuwt268Pe5sXW0\n",
            "To: /content/vocab.txt\n",
            "100% 28.7k/28.7k [00:00<00:00, 57.9MB/s]\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "tokenizer = BertTokenizer.from_pretrained('6-new-12w-0/')\n",
        "model = BertForTokenClassification.from_pretrained('6-new-12w-0/')\n",
        "model.to('cuda')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "xNNu9L6aTvyA",
        "outputId": "a3ea762e-65ff-4306-d6d0-efd72d50ca67"
      },
      "execution_count": 43,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "BertForTokenClassification(\n",
              "  (bert): BertModel(\n",
              "    (embeddings): BertEmbeddings(\n",
              "      (word_embeddings): Embedding(4101, 768, padding_idx=0)\n",
              "      (position_embeddings): Embedding(512, 768)\n",
              "      (token_type_embeddings): Embedding(2, 768)\n",
              "      (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)\n",
              "      (dropout): Dropout(p=0.1, inplace=False)\n",
              "    )\n",
              "    (encoder): BertEncoder(\n",
              "      (layer): ModuleList(\n",
              "        (0-11): 12 x BertLayer(\n",
              "          (attention): BertAttention(\n",
              "            (self): BertSdpaSelfAttention(\n",
              "              (query): Linear(in_features=768, out_features=768, bias=True)\n",
              "              (key): Linear(in_features=768, out_features=768, bias=True)\n",
              "              (value): Linear(in_features=768, out_features=768, bias=True)\n",
              "              (dropout): Dropout(p=0.1, inplace=False)\n",
              "            )\n",
              "            (output): BertSelfOutput(\n",
              "              (dense): Linear(in_features=768, out_features=768, bias=True)\n",
              "              (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)\n",
              "              (dropout): Dropout(p=0.1, inplace=False)\n",
              "            )\n",
              "          )\n",
              "          (intermediate): BertIntermediate(\n",
              "            (dense): Linear(in_features=768, out_features=3072, bias=True)\n",
              "            (intermediate_act_fn): GELUActivation()\n",
              "          )\n",
              "          (output): BertOutput(\n",
              "            (dense): Linear(in_features=3072, out_features=768, bias=True)\n",
              "            (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)\n",
              "            (dropout): Dropout(p=0.1, inplace=False)\n",
              "          )\n",
              "        )\n",
              "      )\n",
              "    )\n",
              "  )\n",
              "  (dropout): Dropout(p=0.1, inplace=False)\n",
              "  (classifier): Linear(in_features=768, out_features=2, bias=True)\n",
              ")"
            ]
          },
          "metadata": {},
          "execution_count": 43
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#### my data"
      ],
      "metadata": {
        "id": "Cs-LaRxyT5o9"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "genomic_file = 'GCF_000151665.1_ASM15166v1_genomic.fna.gz'\n",
        "\n",
        "!wget {base_path}/{genomic_file} -q && gunzip -d {genomic_file} -q && mv {genomic_file[:-3]} genome.fna"
      ],
      "metadata": {
        "id": "axieArQuT6-s"
      },
      "execution_count": 45,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "seq = []\n",
        "cnt = 0\n",
        "\n",
        "with open('genome.fna', 'r') as f:\n",
        "    for i in f.readlines():\n",
        "        if i.startswith('>'):\n",
        "            if seq != []:\n",
        "                sequence = ''.join(seq).lower()\n",
        "                if cnt <= 1:\n",
        "                    with open('scaffold{}.txt'.format(cnt), 'w') as f:\n",
        "                        f.write('>scaffold1\\n' + sequence)\n",
        "            seq = [i]\n",
        "            cnt += 1\n",
        "        else:\n",
        "            seq += [i]\n",
        "\n",
        "name = 'scaffold1.txt'\n",
        "with open(name, 'rb') as f:\n",
        "    uploaded={name: f.read()}"
      ],
      "metadata": {
        "id": "YzcxnNr6UtQY"
      },
      "execution_count": 46,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "#### Predict and save raw outputs"
      ],
      "metadata": {
        "id": "-0jR1zLFT2Ht"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import pickle\n",
        "from tqdm import tqdm\n",
        "import numpy as np\n",
        "from io import StringIO, BytesIO\n",
        "import scipy.ndimage\n",
        "from Bio import SeqIO\n",
        "import torch\n",
        "\n",
        "output_lines = []\n",
        "\n",
        "for filename, file_data in uploaded.items():\n",
        "    print(f\"PROCESSING: {filename}\")\n",
        "    output_lines.append(filename)\n",
        "    results = {}\n",
        "    seq_records = SeqIO.parse(StringIO(BytesIO(file_data).read().decode('UTF-8')), 'fasta')\n",
        "\n",
        "    for record in seq_records:\n",
        "        sequence = str(record.seq).upper()\n",
        "        kmer_sequence = seq2kmer(sequence, 6)\n",
        "        split_sequences = split_seq(kmer_sequence)\n",
        "\n",
        "        output_lines.append(record.name)\n",
        "\n",
        "        with torch.no_grad():\n",
        "            predictions = []\n",
        "\n",
        "            for split_seq_piece in tqdm(split_sequences):\n",
        "                input_ids = torch.LongTensor(tokenizer.encode(' '.join(split_seq_piece), add_special_tokens=False))\n",
        "                output = model(input_ids.to('cuda').unsqueeze(0))[-1]\n",
        "                probabilities = torch.softmax(output, axis=-1)[0, :, 1]\n",
        "                predictions.append(probabilities.cpu().numpy())\n",
        "\n",
        "        stitched_preds = stitch_np_seq(predictions)\n",
        "        results[record.name] = stitched_preds\n",
        "\n",
        "        labeled_regions, max_region_label = scipy.ndimage.label(stitched_preds > model_confidence_threshold)\n",
        "        output_lines.append('  start     end')\n",
        "\n",
        "        for region_label in range(1, max_region_label + 1):\n",
        "            region_indices = np.where(labeled_regions == region_label)[0]\n",
        "            if len(region_indices) > minimum_sequence_length:\n",
        "                start_pos = region_indices[0]\n",
        "                end_pos = region_indices[-1]\n",
        "                print(f'{start_pos:8} {end_pos:8}')\n",
        "                output_lines.append(f'{start_pos:8} {end_pos:8}')\n",
        "\n",
        "    with open(f\"{filename}.preds.pkl\", \"wb\") as file_handle:\n",
        "        pickle.dump(results, file_handle)\n",
        "\n",
        "    print()\n",
        "\n",
        "with open('text_predictions.txt', \"w\") as file_handle:\n",
        "    for line in output_lines:\n",
        "        file_handle.write(f\"{line}\\n\")"
      ],
      "metadata": {
        "collapsed": true,
        "id": "UXQq6DxYT9Xq"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!head text_predictions.txt"
      ],
      "metadata": {
        "id": "6bSqVQN3WJ24",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "c4cf95ba-28cb-4f64-d5db-7089dacd9862"
      },
      "execution_count": 50,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "scaffold1.txt\n",
            "scaffold1\n",
            "  start     end\n",
            "nw_013171815.1\n",
            "  start     end\n",
            "    6091    6108\n",
            "   40644   40657\n",
            "   67868   67879\n",
            "   73666   73682\n",
            "   74987   74997\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#### Download text file with predictions"
      ],
      "metadata": {
        "id": "Ug-B7RYUUBfi"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "res = []\n",
        "\n",
        "with open('text_predictions.txt', 'r') as f:\n",
        "    for line in f.readlines()[5:]:\n",
        "        res.append('\\t'.join(['scaffold_1', line.split()[0], line.split()[1], '\\n']))\n",
        "with open('zdnabert_scaffold_1.bed', 'w') as f:\n",
        "    f.write(''.join(res))"
      ],
      "metadata": {
        "id": "sPraCXToZLMT"
      },
      "execution_count": 51,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "! head zdnabert_scaffold_1.bed"
      ],
      "metadata": {
        "id": "BypWcczKaBV4",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "2b507ca5-4267-471d-c60d-4006ec0fac7d"
      },
      "execution_count": 52,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "scaffold_1\t6091\t6108\t\n",
            "scaffold_1\t40644\t40657\t\n",
            "scaffold_1\t67868\t67879\t\n",
            "scaffold_1\t73666\t73682\t\n",
            "scaffold_1\t74987\t74997\t\n",
            "scaffold_1\t75509\t75524\t\n",
            "scaffold_1\t75537\t75554\t\n",
            "scaffold_1\t75618\t75634\t\n",
            "scaffold_1\t75702\t75712\t\n",
            "scaffold_1\t75969\t75983\t\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "files.download('text_predictions.txt')"
      ],
      "metadata": {
        "id": "7tg6I0YwUBHp",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 17
        },
        "outputId": "7335abf0-84db-4e33-b414-97b9446e2674"
      },
      "execution_count": 53,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ],
            "application/javascript": [
              "\n",
              "    async function download(id, filename, size) {\n",
              "      if (!google.colab.kernel.accessAllowed) {\n",
              "        return;\n",
              "      }\n",
              "      const div = document.createElement('div');\n",
              "      const label = document.createElement('label');\n",
              "      label.textContent = `Downloading \"${filename}\": `;\n",
              "      div.appendChild(label);\n",
              "      const progress = document.createElement('progress');\n",
              "      progress.max = size;\n",
              "      div.appendChild(progress);\n",
              "      document.body.appendChild(div);\n",
              "\n",
              "      const buffers = [];\n",
              "      let downloaded = 0;\n",
              "\n",
              "      const channel = await google.colab.kernel.comms.open(id);\n",
              "      // Send a message to notify the kernel that we're ready.\n",
              "      channel.send({})\n",
              "\n",
              "      for await (const message of channel.messages) {\n",
              "        // Send a message to notify the kernel that we're ready.\n",
              "        channel.send({})\n",
              "        if (message.buffers) {\n",
              "          for (const buffer of message.buffers) {\n",
              "            buffers.push(buffer);\n",
              "            downloaded += buffer.byteLength;\n",
              "            progress.value = downloaded;\n",
              "          }\n",
              "        }\n",
              "      }\n",
              "      const blob = new Blob(buffers, {type: 'application/binary'});\n",
              "      const a = document.createElement('a');\n",
              "      a.href = window.URL.createObjectURL(blob);\n",
              "      a.download = filename;\n",
              "      div.appendChild(a);\n",
              "      a.click();\n",
              "      div.remove();\n",
              "    }\n",
              "  "
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ],
            "application/javascript": [
              "download(\"download_b5d48223-0ae0-4c2f-b5c8-91a2b6a5c968\", \"text_predictions.txt\", 20097)"
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#### Download raw prediction files in numpy format"
      ],
      "metadata": {
        "id": "zyylrNnSUCVn"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "for key in uploaded.keys():\n",
        "    files.download(key + '.preds.pkl')"
      ],
      "metadata": {
        "id": "37Z_bPdJUCEP",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 17
        },
        "outputId": "2dadab01-163d-4279-803a-d14ad76b5fa6"
      },
      "execution_count": 54,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ],
            "application/javascript": [
              "\n",
              "    async function download(id, filename, size) {\n",
              "      if (!google.colab.kernel.accessAllowed) {\n",
              "        return;\n",
              "      }\n",
              "      const div = document.createElement('div');\n",
              "      const label = document.createElement('label');\n",
              "      label.textContent = `Downloading \"${filename}\": `;\n",
              "      div.appendChild(label);\n",
              "      const progress = document.createElement('progress');\n",
              "      progress.max = size;\n",
              "      div.appendChild(progress);\n",
              "      document.body.appendChild(div);\n",
              "\n",
              "      const buffers = [];\n",
              "      let downloaded = 0;\n",
              "\n",
              "      const channel = await google.colab.kernel.comms.open(id);\n",
              "      // Send a message to notify the kernel that we're ready.\n",
              "      channel.send({})\n",
              "\n",
              "      for await (const message of channel.messages) {\n",
              "        // Send a message to notify the kernel that we're ready.\n",
              "        channel.send({})\n",
              "        if (message.buffers) {\n",
              "          for (const buffer of message.buffers) {\n",
              "            buffers.push(buffer);\n",
              "            downloaded += buffer.byteLength;\n",
              "            progress.value = downloaded;\n",
              "          }\n",
              "        }\n",
              "      }\n",
              "      const blob = new Blob(buffers, {type: 'application/binary'});\n",
              "      const a = document.createElement('a');\n",
              "      a.href = window.URL.createObjectURL(blob);\n",
              "      a.download = filename;\n",
              "      div.appendChild(a);\n",
              "      a.click();\n",
              "      div.remove();\n",
              "    }\n",
              "  "
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ],
            "application/javascript": [
              "download(\"download_b4bca309-bb59-450c-9115-bb4ce8fba1f5\", \"scaffold1.txt.preds.pkl\", 15731065)"
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### alignment"
      ],
      "metadata": {
        "id": "YczaA-JU-ZmD"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "bedtools https://colab.research.google.com/drive/12KTBm9wAQGoLmKjPIkublh6zjw4njlRl?usp=sharing"
      ],
      "metadata": {
        "id": "IJceXtBSkmjh"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "#### download bedtools"
      ],
      "metadata": {
        "id": "RJ_rKsJ0aZPV"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!wget github.com/arq5x/bedtools2/releases/download/v2.30.0/bedtools-2.30.0.tar.gz\n",
        "!tar -zxvf bedtools-2.30.0.tar.gz"
      ],
      "metadata": {
        "id": "9UD9LxwjacZG",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "ddcc3cc8-556c-4946-bfb2-689f876bb102"
      },
      "execution_count": 55,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2024-06-16 23:05:29--  http://github.com/arq5x/bedtools2/releases/download/v2.30.0/bedtools-2.30.0.tar.gz\n",
            "Resolving github.com (github.com)... 140.82.116.3\n",
            "Connecting to github.com (github.com)|140.82.116.3|:80... connected.\n",
            "HTTP request sent, awaiting response... 301 Moved Permanently\n",
            "Location: https://github.com/arq5x/bedtools2/releases/download/v2.30.0/bedtools-2.30.0.tar.gz [following]\n",
            "--2024-06-16 23:05:29--  https://github.com/arq5x/bedtools2/releases/download/v2.30.0/bedtools-2.30.0.tar.gz\n",
            "Connecting to github.com (github.com)|140.82.116.3|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/15059334/aecc9080-5d79-11eb-847e-0ff72e8f556a?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20240616%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20240616T230529Z&X-Amz-Expires=300&X-Amz-Signature=2910525cb6037e90c1a0ca7ae49f31b42edc37f7f8afc9241592a0b71fe3e298&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=15059334&response-content-disposition=attachment%3B%20filename%3Dbedtools-2.30.0.tar.gz&response-content-type=application%2Foctet-stream [following]\n",
            "--2024-06-16 23:05:29--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/15059334/aecc9080-5d79-11eb-847e-0ff72e8f556a?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20240616%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20240616T230529Z&X-Amz-Expires=300&X-Amz-Signature=2910525cb6037e90c1a0ca7ae49f31b42edc37f7f8afc9241592a0b71fe3e298&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=15059334&response-content-disposition=attachment%3B%20filename%3Dbedtools-2.30.0.tar.gz&response-content-type=application%2Foctet-stream\n",
            "Resolving objects.githubusercontent.com (objects.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...\n",
            "Connecting to objects.githubusercontent.com (objects.githubusercontent.com)|185.199.108.133|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 20648577 (20M) [application/octet-stream]\n",
            "Saving to: ‘bedtools-2.30.0.tar.gz’\n",
            "\n",
            "bedtools-2.30.0.tar 100%[===================>]  19.69M  --.-KB/s    in 0.1s    \n",
            "\n",
            "2024-06-16 23:05:30 (180 MB/s) - ‘bedtools-2.30.0.tar.gz’ saved [20648577/20648577]\n",
            "\n",
            "bedtools2/\n",
            "bedtools2/obj/\n",
            "bedtools2/tutorial/\n",
            "bedtools2/LICENSE\n",
            "bedtools2/test/\n",
            "bedtools2/bin/\n",
            "bedtools2/Makefile\n",
            "bedtools2/docs/\n",
            "bedtools2/README.md\n",
            "bedtools2/genomes/\n",
            "bedtools2/.gitignore\n",
            "bedtools2/scripts/\n",
            "bedtools2/.github/\n",
            "bedtools2/data/\n",
            "bedtools2/src/\n",
            "bedtools2/src/spacingFile/\n",
            "bedtools2/src/sortBed/\n",
            "bedtools2/src/split/\n",
            "bedtools2/src/bedToBam/\n",
            "bedtools2/src/complementFile/\n",
            "bedtools2/src/expand/\n",
            "bedtools2/src/fisher/\n",
            "bedtools2/src/linksBed/\n",
            "bedtools2/src/pairToPair/\n",
            "bedtools2/src/bedtools.cpp\n",
            "bedtools2/src/mapFile/\n",
            "bedtools2/src/slopBed/\n",
            "bedtools2/src/annotateBed/\n",
            "bedtools2/src/randomBed/\n",
            "bedtools2/src/getOverlap/\n",
            "bedtools2/src/reldist/\n",
            "bedtools2/src/nucBed/\n",
            "bedtools2/src/bedToIgv/\n",
            "bedtools2/src/regressTest/\n",
            "bedtools2/src/maskFastaFromBed/\n",
            "bedtools2/src/genomeCoverageBed/\n",
            "bedtools2/src/utils/\n",
            "bedtools2/src/windowBed/\n",
            "bedtools2/src/jaccard/\n",
            "bedtools2/src/groupBy/\n",
            "bedtools2/src/bamToFastq/\n",
            "bedtools2/src/htsutil.cpp\n",
            "bedtools2/src/fastaFromBed/\n",
            "bedtools2/src/nekSandbox1/\n",
            "bedtools2/src/unionBedGraphs/\n",
            "bedtools2/src/bed12ToBed6/\n",
            "bedtools2/src/subtractFile/\n",
            "bedtools2/src/bedpeToBam/\n",
            "bedtools2/src/windowMaker/\n",
            "bedtools2/src/coverageFile/\n",
            "bedtools2/src/summaryFile/\n",
            "bedtools2/src/multiIntersectBed/\n",
            "bedtools2/src/intersectFile/\n",
            "bedtools2/src/closestFile/\n",
            "bedtools2/src/shuffleBed/\n",
            "bedtools2/src/flankBed/\n",
            "bedtools2/src/tagBam/\n",
            "bedtools2/src/multiBamCov/\n",
            "bedtools2/src/sampleFile/\n",
            "bedtools2/src/clusterBed/\n",
            "bedtools2/src/shiftBed/\n",
            "bedtools2/src/mergeFile/\n",
            "bedtools2/src/bamToBed/\n",
            "bedtools2/src/pairToBed/\n",
            "bedtools2/src/pairToBed/pairToBed.h\n",
            "bedtools2/src/pairToBed/pairToBedMain.cpp\n",
            "bedtools2/src/pairToBed/Makefile.frag\n",
            "bedtools2/src/pairToBed/pairToBed.cpp\n",
            "bedtools2/src/bamToBed/bamToBed.cpp\n",
            "bedtools2/src/bamToBed/Makefile.frag\n",
            "bedtools2/src/mergeFile/mergeFile.cpp\n",
            "bedtools2/src/mergeFile/Makefile.frag\n",
            "bedtools2/src/mergeFile/mergeHelp.cpp\n",
            "bedtools2/src/mergeFile/mergeFile.h\n",
            "bedtools2/src/shiftBed/shiftBed.cpp\n",
            "bedtools2/src/shiftBed/shiftBedMain.cpp\n",
            "bedtools2/src/shiftBed/Makefile.frag\n",
            "bedtools2/src/shiftBed/shiftBed.h\n",
            "bedtools2/src/clusterBed/clusterBed.h\n",
            "bedtools2/src/clusterBed/clusterBed.cpp\n",
            "bedtools2/src/clusterBed/clusterMain.cpp\n",
            "bedtools2/src/clusterBed/Makefile.frag\n",
            "bedtools2/src/sampleFile/sampleFile.cpp\n",
            "bedtools2/src/sampleFile/sampleFile.h\n",
            "bedtools2/src/sampleFile/Makefile.frag\n",
            "bedtools2/src/sampleFile/sampleHelp.cpp\n",
            "bedtools2/src/multiBamCov/multiBamCov.cpp\n",
            "bedtools2/src/multiBamCov/Makefile.frag\n",
            "bedtools2/src/multiBamCov/multiBamCov.h\n",
            "bedtools2/src/multiBamCov/multiBamCovMain.cpp\n",
            "bedtools2/src/tagBam/tagBam.cpp\n",
            "bedtools2/src/tagBam/tagBamMain.cpp\n",
            "bedtools2/src/tagBam/tagBam.h\n",
            "bedtools2/src/tagBam/Makefile.frag\n",
            "bedtools2/src/flankBed/flankBed.cpp\n",
            "bedtools2/src/flankBed/flankBedMain.cpp\n",
            "bedtools2/src/flankBed/test.fa.fai\n",
            "bedtools2/src/flankBed/test.gff\n",
            "bedtools2/src/flankBed/Makefile.frag\n",
            "bedtools2/src/flankBed/test.fa\n",
            "bedtools2/src/flankBed/flankBed.h\n",
            "bedtools2/src/shuffleBed/shuffleBed.h\n",
            "bedtools2/src/shuffleBed/shuffleBed.cpp\n",
            "bedtools2/src/shuffleBed/Makefile.frag\n",
            "bedtools2/src/shuffleBed/shuffleBedMain.cpp\n",
            "bedtools2/src/closestFile/closestHelp.cpp\n",
            "bedtools2/src/closestFile/closestFile.h\n",
            "bedtools2/src/closestFile/closestFile.cpp\n",
            "bedtools2/src/closestFile/Makefile.frag\n",
            "bedtools2/src/intersectFile/intersectHelp.cpp\n",
            "bedtools2/src/intersectFile/Makefile.frag\n",
            "bedtools2/src/intersectFile/intersectFile.cpp\n",
            "bedtools2/src/intersectFile/intersectFile.h\n",
            "bedtools2/src/multiIntersectBed/multiIntersectBed.cpp\n",
            "bedtools2/src/multiIntersectBed/multiIntersectBed.h\n",
            "bedtools2/src/multiIntersectBed/Makefile.frag\n",
            "bedtools2/src/multiIntersectBed/multiIntersectBedMain.cpp\n",
            "bedtools2/src/summaryFile/summaryHelp.cpp\n",
            "bedtools2/src/summaryFile/Makefile.frag\n",
            "bedtools2/src/summaryFile/summaryFile.h\n",
            "bedtools2/src/summaryFile/summaryFile.cpp\n",
            "bedtools2/src/coverageFile/coverageFile.cpp\n",
            "bedtools2/src/coverageFile/coverageFile.h\n",
            "bedtools2/src/coverageFile/Makefile.frag\n",
            "bedtools2/src/coverageFile/coverageHelp.cpp\n",
            "bedtools2/src/windowMaker/windowMakerMain.cpp\n",
            "bedtools2/src/windowMaker/windowMaker.h\n",
            "bedtools2/src/windowMaker/windowMaker.cpp\n",
            "bedtools2/src/windowMaker/Makefile.frag\n",
            "bedtools2/src/bedpeToBam/notes.txt\n",
            "bedtools2/src/bedpeToBam/bedpeToBam.cpp\n",
            "bedtools2/src/bedpeToBam/Makefile.frag\n",
            "bedtools2/src/subtractFile/subtractHelp.cpp\n",
            "bedtools2/src/subtractFile/subtractFile.h\n",
            "bedtools2/src/subtractFile/Makefile.frag\n",
            "bedtools2/src/subtractFile/subtractFile.cpp\n",
            "bedtools2/src/bed12ToBed6/bed12ToBed6.cpp\n",
            "bedtools2/src/bed12ToBed6/Makefile.frag\n",
            "bedtools2/src/unionBedGraphs/unionBedGraphs.cpp\n",
            "bedtools2/src/unionBedGraphs/unionBedGraphsMain.cpp\n",
            "bedtools2/src/unionBedGraphs/Makefile.frag\n",
            "bedtools2/src/unionBedGraphs/unionBedGraphs.h\n",
            "bedtools2/src/nekSandbox1/FormatGuess.h\n",
            "bedtools2/src/nekSandbox1/Makefile\n",
            "bedtools2/src/nekSandbox1/nekSandboxMain.cpp\n",
            "bedtools2/src/fastaFromBed/fastaFromBedMain.cpp\n",
            "bedtools2/src/fastaFromBed/fastaFromBed.cpp\n",
            "bedtools2/src/fastaFromBed/fastaFromBed.h\n",
            "bedtools2/src/fastaFromBed/Makefile.frag\n",
            "bedtools2/src/bamToFastq/bamToFastq.h\n",
            "bedtools2/src/bamToFastq/bamToFastq.cpp\n",
            "bedtools2/src/bamToFastq/bamToFastqMain.cpp\n",
            "bedtools2/src/bamToFastq/Makefile.frag\n",
            "bedtools2/src/groupBy/groupBy.cpp\n",
            "bedtools2/src/groupBy/groupBy.h\n",
            "bedtools2/src/groupBy/Makefile.frag\n",
            "bedtools2/src/groupBy/groupByHelp.cpp\n",
            "bedtools2/src/jaccard/jaccard.cpp\n",
            "bedtools2/src/jaccard/Makefile.frag\n",
            "bedtools2/src/jaccard/jaccardHelp.cpp\n",
            "bedtools2/src/jaccard/jaccard.h\n",
            "bedtools2/src/windowBed/windowBed.cpp\n",
            "bedtools2/src/windowBed/windowBed.h\n",
            "bedtools2/src/windowBed/windowMain.cpp\n",
            "bedtools2/src/windowBed/Makefile.frag\n",
            "bedtools2/src/utils/tabFile/\n",
            "bedtools2/src/utils/htslib/\n",
            "bedtools2/src/utils/NewChromsweep/\n",
            "bedtools2/src/utils/RecordOutputMgr/\n",
            "bedtools2/src/utils/bedFile/\n",
            "bedtools2/src/utils/Contexts/\n",
            "bedtools2/src/utils/BamTools/\n",
            "bedtools2/src/utils/BlockedIntervals/\n",
            "bedtools2/src/utils/general/\n",
            "bedtools2/src/utils/driver/\n",
            "bedtools2/src/utils/Point/\n",
            "bedtools2/src/utils/bedGraphFile/\n",
            "bedtools2/src/utils/sequenceUtilities/\n",
            "bedtools2/src/utils/Fasta/\n",
            "bedtools2/src/utils/gzstream/\n",
            "bedtools2/src/utils/fileType/\n",
            "bedtools2/src/utils/KeyListOps/\n",
            "bedtools2/src/utils/BamTools-Ancillary/\n",
            "bedtools2/src/utils/GenomeFile/\n",
            "bedtools2/src/utils/version/\n",
            "bedtools2/src/utils/ToolBase/\n",
            "bedtools2/src/utils/lineFileUtilities/\n",
            "bedtools2/src/utils/bedFilePE/\n",
            "bedtools2/src/utils/chromsweep/\n",
            "bedtools2/src/utils/FileRecordTools/\n",
            "bedtools2/src/utils/BinTree/\n",
            "bedtools2/src/utils/VectorOps/\n",
            "bedtools2/src/utils/stringUtilities/\n",
            "bedtools2/src/utils/stringUtilities/stringUtilities.h\n",
            "bedtools2/src/utils/VectorOps/VectorOps.cpp\n",
            "bedtools2/src/utils/VectorOps/Makefile.frag\n",
            "bedtools2/src/utils/VectorOps/VectorOps.h\n",
            "bedtools2/src/utils/BinTree/BinTree.cpp\n",
            "bedtools2/src/utils/BinTree/BinTree.h\n",
            "bedtools2/src/utils/BinTree/Makefile.frag\n",
            "bedtools2/src/utils/FileRecordTools/Records/\n",
            "bedtools2/src/utils/FileRecordTools/FileRecordMergeMgr.h\n",
            "bedtools2/src/utils/FileRecordTools/FileRecordMgr.cpp\n",
            "bedtools2/src/utils/FileRecordTools/Makefile.frag\n",
            "bedtools2/src/utils/FileRecordTools/FileRecordMgr.h\n",
            "bedtools2/src/utils/FileRecordTools/FileReaders/\n",
            "bedtools2/src/utils/FileRecordTools/FileRecordMergeMgr.cpp\n",
            "bedtools2/src/utils/FileRecordTools/FileReaders/InputStreamMgr.h\n",
            "bedtools2/src/utils/FileRecordTools/FileReaders/BamFileReader.h\n",
            "bedtools2/src/utils/FileRecordTools/FileReaders/BufferedStreamMgr.cpp\n",
            "bedtools2/src/utils/FileRecordTools/FileReaders/SingleLineDelimTextFileReader.cpp\n",
            "bedtools2/src/utils/FileRecordTools/FileReaders/SingleLineDelimTransferBuffer.cpp\n",
            "bedtools2/src/utils/FileRecordTools/FileReaders/SingleLineDelimTransferBuffer.h\n",
            "bedtools2/src/utils/FileRecordTools/FileReaders/InputStreamMgr.cpp\n",
            "bedtools2/src/utils/FileRecordTools/FileReaders/FileReader.cpp\n",
            "bedtools2/src/utils/FileRecordTools/FileReaders/BamFileReader.cpp\n",
            "bedtools2/src/utils/FileRecordTools/FileReaders/Makefile.frag\n",
            "bedtools2/src/utils/FileRecordTools/FileReaders/SingleLineDelimTextFileReader.h\n",
            "bedtools2/src/utils/FileRecordTools/FileReaders/FileReader.h\n",
            "bedtools2/src/utils/FileRecordTools/FileReaders/BufferedStreamMgr.h\n",
            "bedtools2/src/utils/FileRecordTools/Records/BedGraphInterval.cpp\n",
            "bedtools2/src/utils/FileRecordTools/Records/RecordMgr.cpp\n",
            "bedtools2/src/utils/FileRecordTools/Records/RecordList.cpp\n",
            "bedtools2/src/utils/FileRecordTools/Records/Bed12Interval.h\n",
            "bedtools2/src/utils/FileRecordTools/Records/GffRecord.h\n",
            "bedtools2/src/utils/FileRecordTools/Records/Bed4Interval.h\n",
            "bedtools2/src/utils/FileRecordTools/Records/BedGraphInterval.h\n",
            "bedtools2/src/utils/FileRecordTools/Records/RecordList.h\n",
            "bedtools2/src/utils/FileRecordTools/Records/BamRecord.cpp\n",
            "bedtools2/src/utils/FileRecordTools/Records/RecordKeyVector.cpp\n",
            "bedtools2/src/utils/FileRecordTools/Records/PlusFields.h\n",
            "bedtools2/src/utils/FileRecordTools/Records/StrandQueue.cpp\n",
            "bedtools2/src/utils/FileRecordTools/Records/Bed5Interval.h\n",
            "bedtools2/src/utils/FileRecordTools/Records/Record.h\n",
            "bedtools2/src/utils/FileRecordTools/Records/GffRecord.cpp\n",
            "bedtools2/src/utils/FileRecordTools/Records/Bed4Interval.cpp\n",
            "bedtools2/src/utils/FileRecordTools/Records/VcfRecord.h\n",
            "bedtools2/src/utils/FileRecordTools/Records/Bed3Interval.h\n",
            "bedtools2/src/utils/FileRecordTools/Records/RecordKeyList.h\n",
            "bedtools2/src/utils/FileRecordTools/Records/VcfRecord.cpp\n",
            "bedtools2/src/utils/FileRecordTools/Records/EmptyRecord.h\n",
            "bedtools2/src/utils/FileRecordTools/Records/Bed3Interval.cpp\n",
            "bedtools2/src/utils/FileRecordTools/Records/RecordKeyVector.h\n",
            "bedtools2/src/utils/FileRecordTools/Records/Bed6Interval.h\n",
            "bedtools2/src/utils/FileRecordTools/Records/BlockMgr.h\n",
            "bedtools2/src/utils/FileRecordTools/Records/RecordMgr.h\n",
            "bedtools2/src/utils/FileRecordTools/Records/Bed6Interval.cpp\n",
            "bedtools2/src/utils/FileRecordTools/Records/GffPlusRecord.h\n",
            "bedtools2/src/utils/FileRecordTools/Records/GffPlusRecord.cpp\n",
            "bedtools2/src/utils/FileRecordTools/Records/BedPlusInterval.cpp\n",
            "bedtools2/src/utils/FileRecordTools/Records/Makefile.frag\n",
            "bedtools2/src/utils/FileRecordTools/Records/StrandQueue.h\n",
            "bedtools2/src/utils/FileRecordTools/Records/Record.cpp\n",
            "bedtools2/src/utils/FileRecordTools/Records/Bed5Interval.cpp\n",
            "bedtools2/src/utils/FileRecordTools/Records/NoPosPlusRecord.h\n",
            "bedtools2/src/utils/FileRecordTools/Records/Bed12Interval.cpp\n",
            "bedtools2/src/utils/FileRecordTools/Records/BlockMgr.cpp\n",
            "bedtools2/src/utils/FileRecordTools/Records/RecordKeyList.cpp\n",
            "bedtools2/src/utils/FileRecordTools/Records/PlusFields.cpp\n",
            "bedtools2/src/utils/FileRecordTools/Records/BamRecord.h\n",
            "bedtools2/src/utils/FileRecordTools/Records/NoPosPlusRecord.cpp\n",
            "bedtools2/src/utils/FileRecordTools/Records/EmptyRecord.cpp\n",
            "bedtools2/src/utils/FileRecordTools/Records/BedPlusInterval.h\n",
            "bedtools2/src/utils/chromsweep/chromsweep.cpp\n",
            "bedtools2/src/utils/chromsweep/Makefile.frag\n",
            "bedtools2/src/utils/chromsweep/chromsweep.h\n",
            "bedtools2/src/utils/bedFilePE/bedFilePE.cpp\n",
            "bedtools2/src/utils/bedFilePE/Makefile.frag\n",
            "bedtools2/src/utils/bedFilePE/bedFilePE.h\n",
            "bedtools2/src/utils/lineFileUtilities/lineFileUtilities.h\n",
            "bedtools2/src/utils/ToolBase/ToolBase.cpp\n",
            "bedtools2/src/utils/ToolBase/Makefile.frag\n",
            "bedtools2/src/utils/ToolBase/ToolBase.h\n",
            "bedtools2/src/utils/version/version.h\n",
            "bedtools2/src/utils/version/version.cpp\n",
            "bedtools2/src/utils/version/version_release.txt\n",
            "bedtools2/src/utils/version/Makefile.frag\n",
            "bedtools2/src/utils/GenomeFile/NewGenomeFile.cpp\n",
            "bedtools2/src/utils/GenomeFile/NewGenomeFile.h\n",
            "bedtools2/src/utils/GenomeFile/GenomeFile.cpp\n",
            "bedtools2/src/utils/GenomeFile/Makefile.frag\n",
            "bedtools2/src/utils/GenomeFile/GenomeFile.h\n",
            "bedtools2/src/utils/BamTools-Ancillary/BamAncillary.h\n",
            "bedtools2/src/utils/BamTools-Ancillary/BamAncillary.cpp\n",
            "bedtools2/src/utils/BamTools-Ancillary/Makefile.frag\n",
            "bedtools2/src/utils/KeyListOps/KeyListOps.h\n",
            "bedtools2/src/utils/KeyListOps/KeyListOps.cpp\n",
            "bedtools2/src/utils/KeyListOps/KeyListOpsMethods.cpp\n",
            "bedtools2/src/utils/KeyListOps/Makefile.frag\n",
            "bedtools2/src/utils/KeyListOps/KeyListOpsMethods.h\n",
            "bedtools2/src/utils/fileType/fileType.cpp\n",
            "bedtools2/src/utils/fileType/FileRecordTypeChecker.h\n",
            "bedtools2/src/utils/fileType/FileRecordTypeChecker.cpp\n",
            "bedtools2/src/utils/fileType/Makefile.frag\n",
            "bedtools2/src/utils/fileType/fileType.h\n",
            "bedtools2/src/utils/gzstream/gzstream.C\n",
            "bedtools2/src/utils/gzstream/README\n",
            "bedtools2/src/utils/gzstream/COPYING.LIB\n",
            "bedtools2/src/utils/gzstream/Makefile.frag\n",
            "bedtools2/src/utils/gzstream/version.txt\n",
            "bedtools2/src/utils/gzstream/gzstream.h\n",
            "bedtools2/src/utils/Fasta/Fasta.h\n",
            "bedtools2/src/utils/Fasta/split.cpp\n",
            "bedtools2/src/utils/Fasta/Fasta.cpp\n",
            "bedtools2/src/utils/Fasta/LargeFileSupport.h\n",
            "bedtools2/src/utils/Fasta/Makefile.frag\n",
            "bedtools2/src/utils/Fasta/split.h\n",
            "bedtools2/src/utils/sequenceUtilities/sequenceUtils.h\n",
            "bedtools2/src/utils/sequenceUtilities/Makefile.frag\n",
            "bedtools2/src/utils/sequenceUtilities/sequenceUtils.cpp\n",
            "bedtools2/src/utils/bedGraphFile/bedGraphFile.h\n",
            "bedtools2/src/utils/bedGraphFile/Makefile.frag\n",
            "bedtools2/src/utils/bedGraphFile/bedGraphFile.cpp\n",
            "bedtools2/src/utils/Point/Point.h\n",
            "bedtools2/src/utils/driver/BedtoolsDriver.h\n",
            "bedtools2/src/utils/driver/BedtoolsDriver.cpp\n",
            "bedtools2/src/utils/driver/Makefile.frag\n",
            "bedtools2/src/utils/general/QuickString.cpp\n",
            "bedtools2/src/utils/general/ParseTools.cpp\n",
            "bedtools2/src/utils/general/CommonHelp.h\n",
            "bedtools2/src/utils/general/FreeList.h\n",
            "bedtools2/src/utils/general/CompressionTools.cpp\n",
            "bedtools2/src/utils/general/Random.cpp\n",
            "bedtools2/src/utils/general/BedtoolsTypes.h\n",
            "bedtools2/src/utils/general/ErrorMsg.h\n",
            "bedtools2/src/utils/general/PushBackStreamBuf.h\n",
            "bedtools2/src/utils/general/Tokenizer.h\n",
            "bedtools2/src/utils/general/ParseTools.h\n",
            "bedtools2/src/utils/general/QuickString.h\n",
            "bedtools2/src/utils/general/Tokenizer.cpp\n",
            "bedtools2/src/utils/general/ErrorMsg.cpp\n",
            "bedtools2/src/utils/general/CommonHelp.cpp\n",
            "bedtools2/src/utils/general/Makefile.frag\n",
            "bedtools2/src/utils/general/BTlist.h\n",
            "bedtools2/src/utils/general/CompressionTools.h\n",
            "bedtools2/src/utils/general/PushBackStreamBuf.cpp\n",
            "bedtools2/src/utils/general/InflateStreamBuf.h\n",
            "bedtools2/src/utils/general/Random.h\n",
            "bedtools2/src/utils/BlockedIntervals/BlockedIntervals.cpp\n",
            "bedtools2/src/utils/BlockedIntervals/BlockedIntervals.h\n",
            "bedtools2/src/utils/BlockedIntervals/Makefile.frag\n",
            "bedtools2/src/utils/BamTools/mapping/\n",
            "bedtools2/src/utils/BamTools/include/\n",
            "bedtools2/src/utils/BamTools/.gitignore\n",
            "bedtools2/src/utils/BamTools/Makefile.frag\n",
            "bedtools2/src/utils/BamTools/include/BamAncillary.h\n",
            "bedtools2/src/utils/BamTools/include/SamHeader.hpp\n",
            "bedtools2/src/utils/BamTools/include/BamReader.hpp\n",
            "bedtools2/src/utils/BamTools/include/api/\n",
            "bedtools2/src/utils/BamTools/include/BamWriter.hpp\n",
            "bedtools2/src/utils/BamTools/include/BamAlignment.mapping.hpp\n",
            "bedtools2/src/utils/BamTools/include/BamAlignment.hpp\n",
            "bedtools2/src/utils/BamTools/include/api/BamWriter.h\n",
            "bedtools2/src/utils/BamTools/include/api/BamMultiReader.h\n",
            "bedtools2/src/utils/BamTools/include/api/SamConstants.h\n",
            "bedtools2/src/utils/BamTools/include/api/BamAlignment.h\n",
            "bedtools2/src/utils/BamTools/include/api/BamReader.h\n",
            "bedtools2/src/utils/BamTools/include/api/BamAux.h\n",
            "bedtools2/src/utils/BamTools/mapping/BamAlignment.map\n",
            "bedtools2/src/utils/BamTools/mapping/BamAlignment.py\n",
            "bedtools2/src/utils/Contexts/ContextBase.h\n",
            "bedtools2/src/utils/Contexts/ContextClosest.h\n",
            "bedtools2/src/utils/Contexts/ContextIntersect.h\n",
            "bedtools2/src/utils/Contexts/ContextIntersect.cpp\n",
            "bedtools2/src/utils/Contexts/ContextFisher.cpp\n",
            "bedtools2/src/utils/Contexts/ContextMerge.h\n",
            "bedtools2/src/utils/Contexts/ContextGroupBy.h\n",
            "bedtools2/src/utils/Contexts/ContextComplement.h\n",
            "bedtools2/src/utils/Contexts/ContextSubtract.h\n",
            "bedtools2/src/utils/Contexts/ContextCoverage.cpp\n",
            "bedtools2/src/utils/Contexts/ContextBase.cpp\n",
            "bedtools2/src/utils/Contexts/ContextSummary.h\n",
            "bedtools2/src/utils/Contexts/ContextFisher.h\n",
            "bedtools2/src/utils/Contexts/ContextSpacing.cpp\n",
            "bedtools2/src/utils/Contexts/ContextMap.h\n",
            "bedtools2/src/utils/Contexts/ContextSubtract.cpp\n",
            "bedtools2/src/utils/Contexts/ContextJaccard.h\n",
            "bedtools2/src/utils/Contexts/ContextComplement.cpp\n",
            "bedtools2/src/utils/Contexts/ContextCoverage.h\n",
            "bedtools2/src/utils/Contexts/ContextSample.cpp\n",
            "bedtools2/src/utils/Contexts/ContextClosest.cpp\n",
            "bedtools2/src/utils/Contexts/Makefile.frag\n",
            "bedtools2/src/utils/Contexts/ContextJaccard.cpp\n",
            "bedtools2/src/utils/Contexts/ContextSample.h\n",
            "bedtools2/src/utils/Contexts/ContextSummary.cpp\n",
            "bedtools2/src/utils/Contexts/ContextGroupBy.cpp\n",
            "bedtools2/src/utils/Contexts/ContextMap.cpp\n",
            "bedtools2/src/utils/Contexts/ContextSpacing.h\n",
            "bedtools2/src/utils/Contexts/ContextMerge.cpp\n",
            "bedtools2/src/utils/bedFile/bedFile.h\n",
            "bedtools2/src/utils/bedFile/Makefile.frag\n",
            "bedtools2/src/utils/bedFile/bedFile.cpp\n",
            "bedtools2/src/utils/RecordOutputMgr/RecordOutputMgr.h\n",
            "bedtools2/src/utils/RecordOutputMgr/Makefile.frag\n",
            "bedtools2/src/utils/RecordOutputMgr/RecordOutputMgr.cpp\n",
            "bedtools2/src/utils/NewChromsweep/NewChromsweep.h\n",
            "bedtools2/src/utils/NewChromsweep/CloseSweep.h\n",
            "bedtools2/src/utils/NewChromsweep/Makefile.frag\n",
            "bedtools2/src/utils/NewChromsweep/NewChromsweep.cpp\n",
            "bedtools2/src/utils/NewChromsweep/CloseSweep.cpp\n",
            "bedtools2/src/utils/htslib/htslib.mk\n",
            "bedtools2/src/utils/htslib/cram/\n",
            "bedtools2/src/utils/htslib/htslib/\n",
            "bedtools2/src/utils/htslib/faidx.c\n",
            "bedtools2/src/utils/htslib/vcfutils.c\n",
            "bedtools2/src/utils/htslib/bcf_sr_sort.h\n",
            "bedtools2/src/utils/htslib/faidx.5\n",
            "bedtools2/src/utils/htslib/regidx.c\n",
            "bedtools2/src/utils/htslib/htslib.pc.in\n",
            "bedtools2/src/utils/htslib/configure.ac\n",
            "bedtools2/src/utils/htslib/INSTALL\n",
            "bedtools2/src/utils/htslib/hts_os.c\n",
            "bedtools2/src/utils/htslib/tbx.c\n",
            "bedtools2/src/utils/htslib/probaln.c\n",
            "bedtools2/src/utils/htslib/LICENSE\n",
            "bedtools2/src/utils/htslib/kfunc.c\n",
            "bedtools2/src/utils/htslib/hfile_gcs.c\n",
            "bedtools2/src/utils/htslib/errmod.c\n",
            "bedtools2/src/utils/htslib/multipart.c\n",
            "bedtools2/src/utils/htslib/hfile_internal.h\n",
            "bedtools2/src/utils/htslib/Makefile\n",
            "bedtools2/src/utils/htslib/htsfile.c\n",
            "bedtools2/src/utils/htslib/realn.c\n",
            "bedtools2/src/utils/htslib/bgzip.c\n",
            "bedtools2/src/utils/htslib/hfile_net.c\n",
            "bedtools2/src/utils/htslib/tabix.1\n",
            "bedtools2/src/utils/htslib/textutils.c\n",
            "bedtools2/src/utils/htslib/hts.c\n",
            "bedtools2/src/utils/htslib/README\n",
            "bedtools2/src/utils/htslib/vcf_sweep.c\n",
            "bedtools2/src/utils/htslib/README.md\n",
            "bedtools2/src/utils/htslib/sam.5\n",
            "bedtools2/src/utils/htslib/bcf_sr_sort.c\n",
            "bedtools2/src/utils/htslib/synced_bcf_reader.c\n",
            "bedtools2/src/utils/htslib/htslib_vars.mk\n",
            "bedtools2/src/utils/htslib/sam.c\n",
            "bedtools2/src/utils/htslib/NEWS\n",
            "bedtools2/src/utils/htslib/.gitignore\n",
            "bedtools2/src/utils/htslib/vcf.5\n",
            "bedtools2/src/utils/htslib/bgzf.c\n",
            "bedtools2/src/utils/htslib/.gitattributes\n",
            "bedtools2/src/utils/htslib/hts_internal.h\n",
            "bedtools2/src/utils/htslib/vcf.c\n",
            "bedtools2/src/utils/htslib/m4/\n",
            "bedtools2/src/utils/htslib/plugin.c\n",
            "bedtools2/src/utils/htslib/version.sh\n",
            "bedtools2/src/utils/htslib/hfile_libcurl.c\n",
            "bedtools2/src/utils/htslib/kstring.c\n",
            "bedtools2/src/utils/htslib/os/\n",
            "bedtools2/src/utils/htslib/md5.c\n",
            "bedtools2/src/utils/htslib/hfile_s3.c\n",
            "bedtools2/src/utils/htslib/htsfile.1\n",
            "bedtools2/src/utils/htslib/thread_pool.c\n",
            "bedtools2/src/utils/htslib/hfile.c\n",
            "bedtools2/src/utils/htslib/config.mk.in\n",
            "bedtools2/src/utils/htslib/tabix.c\n",
            "bedtools2/src/utils/htslib/textutils_internal.h\n",
            "bedtools2/src/utils/htslib/knetfile.c\n",
            "bedtools2/src/utils/htslib/bgzip.1\n",
            "bedtools2/src/utils/htslib/thread_pool_internal.h\n",
            "bedtools2/src/utils/htslib/os/rand.c\n",
            "bedtools2/src/utils/htslib/os/lzma_stub.h\n",
            "bedtools2/src/utils/htslib/m4/hts_prog_cc_warnings.m4\n",
            "bedtools2/src/utils/htslib/htslib/sam.h\n",
            "bedtools2/src/utils/htslib/htslib/synced_bcf_reader.h\n",
            "bedtools2/src/utils/htslib/htslib/vcf_sweep.h\n",
            "bedtools2/src/utils/htslib/htslib/vcf.h\n",
            "bedtools2/src/utils/htslib/htslib/khash_str2int.h\n",
            "bedtools2/src/utils/htslib/htslib/bgzf.h\n",
            "bedtools2/src/utils/htslib/htslib/cram.h\n",
            "bedtools2/src/utils/htslib/htslib/kstring.h\n",
            "bedtools2/src/utils/htslib/htslib/knetfile.h\n",
            "bedtools2/src/utils/htslib/htslib/hts_log.h\n",
            "bedtools2/src/utils/htslib/htslib/hfile.h\n",
            "bedtools2/src/utils/htslib/htslib/thread_pool.h\n",
            "bedtools2/src/utils/htslib/htslib/hts_defs.h\n",
            "bedtools2/src/utils/htslib/htslib/khash.h\n",
            "bedtools2/src/utils/htslib/htslib/regidx.h\n",
            "bedtools2/src/utils/htslib/htslib/faidx.h\n",
            "bedtools2/src/utils/htslib/htslib/vcfutils.h\n",
            "bedtools2/src/utils/htslib/htslib/kseq.h\n",
            "bedtools2/src/utils/htslib/htslib/kfunc.h\n",
            "bedtools2/src/utils/htslib/htslib/tbx.h\n",
            "bedtools2/src/utils/htslib/htslib/ksort.h\n",
            "bedtools2/src/utils/htslib/htslib/kbitset.h\n",
            "bedtools2/src/utils/htslib/htslib/hts_os.h\n",
            "bedtools2/src/utils/htslib/htslib/klist.h\n",
            "bedtools2/src/utils/htslib/htslib/hts.h\n",
            "bedtools2/src/utils/htslib/htslib/hts_endian.h\n",
            "bedtools2/src/utils/htslib/cram/cram_codecs.h\n",
            "bedtools2/src/utils/htslib/cram/sam_header.c\n",
            "bedtools2/src/utils/htslib/cram/cram_index.c\n",
            "bedtools2/src/utils/htslib/cram/open_trace_file.c\n",
            "bedtools2/src/utils/htslib/cram/mFILE.h\n",
            "bedtools2/src/utils/htslib/cram/cram_external.c\n",
            "bedtools2/src/utils/htslib/cram/cram.h\n",
            "bedtools2/src/utils/htslib/cram/misc.h\n",
            "bedtools2/src/utils/htslib/cram/files.c\n",
            "bedtools2/src/utils/htslib/cram/os.h\n",
            "bedtools2/src/utils/htslib/cram/rANS_static.h\n",
            "bedtools2/src/utils/htslib/cram/cram_samtools.c\n",
            "bedtools2/src/utils/htslib/cram/pooled_alloc.c\n",
            "bedtools2/src/utils/htslib/cram/string_alloc.c\n",
            "bedtools2/src/utils/htslib/cram/cram_io.h\n",
            "bedtools2/src/utils/htslib/cram/cram_decode.h\n",
            "bedtools2/src/utils/htslib/cram/cram_encode.h\n",
            "bedtools2/src/utils/htslib/cram/cram_stats.h\n",
            "bedtools2/src/utils/htslib/cram/mFILE.c\n",
            "bedtools2/src/utils/htslib/cram/cram_index.h\n",
            "bedtools2/src/utils/htslib/cram/sam_header.h\n",
            "bedtools2/src/utils/htslib/cram/open_trace_file.h\n",
            "bedtools2/src/utils/htslib/cram/cram_codecs.c\n",
            "bedtools2/src/utils/htslib/cram/rANS_byte.h\n",
            "bedtools2/src/utils/htslib/cram/string_alloc.h\n",
            "bedtools2/src/utils/htslib/cram/rANS_static.c\n",
            "bedtools2/src/utils/htslib/cram/pooled_alloc.h\n",
            "bedtools2/src/utils/htslib/cram/cram_samtools.h\n",
            "bedtools2/src/utils/htslib/cram/cram_stats.c\n",
            "bedtools2/src/utils/htslib/cram/cram_encode.c\n",
            "bedtools2/src/utils/htslib/cram/cram_structs.h\n",
            "bedtools2/src/utils/htslib/cram/cram_io.c\n",
            "bedtools2/src/utils/htslib/cram/cram_decode.c\n",
            "bedtools2/src/utils/tabFile/tabFile.cpp\n",
            "bedtools2/src/utils/tabFile/tabFile.h\n",
            "bedtools2/src/utils/tabFile/Makefile.frag\n",
            "bedtools2/src/genomeCoverageBed/genomeCoverageMain.cpp\n",
            "bedtools2/src/genomeCoverageBed/genomeCoverageBed.h\n",
            "bedtools2/src/genomeCoverageBed/Makefile.frag\n",
            "bedtools2/src/genomeCoverageBed/genomeCoverageBed.cpp\n",
            "bedtools2/src/maskFastaFromBed/maskFastaFromBed.cpp\n",
            "bedtools2/src/maskFastaFromBed/maskFastaFromBed.h\n",
            "bedtools2/src/maskFastaFromBed/maskFastaFromBedMain.cpp\n",
            "bedtools2/src/maskFastaFromBed/Makefile.frag\n",
            "bedtools2/src/regressTest/RegressTest.h\n",
            "bedtools2/src/regressTest/RegressTest.cpp\n",
            "bedtools2/src/regressTest/regressTestMain.cpp\n",
            "bedtools2/src/regressTest/Makefile.frag\n",
            "bedtools2/src/bedToIgv/bedToIgv.cpp\n",
            "bedtools2/src/bedToIgv/Makefile.frag\n",
            "bedtools2/src/nucBed/nucBed.h\n",
            "bedtools2/src/nucBed/nucBedMain.cpp\n",
            "bedtools2/src/nucBed/LargeFileSupport.h\n",
            "bedtools2/src/nucBed/Makefile.frag\n",
            "bedtools2/src/nucBed/nucBed.cpp\n",
            "bedtools2/src/reldist/reldist.cpp\n",
            "bedtools2/src/reldist/Makefile.frag\n",
            "bedtools2/src/reldist/reldistMain.cpp\n",
            "bedtools2/src/reldist/reldist.h\n",
            "bedtools2/src/getOverlap/getOverlap.cpp\n",
            "bedtools2/src/getOverlap/Makefile.frag\n",
            "bedtools2/src/randomBed/randomBedMain.cpp\n",
            "bedtools2/src/randomBed/randomBed.cpp\n",
            "bedtools2/src/randomBed/Makefile.frag\n",
            "bedtools2/src/randomBed/randomBed.h\n",
            "bedtools2/src/annotateBed/annotateBed.cpp\n",
            "bedtools2/src/annotateBed/annotateBed.h\n",
            "bedtools2/src/annotateBed/annotateMain.cpp\n",
            "bedtools2/src/annotateBed/Makefile.frag\n",
            "bedtools2/src/slopBed/slopBed.h\n",
            "bedtools2/src/slopBed/slopBed.cpp\n",
            "bedtools2/src/slopBed/Makefile.frag\n",
            "bedtools2/src/slopBed/slopBedMain.cpp\n",
            "bedtools2/src/mapFile/mapHelp.cpp\n",
            "bedtools2/src/mapFile/mapFile.cpp\n",
            "bedtools2/src/mapFile/Makefile.frag\n",
            "bedtools2/src/mapFile/mapFile.h\n",
            "bedtools2/src/pairToPair/pairToPair.h\n",
            "bedtools2/src/pairToPair/pairToPairMain.cpp\n",
            "bedtools2/src/pairToPair/pairToPair.cpp\n",
            "bedtools2/src/pairToPair/Makefile.frag\n",
            "bedtools2/src/linksBed/linksBed.cpp\n",
            "bedtools2/src/linksBed/Makefile.frag\n",
            "bedtools2/src/linksBed/linksBed.h\n",
            "bedtools2/src/linksBed/linksMain.cpp\n",
            "bedtools2/src/fisher/fisher.h\n",
            "bedtools2/src/fisher/kfunc.h\n",
            "bedtools2/src/fisher/kfunc.cpp\n",
            "bedtools2/src/fisher/Makefile.frag\n",
            "bedtools2/src/fisher/fisher.cpp\n",
            "bedtools2/src/fisher/fisherHelp.cpp\n",
            "bedtools2/src/expand/Makefile.frag\n",
            "bedtools2/src/expand/expand.cpp\n",
            "bedtools2/src/complementFile/complementHelp.cpp\n",
            "bedtools2/src/complementFile/complementFile.h\n",
            "bedtools2/src/complementFile/Makefile.frag\n",
            "bedtools2/src/complementFile/complementFile.cpp\n",
            "bedtools2/src/bedToBam/bedToBam.cpp\n",
            "bedtools2/src/bedToBam/Makefile.frag\n",
            "bedtools2/src/split/splitBed.h\n",
            "bedtools2/src/split/splitBed.cpp\n",
            "bedtools2/src/split/splitBedMain.cpp\n",
            "bedtools2/src/split/Makefile.frag\n",
            "bedtools2/src/sortBed/sortMain.cpp\n",
            "bedtools2/src/sortBed/sortBed.cpp\n",
            "bedtools2/src/sortBed/sortBed.h\n",
            "bedtools2/src/sortBed/Makefile.frag\n",
            "bedtools2/src/spacingFile/spacingHelp.cpp\n",
            "bedtools2/src/spacingFile/spacingFile.h\n",
            "bedtools2/src/spacingFile/Makefile.frag\n",
            "bedtools2/src/spacingFile/spacingFile.cpp\n",
            "bedtools2/data/gerp.chr1.bed.gz\n",
            "bedtools2/data/simpleRepeats.chr1.bed.gz\n",
            "bedtools2/data/aluY.chr1.bed.gz\n",
            "bedtools2/data/refseq.chr1.exons.bed.gz\n",
            "bedtools2/data/knownGene.hg18.chr21.short.bed\n",
            "bedtools2/data/knownGene.hg18.chr21.bed\n",
            "bedtools2/.github/workflows/\n",
            "bedtools2/.github/workflows/main.yml\n",
            "bedtools2/scripts/makeBashScripts.py\n",
            "bedtools2/scripts/vcfsort.sh\n",
            "bedtools2/genomes/human.hg38.genome\n",
            "bedtools2/genomes/mouse.mm10.genome\n",
            "bedtools2/genomes/mouse.mm8.genome\n",
            "bedtools2/genomes/human.hg18.genome\n",
            "bedtools2/genomes/human.hg19.genome\n",
            "bedtools2/genomes/mouse.mm9.genome\n",
            "bedtools2/docs/index.rst\n",
            "bedtools2/docs/Makefile\n",
            "bedtools2/docs/conf.py\n",
            "bedtools2/docs/_static/\n",
            "bedtools2/docs/content/\n",
            "bedtools2/docs/bedtools.swiss.png\n",
            "bedtools2/docs/build.sh\n",
            "bedtools2/docs/make.bat\n",
            "bedtools2/docs/bedtools.ai\n",
            "bedtools2/docs/templates/\n",
            "bedtools2/docs/themes/\n",
            "bedtools2/docs/bedtools.png\n",
            "bedtools2/docs/themes/rtd/\n",
            "bedtools2/docs/themes/rtd/theme.conf\n",
            "bedtools2/docs/themes/rtd/static/\n",
            "bedtools2/docs/themes/rtd/static/rtd.css_t\n",
            "bedtools2/docs/templates/page.html\n",
            "bedtools2/docs/templates/sidebar-intro.html\n",
            "bedtools2/docs/content/general-usage.rst\n",
            "bedtools2/docs/content/tools/\n",
            "bedtools2/docs/content/images/\n",
            "bedtools2/docs/content/quick-start.rst\n",
            "bedtools2/docs/content/example-usage.rst\n",
            "bedtools2/docs/content/overview.rst\n",
            "bedtools2/docs/content/advanced-usage.rst\n",
            "bedtools2/docs/content/history.rst\n",
            "bedtools2/docs/content/installation.rst\n",
            "bedtools2/docs/content/tips-and-tricks.rst\n",
            "bedtools2/docs/content/bedtools-suite.rst\n",
            "bedtools2/docs/content/faq.rst\n",
            "bedtools2/docs/content/related-tools.rst\n",
            "bedtools2/docs/content/images/memory-comparo.png\n",
            "bedtools2/docs/content/images/speed-comparo.png\n",
            "bedtools2/docs/content/images/fisher/\n",
            "bedtools2/docs/content/images/github-zip-button.png\n",
            "bedtools2/docs/content/images/galaxy-bedtools.png\n",
            "bedtools2/docs/content/images/bedtoolsMenu.png\n",
            "bedtools2/docs/content/images/tool-glyphs/\n",
            "bedtools2/docs/content/images/tool-glyphs/map-glyph.png\n",
            "bedtools2/docs/content/images/tool-glyphs/jaccard-glyph.png\n",
            "bedtools2/docs/content/images/tool-glyphs/merge-glyph.png\n",
            "bedtools2/docs/content/images/tool-glyphs/reldist-glyph.ai\n",
            "bedtools2/docs/content/images/tool-glyphs/getfasta-glyph.png\n",
            "bedtools2/docs/content/images/tool-glyphs/map-speed-comparo.png\n",
            "bedtools2/docs/content/images/tool-glyphs/reldist-plot.png\n",
            "bedtools2/docs/content/images/tool-glyphs/subtract-glyph.png\n",
            "bedtools2/docs/content/images/tool-glyphs/flank-glyph.png\n",
            "bedtools2/docs/content/images/tool-glyphs/jaccard-glyph.ai\n",
            "bedtools2/docs/content/images/tool-glyphs/map-glyph.svg\n",
            "bedtools2/docs/content/images/tool-glyphs/random-glyph.png\n",
            "bedtools2/docs/content/images/tool-glyphs/genomecov-glyph.png\n",
            "bedtools2/docs/content/images/tool-glyphs/cluster-glyph.png\n",
            "bedtools2/docs/content/images/tool-glyphs/maskfasta-glyph.svg\n",
            "bedtools2/docs/content/images/tool-glyphs/random-glyph.ai\n",
            "bedtools2/docs/content/images/tool-glyphs/complement-glyph.png\n",
            "bedtools2/docs/content/images/tool-glyphs/slop-glyph.png\n",
            "bedtools2/docs/content/images/tool-glyphs/window-glyph.png\n",
            "bedtools2/docs/content/images/tool-glyphs/shuffle-glyph.png\n",
            "bedtools2/docs/content/images/tool-glyphs/reldist-glyph.png\n",
            "bedtools2/docs/content/images/tool-glyphs/closest-glyph.png\n",
            "bedtools2/docs/content/images/tool-glyphs/maskfasta-glyph.png\n",
            "bedtools2/docs/content/images/tool-glyphs/intersect-glyph.png\n",
            "bedtools2/docs/content/images/tool-glyphs/barski_binding_site.png\n",
            "bedtools2/docs/content/images/fisher/res-genes.png\n",
            "bedtools2/docs/content/tools/expand.rst\n",
            "bedtools2/docs/content/tools/genomecov.rst\n",
            "bedtools2/docs/content/tools/makewindows.rst\n",
            "bedtools2/docs/content/tools/igv.rst\n",
            "bedtools2/docs/content/tools/overlap.rst\n",
            "bedtools2/docs/content/tools/nuc.rst\n",
            "bedtools2/docs/content/tools/fisher.rst\n",
            "bedtools2/docs/content/tools/cluster.rst\n",
            "bedtools2/docs/content/tools/pairtobed.rst\n",
            "bedtools2/docs/content/tools/shift.rst\n",
            "bedtools2/docs/content/tools/complement.rst\n",
            "bedtools2/docs/content/tools/bed12tobed6.rst\n",
            "bedtools2/docs/content/tools/map.rst\n",
            "bedtools2/docs/content/tools/tag.rst\n",
            "bedtools2/docs/content/tools/pairtopair.rst\n",
            "bedtools2/docs/content/tools/slop.rst\n",
            "bedtools2/docs/content/tools/sort.rst\n",
            "bedtools2/docs/content/tools/reldist.rst\n",
            "bedtools2/docs/content/tools/closest.rst\n",
            "bedtools2/docs/content/tools/bedtobam.rst\n",
            "bedtools2/docs/content/tools/maskfasta.rst\n",
            "bedtools2/docs/content/tools/bamtofastq.rst\n",
            "bedtools2/docs/content/tools/multicov.rst\n",
            "bedtools2/docs/content/tools/flank.rst\n",
            "bedtools2/docs/content/tools/random.rst\n",
            "bedtools2/docs/content/tools/merge.rst\n",
            "bedtools2/docs/content/tools/shuffle.rst\n",
            "bedtools2/docs/content/tools/window.rst\n",
            "bedtools2/docs/content/tools/bedpetobam.rst\n",
            "bedtools2/docs/content/tools/groupby.rst\n",
            "bedtools2/docs/content/tools/multiinter.rst\n",
            "bedtools2/docs/content/tools/bamtobed.rst\n",
            "bedtools2/docs/content/tools/getfasta.rst\n",
            "bedtools2/docs/content/tools/unionbedg.rst\n",
            "bedtools2/docs/content/tools/subtract.rst\n",
            "bedtools2/docs/content/tools/coverage.rst\n",
            "bedtools2/docs/content/tools/jaccard.rst\n",
            "bedtools2/docs/content/tools/links.rst\n",
            "bedtools2/docs/content/tools/annotate.rst\n",
            "bedtools2/docs/content/tools/intersect.rst\n",
            "bedtools2/docs/_static/labibi.css\n",
            "bedtools2/docs/_static/labibi.js\n",
            "bedtools2/test/split/\n",
            "bedtools2/test/bedtobam/\n",
            "bedtools2/test/cluster/\n",
            "bedtools2/test/expand/\n",
            "bedtools2/test/shuffle/\n",
            "bedtools2/test/fisher/\n",
            "bedtools2/test/complement/\n",
            "bedtools2/test/intersect/\n",
            "bedtools2/test/sort/\n",
            "bedtools2/test/subtract/\n",
            "bedtools2/test/merge/\n",
            "bedtools2/test/general/\n",
            "bedtools2/test/slop/\n",
            "bedtools2/test/reldist/\n",
            "bedtools2/test/shift/\n",
            "bedtools2/test/genomecov/\n",
            "bedtools2/test/jaccard/\n",
            "bedtools2/test/groupby/\n",
            "bedtools2/test/bamtofastq/\n",
            "bedtools2/test/bed12tobed6/\n",
            "bedtools2/test/getfasta/\n",
            "bedtools2/test/map/\n",
            "bedtools2/test/bigchroms/\n",
            "bedtools2/test/spacing/\n",
            "bedtools2/test/sample/\n",
            "bedtools2/test/flank/\n",
            "bedtools2/test/negativecontrol/\n",
            "bedtools2/test/multicov/\n",
            "bedtools2/test/closest/\n",
            "bedtools2/test/test.sh\n",
            "bedtools2/test/coverage/\n",
            "bedtools2/test/bamtobed/\n",
            "bedtools2/test/makewindows/\n",
            "bedtools2/test/makewindows/test-makewindows.sh\n",
            "bedtools2/test/makewindows/a.33bp.bed\n",
            "bedtools2/test/makewindows/a.20bp.bed\n",
            "bedtools2/test/makewindows/input.bed\n",
            "bedtools2/test/makewindows/a.19bp.bed\n",
            "bedtools2/test/bamtobed/two_blocks.sam\n",
            "bedtools2/test/bamtobed/three_blocks.sam\n",
            "bedtools2/test/bamtobed/test-bamtobed.sh\n",
            "bedtools2/test/bamtobed/numeric_tag.sam\n",
            "bedtools2/test/bamtobed/extra-long-header.sam.gz\n",
            "bedtools2/test/bamtobed/two_blocks_w_D.sam\n",
            "bedtools2/test/bamtobed/one_block.sam\n",
            "bedtools2/test/coverage/two_blocks.sam\n",
            "bedtools2/test/coverage/chr_0-100.bed\n",
            "bedtools2/test/coverage/three_blocks.bam.bai\n",
            "bedtools2/test/coverage/sam-w-del.sam\n",
            "bedtools2/test/coverage/three_blocks_nomatch.bed\n",
            "bedtools2/test/coverage/two_blocks.bam.bai\n",
            "bedtools2/test/coverage/three_blocks.sam\n",
            "bedtools2/test/coverage/one_block.bam.bai\n",
            "bedtools2/test/coverage/test-coverage.sh\n",
            "bedtools2/test/coverage/x.bed\n",
            "bedtools2/test/coverage/y.bed\n",
            "bedtools2/test/coverage/a.bed\n",
            "bedtools2/test/coverage/c.bed\n",
            "bedtools2/test/coverage/three_blocks_match.bam\n",
            "bedtools2/test/coverage/sam-w-del.bam.bai\n",
            "bedtools2/test/coverage/b.bed\n",
            "bedtools2/test/coverage/three_blocks_match.bed\n",
            "bedtools2/test/coverage/one_block.sam\n",
            "bedtools2/test/coverage/three_blocks_match_1bp.bed\n",
            "bedtools2/test/closest/d_d1F.bed\n",
            "bedtools2/test/closest/d_q1.bed\n",
            "bedtools2/test/closest/d_iu.bed\n",
            "bedtools2/test/closest/d_d1R.bed\n",
            "bedtools2/test/closest/d_q2.bed\n",
            "bedtools2/test/closest/tfbs.bed\n",
            "bedtools2/test/closest/null_b.bed\n",
            "bedtools2/test/closest/bug281_a.medium.bed\n",
            "bedtools2/test/closest/null_c.bed\n",
            "bedtools2/test/closest/null_a.bed\n",
            "bedtools2/test/closest/d_id.bed\n",
            "bedtools2/test/closest/a2.bed\n",
            "bedtools2/test/closest/test-closest.sh\n",
            "bedtools2/test/closest/kclosest/\n",
            "bedtools2/test/closest/close-a.bed\n",
            "bedtools2/test/closest/b.names.bed\n",
            "bedtools2/test/closest/close-b.bed\n",
            "bedtools2/test/closest/strand-test-d.bed\n",
            "bedtools2/test/closest/b-one-bp-closer.bed\n",
            "bedtools2/test/closest/islands.2.bed\n",
            "bedtools2/test/closest/bug157_b.bed\n",
            "bedtools2/test/closest/shores.bed\n",
            "bedtools2/test/closest/strand-test-a.bed\n",
            "bedtools2/test/closest/strand-test-c.bed\n",
            "bedtools2/test/closest/bug157_a.bed\n",
            "bedtools2/test/closest/strand-test-b.bed\n",
            "bedtools2/test/closest/bug281_b.bed\n",
            "bedtools2/test/closest/dmr.bed\n",
            "bedtools2/test/closest/a.bed\n",
            "bedtools2/test/closest/b.bed\n",
            "bedtools2/test/closest/a.names.bed\n",
            "bedtools2/test/closest/mdb3.bed\n",
            "bedtools2/test/closest/mdb2.bed\n",
            "bedtools2/test/closest/b2.bed\n",
            "bedtools2/test/closest/mdb1.bed\n",
            "bedtools2/test/closest/mq1.bed\n",
            "bedtools2/test/closest/d.bed\n",
            "bedtools2/test/closest/d4.bed\n",
            "bedtools2/test/closest/sortAndNaming/\n",
            "bedtools2/test/closest/islands.bed\n",
            "bedtools2/test/closest/bug281_b.flip.bed\n",
            "bedtools2/test/closest/bug281_a.flip.medium.bed\n",
            "bedtools2/test/closest/d_d2R.bed\n",
            "bedtools2/test/closest/d_d2F.bed\n",
            "bedtools2/test/closest/sortAndNaming/num_all.bed\n",
            "bedtools2/test/closest/sortAndNaming/db1_noChr.bed\n",
            "bedtools2/test/closest/sortAndNaming/alpha_all.bed\n",
            "bedtools2/test/closest/sortAndNaming/test-sort-and-naming.sh\n",
            "bedtools2/test/closest/sortAndNaming/q1_gls.bed\n",
            "bedtools2/test/closest/sortAndNaming/db1_num.bed\n",
            "bedtools2/test/closest/sortAndNaming/db2_numBackwards.bed\n",
            "bedtools2/test/closest/sortAndNaming/num_missing.bed\n",
            "bedtools2/test/closest/sortAndNaming/q1_num.bed\n",
            "bedtools2/test/closest/sortAndNaming/db2_num.bed\n",
            "bedtools2/test/closest/sortAndNaming/db1_leadingZero.txt\n",
            "bedtools2/test/closest/sortAndNaming/sq1.bed\n",
            "bedtools2/test/closest/sortAndNaming/db1_gls.bed\n",
            "bedtools2/test/closest/sortAndNaming/alpha_missing.bed\n",
            "bedtools2/test/closest/sortAndNaming/sdb1.bed\n",
            "bedtools2/test/closest/sortAndNaming/db3_numBackwards.bed\n",
            "bedtools2/test/closest/sortAndNaming/q1a_num.bed\n",
            "bedtools2/test/closest/kclosest/test-kclosest.sh\n",
            "bedtools2/test/closest/kclosest/bug471_b.bed\n",
            "bedtools2/test/closest/kclosest/bug471_a.bed\n",
            "bedtools2/test/closest/kclosest/d2.bed\n",
            "bedtools2/test/closest/kclosest/q2.bed\n",
            "bedtools2/test/closest/kclosest/d3.bed\n",
            "bedtools2/test/closest/kclosest/d1.bed\n",
            "bedtools2/test/closest/kclosest/q1.bed\n",
            "bedtools2/test/multicov/two_blocks.sam\n",
            "bedtools2/test/multicov/test-multicov.sh\n",
            "bedtools2/test/multicov/test-multi.bed\n",
            "bedtools2/test/multicov/test-multi.2.sam\n",
            "bedtools2/test/multicov/multicov.bed\n",
            "bedtools2/test/multicov/test-multi.sam\n",
            "bedtools2/test/multicov/one_block.sam\n",
            "bedtools2/test/negativecontrol/test-negativecontrol.sh\n",
            "bedtools2/test/flank/test-flank.sh\n",
            "bedtools2/test/flank/a.bed\n",
            "bedtools2/test/flank/tiny.genome\n",
            "bedtools2/test/sample/test-sample.sh\n",
            "bedtools2/test/sample/human.hg19.genome\n",
            "bedtools2/test/spacing/a.bed\n",
            "bedtools2/test/spacing/test-spacing.sh\n",
            "bedtools2/test/bigchroms/test-bigchroms.sh\n",
            "bedtools2/test/bigchroms/big4.bed\n",
            "bedtools2/test/bigchroms/make-big-chrom.py\n",
            "bedtools2/test/bigchroms/big4c.bed\n",
            "bedtools2/test/bigchroms/bigx.bed\n",
            "bedtools2/test/bigchroms/abig.bed\n",
            "bedtools2/test/map/values.bam\n",
            "bedtools2/test/map/.Rhistory\n",
            "bedtools2/test/map/a.vsorted.bed\n",
            "bedtools2/test/map/three_blocks_nomatch.bed\n",
            "bedtools2/test/map/b.vsorted.bed\n",
            "bedtools2/test/map/test.vcf\n",
            "bedtools2/test/map/values.bed\n",
            "bedtools2/test/map/test.gff\n",
            "bedtools2/test/map/values5.bed\n",
            "bedtools2/test/map/ivls.bed\n",
            "bedtools2/test/map/values4.bed\n",
            "bedtools2/test/map/t.bed\n",
            "bedtools2/test/map/ivls2.bed\n",
            "bedtools2/test/map/values3.bed\n",
            "bedtools2/test/map/.RData\n",
            "bedtools2/test/map/values2.bed\n",
            "bedtools2/test/map/d.bed\n",
            "bedtools2/test/map/genome\n",
            "bedtools2/test/map/three_blocks_match.bed\n",
            "bedtools2/test/map/test-map.sh\n",
            "bedtools2/test/map/bug262_a.bed\n",
            "bedtools2/test/map/bug262_b.bed\n",
            "bedtools2/test/map/fullFields.bam\n",
            "bedtools2/test/getfasta/test.iupac.bed\n",
            "bedtools2/test/getfasta/rna.fasta\n",
            "bedtools2/test/getfasta/t.fa\n",
            "bedtools2/test/getfasta/test.iupac.fa\n",
            "bedtools2/test/getfasta/blocks.bed\n",
            "bedtools2/test/getfasta/t_fH.fa.fai\n",
            "bedtools2/test/getfasta/t.bed\n",
            "bedtools2/test/getfasta/t.fa.fai\n",
            "bedtools2/test/getfasta/test-getfasta.sh\n",
            "bedtools2/test/getfasta/t_fH.fa\n",
            "bedtools2/test/getfasta/rna.bed\n",
            "bedtools2/test/bed12tobed6/two_blocks.bed\n",
            "bedtools2/test/bed12tobed6/bed6.bed\n",
            "bedtools2/test/bed12tobed6/three_blocks.bed\n",
            "bedtools2/test/bed12tobed6/test-bed12tobed6.sh\n",
            "bedtools2/test/bed12tobed6/one_blocks.bed\n",
            "bedtools2/test/bamtofastq/test-bamtofastq.sh\n",
            "bedtools2/test/bamtofastq/test2.sam\n",
            "bedtools2/test/bamtofastq/golden.fq\n",
            "bedtools2/test/bamtofastq/golden.fq2\n",
            "bedtools2/test/bamtofastq/test.sam\n",
            "bedtools2/test/groupby/gdc.bam\n",
            "bedtools2/test/groupby/values3_case.header.bed\n",
            "bedtools2/test/groupby/test-groupby.sh\n",
            "bedtools2/test/groupby/values3.unmarked_header.bed\n",
            "bedtools2/test/groupby/a_vcfSVtest.vcf\n",
            "bedtools2/test/groupby/values3.7fields.header.bed\n",
            "bedtools2/test/groupby/values3.unmarked_header.bed.2\n",
            "bedtools2/test/groupby/values3.no_header.bed\n",
            "bedtools2/test/groupby/values3.header.bed\n",
            "bedtools2/test/groupby/noPosvalues.header.bed\n",
            "bedtools2/test/groupby/test.bed\n",
            "bedtools2/test/groupby/bug569_problem.txt\n",
            "bedtools2/test/jaccard/a645.bed\n",
            "bedtools2/test/jaccard/short.bed\n",
            "bedtools2/test/jaccard/bMixedStrands.bed\n",
            "bedtools2/test/jaccard/a.bed\n",
            "bedtools2/test/jaccard/c.bed\n",
            "bedtools2/test/jaccard/three_blocks_match.bam\n",
            "bedtools2/test/jaccard/b.bed\n",
            "bedtools2/test/jaccard/b645.bed\n",
            "bedtools2/test/jaccard/e.bed\n",
            "bedtools2/test/jaccard/test-jaccard.sh\n",
            "bedtools2/test/jaccard/long.bed\n",
            "bedtools2/test/jaccard/three_blocks_match.bed\n",
            "bedtools2/test/jaccard/aMixedStrands.bed\n",
            "bedtools2/test/jaccard/a.bam\n",
            "bedtools2/test/genomecov/two_blocks.sam\n",
            "bedtools2/test/genomecov/test_ref.fa.fai\n",
            "bedtools2/test/genomecov/test-genomecov.sh\n",
            "bedtools2/test/genomecov/chip.sam\n",
            "bedtools2/test/genomecov/y.bam\n",
            "bedtools2/test/genomecov/test_ref.fa\n",
            "bedtools2/test/genomecov/sam-w-del.sam\n",
            "bedtools2/test/genomecov/three_blocks_nomatch.bed\n",
            "bedtools2/test/genomecov/three_blocks.sam\n",
            "bedtools2/test/genomecov/empty.bam\n",
            "bedtools2/test/genomecov/y.bed\n",
            "bedtools2/test/genomecov/empty.cram\n",
            "bedtools2/test/genomecov/genome.txt\n",
            "bedtools2/test/genomecov/three_blocks_match.bed\n",
            "bedtools2/test/genomecov/pair-chip.sam\n",
            "bedtools2/test/genomecov/one_block.sam\n",
            "bedtools2/test/genomecov/three_blocks_match_1bp.bed\n",
            "bedtools2/test/shift/test-shift.sh\n",
            "bedtools2/test/shift/issue_807.bed\n",
            "bedtools2/test/shift/a.bed\n",
            "bedtools2/test/shift/b.bed\n",
            "bedtools2/test/shift/tiny.genome\n",
            "bedtools2/test/shift/huge.genome\n",
            "bedtools2/test/shift/issue_807.genomesize\n",
            "bedtools2/test/reldist/issue_711.a.bed\n",
            "bedtools2/test/reldist/issue_711.b.bed\n",
            "bedtools2/test/reldist/test-reldist.sh\n",
            "bedtools2/test/slop/a.bed\n",
            "bedtools2/test/slop/tiny.genome\n",
            "bedtools2/test/slop/test-slop.sh\n",
            "bedtools2/test/general/hg19.fa.fai\n",
            "bedtools2/test/general/empty.bed\n",
            "bedtools2/test/general/a.chromheader.bed\n",
            "bedtools2/test/general/test-general.sh\n",
            "bedtools2/test/general/non-empty.bed.gz\n",
            "bedtools2/test/general/t.bed\n",
            "bedtools2/test/general/a.trackheader.bed\n",
            "bedtools2/test/merge/bamCol8Mean.txt\n",
            "bedtools2/test/merge/mergeBugTest.bed\n",
            "bedtools2/test/merge/expFormat.bed\n",
            "bedtools2/test/merge/bamCol9Mean.txt\n",
            "bedtools2/test/merge/unsorted.bed\n",
            "bedtools2/test/merge/bug254_d.bed\n",
            "bedtools2/test/merge/bug254_e.bed\n",
            "bedtools2/test/merge/bamCol6Collapse.txt\n",
            "bedtools2/test/merge/bamCol3Collapse.txt\n",
            "bedtools2/test/merge/vcfSVtest.vcf\n",
            "bedtools2/test/merge/bug254_c.bed\n",
            "bedtools2/test/merge/bamCol11Collapse.txt\n",
            "bedtools2/test/merge/a.full.bed\n",
            "bedtools2/test/merge/a.bed\n",
            "bedtools2/test/merge/test-merge.sh\n",
            "bedtools2/test/merge/precisionTest.bed\n",
            "bedtools2/test/merge/b.bed\n",
            "bedtools2/test/merge/a.names.bed\n",
            "bedtools2/test/merge/testA.vcf\n",
            "bedtools2/test/merge/bamCol1Collapse.txt\n",
            "bedtools2/test/merge/precisionTest2.bed\n",
            "bedtools2/test/merge/bamCol10Collapse.txt\n",
            "bedtools2/test/merge/bamCol5Mean.txt\n",
            "bedtools2/test/merge/a.gff\n",
            "bedtools2/test/merge/bamCol7Collapse.txt\n",
            "bedtools2/test/merge/bamCol4Mean.txt\n",
            "bedtools2/test/merge/a.full.bam\n",
            "bedtools2/test/merge/mixedStrands.bed\n",
            "bedtools2/test/merge/fullFields.bam\n",
            "bedtools2/test/subtract/test-subtract.sh\n",
            "bedtools2/test/subtract/a.bed\n",
            "bedtools2/test/subtract/c.bed\n",
            "bedtools2/test/subtract/b.bed\n",
            "bedtools2/test/subtract/b2.bed\n",
            "bedtools2/test/subtract/d.bed\n",
            "bedtools2/test/sort/names.txt\n",
            "bedtools2/test/sort/a.bed\n",
            "bedtools2/test/sort/test-sort.sh\n",
            "bedtools2/test/sort/b.bed\n",
            "bedtools2/test/intersect/two_blocks.sam\n",
            "bedtools2/test/intersect/a_with_bothUnmapped.bam\n",
            "bedtools2/test/intersect/gdc.gff\n",
            "bedtools2/test/intersect/bug223_sv1_b.vcf\n",
            "bedtools2/test/intersect/bed3.bed\n",
            "bedtools2/test/intersect/bug150_a.bed\n",
            "bedtools2/test/intersect/test_ref.fa.fai\n",
            "bedtools2/test/intersect/x.bam\n",
            "bedtools2/test/intersect/exons.issue750.bed\n",
            "bedtools2/test/intersect/performanceTest.sh\n",
            "bedtools2/test/intersect/bed6.strand.bed\n",
            "bedtools2/test/intersect/queryUnmappedMateMappedCoordsInvalid.bam\n",
            "bedtools2/test/intersect/bug223_sv1_a.vcf\n",
            "bedtools2/test/intersect/bug150_b.bed\n",
            "bedtools2/test/intersect/headerOnly.vcf\n",
            "bedtools2/test/intersect/gdc_one.bed\n",
            "bedtools2/test/intersect/bed6.strand2.bed\n",
            "bedtools2/test/intersect/null_b.bed\n",
            "bedtools2/test/intersect/a_withLargeHeader.bed\n",
            "bedtools2/test/intersect/human.hg19.vSort.genome\n",
            "bedtools2/test/intersect/gdc.bam\n",
            "bedtools2/test/intersect/bed4.bed\n",
            "bedtools2/test/intersect/chromOrderB.bed\n",
            "bedtools2/test/intersect/bedplus.bed\n",
            "bedtools2/test/intersect/test_ref.fa\n",
            "bedtools2/test/intersect/bed5.bed\n",
            "bedtools2/test/intersect/oneRecordNoNewline.bed\n",
            "bedtools2/test/intersect/aWithHeaderVsB.txt\n",
            "bedtools2/test/intersect/null_c.bed\n",
            "bedtools2/test/intersect/j1.bed\n",
            "bedtools2/test/intersect/b.issue311.vcf\n",
            "bedtools2/test/intersect/null_a.bed\n",
            "bedtools2/test/intersect/oneUnmapped.bam\n",
            "bedtools2/test/intersect/chromOrderA.bed\n",
            "bedtools2/test/intersect/bed6.bed\n",
            "bedtools2/test/intersect/issue_773_b.bed\n",
            "bedtools2/test/intersect/new_test-intersect.sh\n",
            "bedtools2/test/intersect/chromsOutOfOrder.bed\n",
            "bedtools2/test/intersect/jim.bed\n",
            "bedtools2/test/intersect/three_blocks_nomatch.bed\n",
            "bedtools2/test/intersect/b_vcfSVtest.vcf\n",
            "bedtools2/test/intersect/bug223_f.vcf\n",
            "bedtools2/test/intersect/a.issue311.bed\n",
            "bedtools2/test/intersect/three_blocks.sam\n",
            "bedtools2/test/intersect/gdc_exp\n",
            "bedtools2/test/intersect/bug44_b.bed\n",
            "bedtools2/test/intersect/bug223_d.vcf\n",
            "bedtools2/test/intersect/bug223_e.vcf\n",
            "bedtools2/test/intersect/a_testZeroLen.bed\n",
            "bedtools2/test/intersect/a_bgzipped.bed.gz\n",
            "bedtools2/test/intersect/x.bed\n",
            "bedtools2/test/intersect/a_withLargeHeader_bgzipped.bed.gz\n",
            "bedtools2/test/intersect/dosLineChar_a.bed\n",
            "bedtools2/test/intersect/bug44_a.vcf.gz\n",
            "bedtools2/test/intersect/b.cram\n",
            "bedtools2/test/intersect/y.bed\n",
            "bedtools2/test/intersect/test-intersect.sh\n",
            "bedtools2/test/intersect/a_vcfSVtest.vcf\n",
            "bedtools2/test/intersect/blocks.bed12\n",
            "bedtools2/test/intersect/recordsOutOfOrder.bed\n",
            "bedtools2/test/intersect/a.bed\n",
            "bedtools2/test/intersect/jim.vcf\n",
            "bedtools2/test/intersect/c.bed\n",
            "bedtools2/test/intersect/three_blocks_match.bam\n",
            "bedtools2/test/intersect/dosLineCharWithExtraTab_b.bed\n",
            "bedtools2/test/intersect/a.cram\n",
            "bedtools2/test/intersect/b.bed\n",
            "bedtools2/test/intersect/two_blocks_partial.bam.bai\n",
            "bedtools2/test/intersect/foo\n",
            "bedtools2/test/intersect/split.issue750.bam\n",
            "bedtools2/test/intersect/mapped_and_unmapped.sam\n",
            "bedtools2/test/intersect/two_blocks_partial.bam\n",
            "bedtools2/test/intersect/e.bed\n",
            "bedtools2/test/intersect/a_withLargeHeader_gzipped.bed.gz\n",
            "bedtools2/test/intersect/svlen.vcf\n",
            "bedtools2/test/intersect/d.bed\n",
            "bedtools2/test/intersect/splitBug155_a.bed\n",
            "bedtools2/test/intersect/multi_intersect/\n",
            "bedtools2/test/intersect/notexthdr.bam\n",
            "bedtools2/test/intersect/large_b.bed\n",
            "bedtools2/test/intersect/nonamecheck_a.bed\n",
            "bedtools2/test/intersect/issue_773_y.bed\n",
            "bedtools2/test/intersect/sortAndNaming/\n",
            "bedtools2/test/intersect/split.issue750.sam\n",
            "bedtools2/test/intersect/a_gzipped.bed.gz\n",
            "bedtools2/test/intersect/splitBug155_b.bed\n",
            "bedtools2/test/intersect/bug167_strandSweep.bed\n",
            "bedtools2/test/intersect/large_a.bed\n",
            "bedtools2/test/intersect/nonamecheck_b.bed\n",
            "bedtools2/test/intersect/b.issue311.gff\n",
            "bedtools2/test/intersect/two_blocks_partial.bed\n",
            "bedtools2/test/intersect/three_blocks_match.bed\n",
            "bedtools2/test/intersect/human.hg19.genome\n",
            "bedtools2/test/intersect/bug187_a.bed\n",
            "bedtools2/test/intersect/strand_with_star.b.bed\n",
            "bedtools2/test/intersect/b_testZeroLen.bed\n",
            "bedtools2/test/intersect/one_block.sam\n",
            "bedtools2/test/intersect/bug187_b.bed\n",
            "bedtools2/test/intersect/three_blocks_match_1bp.bed\n",
            "bedtools2/test/intersect/strand_with_star.a.bed\n",
            "bedtools2/test/intersect/aVSb.bam\n",
            "bedtools2/test/intersect/a.bam\n",
            "bedtools2/test/intersect/sortAndNaming/num_all.bed\n",
            "bedtools2/test/intersect/sortAndNaming/diffs.txt\n",
            "bedtools2/test/intersect/sortAndNaming/db1_noChr.bed\n",
            "bedtools2/test/intersect/sortAndNaming/alpha_all.bed\n",
            "bedtools2/test/intersect/sortAndNaming/test-sort-and-naming.sh\n",
            "bedtools2/test/intersect/sortAndNaming/exp1\n",
            "bedtools2/test/intersect/sortAndNaming/q1_gls.bed\n",
            "bedtools2/test/intersect/sortAndNaming/db1_num.bed\n",
            "bedtools2/test/intersect/sortAndNaming/db2_numBackwards.bed\n",
            "bedtools2/test/intersect/sortAndNaming/num_missing.bed\n",
            "bedtools2/test/intersect/sortAndNaming/q1_num.bed\n",
            "bedtools2/test/intersect/sortAndNaming/db2_num.bed\n",
            "bedtools2/test/intersect/sortAndNaming/db1_leadingZero.txt\n",
            "bedtools2/test/intersect/sortAndNaming/sq1.bed\n",
            "bedtools2/test/intersect/sortAndNaming/db1_gls.bed\n",
            "bedtools2/test/intersect/sortAndNaming/alpha_missing.bed\n",
            "bedtools2/test/intersect/sortAndNaming/bigTests/\n",
            "bedtools2/test/intersect/sortAndNaming/sdb1.bed\n",
            "bedtools2/test/intersect/sortAndNaming/db3_numBackwards.bed\n",
            "bedtools2/test/intersect/sortAndNaming/q1a_num.bed\n",
            "bedtools2/test/intersect/sortAndNaming/bigTests/curr5v5.txt\n",
            "bedtools2/test/intersect/sortAndNaming/bigTests/2.21_500Kvs500K.txt\n",
            "bedtools2/test/intersect/sortAndNaming/bigTests/q500K.bed\n",
            "bedtools2/test/intersect/sortAndNaming/bigTests/dbchr1.bed\n",
            "bedtools2/test/intersect/sortAndNaming/bigTests/2.21_chr1.txt\n",
            "bedtools2/test/intersect/sortAndNaming/bigTests/diffs.txt\n",
            "bedtools2/test/intersect/sortAndNaming/bigTests/curr500Kvs500K.txt\n",
            "bedtools2/test/intersect/sortAndNaming/bigTests/qchr1.bed\n",
            "bedtools2/test/intersect/sortAndNaming/bigTests/2.21_5v5.txt\n",
            "bedtools2/test/intersect/sortAndNaming/bigTests/currChr1.txt\n",
            "bedtools2/test/intersect/sortAndNaming/bigTests/db500K.bed\n",
            "bedtools2/test/intersect/multi_intersect/query3.bed\n",
            "bedtools2/test/intersect/multi_intersect/query2.bed\n",
            "bedtools2/test/intersect/multi_intersect/d8.bed\n",
            "bedtools2/test/intersect/multi_intersect/d9.bed\n",
            "bedtools2/test/intersect/multi_intersect/query.bed\n",
            "bedtools2/test/intersect/multi_intersect/g.bed\n",
            "bedtools2/test/intersect/multi_intersect/test-multi_intersect.sh\n",
            "bedtools2/test/intersect/multi_intersect/d4.bed\n",
            "bedtools2/test/intersect/multi_intersect/d5.bed\n",
            "bedtools2/test/intersect/multi_intersect/d7.bed\n",
            "bedtools2/test/intersect/multi_intersect/d6.bed\n",
            "bedtools2/test/intersect/multi_intersect/d2.bed\n",
            "bedtools2/test/intersect/multi_intersect/d3.bed\n",
            "bedtools2/test/intersect/multi_intersect/d1.bed\n",
            "bedtools2/test/complement/test-complement.sh\n",
            "bedtools2/test/complement/issue_503.genome\n",
            "bedtools2/test/complement/issue_503.bed\n",
            "bedtools2/test/fisher/test-fisher.sh\n",
            "bedtools2/test/fisher/sim.py\n",
            "bedtools2/test/fisher/plot.py\n",
            "bedtools2/test/fisher/cmp.sh\n",
            "bedtools2/test/fisher/dm6.fai\n",
            "bedtools2/test/fisher/t.60.genome\n",
            "bedtools2/test/fisher/get.sh\n",
            "bedtools2/test/fisher/README.md\n",
            "bedtools2/test/fisher/a.bed\n",
            "bedtools2/test/fisher/b.bed\n",
            "bedtools2/test/fisher/tumor.gff\n",
            "bedtools2/test/fisher/t.500.genome\n",
            "bedtools2/test/fisher/a_merge.bed\n",
            "bedtools2/test/fisher/shuf.sh\n",
            "bedtools2/test/fisher/test.bed\n",
            "bedtools2/test/shuffle/incl.sizedif.bed\n",
            "bedtools2/test/shuffle/test-shuffle.sh\n",
            "bedtools2/test/shuffle/excl.bed\n",
            "bedtools2/test/shuffle/simrep.bed\n",
            "bedtools2/test/shuffle/incl.bed\n",
            "bedtools2/test/expand/expand.txt\n",
            "bedtools2/test/expand/test-expand.sh\n",
            "bedtools2/test/cluster/in.bed\n",
            "bedtools2/test/cluster/test-cluster.sh\n",
            "bedtools2/test/bedtobam/test-bedtobam.sh\n",
            "bedtools2/test/bedtobam/chrsize.tmp\n",
            "bedtools2/test/split/test-split.sh\n",
            "bedtools2/test/split/randData.bed\n",
            "bedtools2/tutorial/answers.md\n",
            "bedtools2/tutorial/template.class.html\n",
            "bedtools2/tutorial/answers.html\n",
            "bedtools2/tutorial/bootstrap.css\n",
            "bedtools2/tutorial/Makefile\n",
            "bedtools2/tutorial/bedtools.html\n",
            "bedtools2/tutorial/bedtools.md\n",
            "bedtools2/tutorial/template.html\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!cd bedtools2; make 2> /dev/null"
      ],
      "metadata": {
        "id": "o4o0PXv2adMK",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "285eacc3-b87c-4828-b14e-20171ad0ed83"
      },
      "execution_count": 56,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Building BEDTools:\n",
            "=========================================================\n",
            "CXXFLAGS is [-g -Wall -O2 -std=c++11 ]\n",
            "DETECTED_VERSION = v2.30.0\n",
            "CURRENT_VERSION  = \n",
            "Updating version file.\n",
            "  * compiling src/bedtools.cpp\n",
            "  * compiling src/annotateBed/annotateMain.cpp\n",
            "  * compiling src/annotateBed/annotateBed.cpp\n",
            "  * compiling src/bamToBed/bamToBed.cpp\n",
            "\u001b[01m\u001b[Ksrc/bamToBed/bamToBed.cpp:\u001b[m\u001b[K In function ‘\u001b[01m\u001b[Kint bamtobed_main(int, char**)\u001b[m\u001b[K’:\n",
            "\u001b[01m\u001b[Ksrc/bamToBed/bamToBed.cpp:88:10:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kvariable ‘\u001b[01m\u001b[KuseAlignmentScore\u001b[m\u001b[K’ set but not used [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wunused-but-set-variable\u0007-Wunused-but-set-variable\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "   88 |     bool \u001b[01;35m\u001b[KuseAlignmentScore\u001b[m\u001b[K = false;\n",
            "      |          \u001b[01;35m\u001b[K^~~~~~~~~~~~~~~~~\u001b[m\u001b[K\n",
            "In file included from \u001b[01m\u001b[Ksrc/utils/BamTools/include/api/BamReader.h:1\u001b[m\u001b[K,\n",
            "                 from \u001b[01m\u001b[Ksrc/bamToBed/bamToBed.cpp:12\u001b[m\u001b[K:\n",
            "\u001b[01m\u001b[Ksrc/utils/BamTools/include/SamHeader.hpp:\u001b[m\u001b[K In function ‘\u001b[01m\u001b[Kint htslib_future::sam_hdr_rebuild(bam_hdr_t*)\u001b[m\u001b[K’:\n",
            "\u001b[01m\u001b[Ksrc/utils/BamTools/include/SamHeader.hpp:32:30:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kignoring return value of ‘\u001b[01m\u001b[Kint sam_hdr_write(samFile*, const bam_hdr_t*)\u001b[m\u001b[K’ declared with attribute ‘\u001b[01m\u001b[Kwarn_unused_result\u001b[m\u001b[K’ [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wunused-result\u0007-Wunused-result\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "   32 |                 \u001b[01;35m\u001b[Ksam_hdr_write(fp, hdr)\u001b[m\u001b[K;\n",
            "      |                 \u001b[01;35m\u001b[K~~~~~~~~~~~~~^~~~~~~~~\u001b[m\u001b[K\n",
            "  * compiling src/bamToFastq/bamToFastqMain.cpp\n",
            "\u001b[01m\u001b[Ksrc/bamToFastq/bamToFastqMain.cpp:\u001b[m\u001b[K In function ‘\u001b[01m\u001b[Kint bamtofastq_main(int, char**)\u001b[m\u001b[K’:\n",
            "\u001b[01m\u001b[Ksrc/bamToFastq/bamToFastqMain.cpp:38:10:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kvariable ‘\u001b[01m\u001b[KhaveFastq2\u001b[m\u001b[K’ set but not used [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wunused-but-set-variable\u0007-Wunused-but-set-variable\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "   38 |     bool \u001b[01;35m\u001b[KhaveFastq2\u001b[m\u001b[K    = false;\n",
            "      |          \u001b[01;35m\u001b[K^~~~~~~~~~\u001b[m\u001b[K\n",
            "  * compiling src/bamToFastq/bamToFastq.cpp\n",
            "In file included from \u001b[01m\u001b[Ksrc/utils/BamTools/include/api/BamReader.h:1\u001b[m\u001b[K,\n",
            "                 from \u001b[01m\u001b[Ksrc/bamToFastq/bamToFastq.h:15\u001b[m\u001b[K,\n",
            "                 from \u001b[01m\u001b[Ksrc/bamToFastq/bamToFastq.cpp:15\u001b[m\u001b[K:\n",
            "\u001b[01m\u001b[Ksrc/utils/BamTools/include/SamHeader.hpp:\u001b[m\u001b[K In function ‘\u001b[01m\u001b[Kint htslib_future::sam_hdr_rebuild(bam_hdr_t*)\u001b[m\u001b[K’:\n",
            "\u001b[01m\u001b[Ksrc/utils/BamTools/include/SamHeader.hpp:32:30:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kignoring return value of ‘\u001b[01m\u001b[Kint sam_hdr_write(samFile*, const bam_hdr_t*)\u001b[m\u001b[K’ declared with attribute ‘\u001b[01m\u001b[Kwarn_unused_result\u001b[m\u001b[K’ [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wunused-result\u0007-Wunused-result\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "   32 |                 \u001b[01;35m\u001b[Ksam_hdr_write(fp, hdr)\u001b[m\u001b[K;\n",
            "      |                 \u001b[01;35m\u001b[K~~~~~~~~~~~~~^~~~~~~~~\u001b[m\u001b[K\n",
            "  * compiling src/bedToBam/bedToBam.cpp\n",
            "\u001b[01m\u001b[Ksrc/bedToBam/bedToBam.cpp:\u001b[m\u001b[K In function ‘\u001b[01m\u001b[Kint bedtobam_main(int, char**)\u001b[m\u001b[K’:\n",
            "\u001b[01m\u001b[Ksrc/bedToBam/bedToBam.cpp:60:10:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kvariable ‘\u001b[01m\u001b[KhaveMapQual\u001b[m\u001b[K’ set but not used [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wunused-but-set-variable\u0007-Wunused-but-set-variable\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "   60 |     bool \u001b[01;35m\u001b[KhaveMapQual\u001b[m\u001b[K     = false;\n",
            "      |          \u001b[01;35m\u001b[K^~~~~~~~~~~\u001b[m\u001b[K\n",
            "  * compiling src/bedpeToBam/bedpeToBam.cpp\n",
            "\u001b[01m\u001b[Ksrc/bedpeToBam/bedpeToBam.cpp:\u001b[m\u001b[K In function ‘\u001b[01m\u001b[Kint bedpetobam_main(int, char**)\u001b[m\u001b[K’:\n",
            "\u001b[01m\u001b[Ksrc/bedpeToBam/bedpeToBam.cpp:61:10:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kvariable ‘\u001b[01m\u001b[KhaveMapQual\u001b[m\u001b[K’ set but not used [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wunused-but-set-variable\u0007-Wunused-but-set-variable\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "   61 |     bool \u001b[01;35m\u001b[KhaveMapQual\u001b[m\u001b[K     = false;\n",
            "      |          \u001b[01;35m\u001b[K^~~~~~~~~~~\u001b[m\u001b[K\n",
            "  * compiling src/bedToIgv/bedToIgv.cpp\n",
            "  * compiling src/bed12ToBed6/bed12ToBed6.cpp\n",
            "  * compiling src/closestFile/closestHelp.cpp\n",
            "  * compiling src/closestFile/closestFile.cpp\n",
            "  * compiling src/clusterBed/clusterMain.cpp\n",
            "\u001b[01m\u001b[Ksrc/clusterBed/clusterMain.cpp:\u001b[m\u001b[K In function ‘\u001b[01m\u001b[Kint cluster_main(int, char**)\u001b[m\u001b[K’:\n",
            "\u001b[01m\u001b[Ksrc/clusterBed/clusterMain.cpp:38:10:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kvariable ‘\u001b[01m\u001b[KhaveMaxDistance\u001b[m\u001b[K’ set but not used [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wunused-but-set-variable\u0007-Wunused-but-set-variable\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "   38 |     bool \u001b[01;35m\u001b[KhaveMaxDistance\u001b[m\u001b[K = false;\n",
            "      |          \u001b[01;35m\u001b[K^~~~~~~~~~~~~~~\u001b[m\u001b[K\n",
            "  * compiling src/clusterBed/clusterBed.cpp\n",
            "  * compiling src/complementFile/complementHelp.cpp\n",
            "  * compiling src/complementFile/complementFile.cpp\n",
            "  * compiling src/coverageFile/coverageHelp.cpp\n",
            "  * compiling src/coverageFile/coverageFile.cpp\n",
            "\u001b[01m\u001b[Ksrc/coverageFile/coverageFile.cpp:\u001b[m\u001b[K In member function ‘\u001b[01m\u001b[Kvirtual void CoverageFile::giveFinalReport(RecordOutputMgr*)\u001b[m\u001b[K’:\n",
            "\u001b[01m\u001b[Ksrc/coverageFile/coverageFile.cpp:100:25:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kignoring return value of ‘\u001b[01m\u001b[Kint asprintf(char**, const char*, ...)\u001b[m\u001b[K’ declared with attribute ‘\u001b[01m\u001b[Kwarn_unused_result\u001b[m\u001b[K’ [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wunused-result\u0007-Wunused-result\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "  100 |                 \u001b[01;35m\u001b[Kasprintf(&depthPctString, \"%0.7f\", depthPct)\u001b[m\u001b[K;\n",
            "      |                 \u001b[01;35m\u001b[K~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\u001b[m\u001b[K\n",
            "\u001b[01m\u001b[Ksrc/coverageFile/coverageFile.cpp:\u001b[m\u001b[K In member function ‘\u001b[01m\u001b[Kvoid CoverageFile::doMean(RecordOutputMgr*, RecordKeyVector&)\u001b[m\u001b[K’:\n",
            "\u001b[01m\u001b[Ksrc/coverageFile/coverageFile.cpp:181:17:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kignoring return value of ‘\u001b[01m\u001b[Kint asprintf(char**, const char*, ...)\u001b[m\u001b[K’ declared with attribute ‘\u001b[01m\u001b[Kwarn_unused_result\u001b[m\u001b[K’ [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wunused-result\u0007-Wunused-result\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "  181 |         \u001b[01;35m\u001b[Kasprintf(&meanString, \"%0.7f\", mean)\u001b[m\u001b[K;\n",
            "      |         \u001b[01;35m\u001b[K~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~\u001b[m\u001b[K\n",
            "\u001b[01m\u001b[Ksrc/coverageFile/coverageFile.cpp:\u001b[m\u001b[K In member function ‘\u001b[01m\u001b[Kvoid CoverageFile::doHist(RecordOutputMgr*, RecordKeyVector&)\u001b[m\u001b[K’:\n",
            "\u001b[01m\u001b[Ksrc/coverageFile/coverageFile.cpp:210:25:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kignoring return value of ‘\u001b[01m\u001b[Kint asprintf(char**, const char*, ...)\u001b[m\u001b[K’ declared with attribute ‘\u001b[01m\u001b[Kwarn_unused_result\u001b[m\u001b[K’ [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wunused-result\u0007-Wunused-result\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "  210 |                 \u001b[01;35m\u001b[Kasprintf(&coveredFractionString, \"%0.7f\", coveredFraction)\u001b[m\u001b[K;\n",
            "      |                 \u001b[01;35m\u001b[K~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\u001b[m\u001b[K\n",
            "\u001b[01m\u001b[Ksrc/coverageFile/coverageFile.cpp:\u001b[m\u001b[K In member function ‘\u001b[01m\u001b[Kvoid CoverageFile::doDefault(RecordOutputMgr*, RecordKeyVector&)\u001b[m\u001b[K’:\n",
            "\u001b[01m\u001b[Ksrc/coverageFile/coverageFile.cpp:230:17:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kignoring return value of ‘\u001b[01m\u001b[Kint asprintf(char**, const char*, ...)\u001b[m\u001b[K’ declared with attribute ‘\u001b[01m\u001b[Kwarn_unused_result\u001b[m\u001b[K’ [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wunused-result\u0007-Wunused-result\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "  230 |         \u001b[01;35m\u001b[Kasprintf(&coveredFractionString, \"%0.7f\", coveredFraction)\u001b[m\u001b[K;\n",
            "      |         \u001b[01;35m\u001b[K~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\u001b[m\u001b[K\n",
            "  * compiling src/expand/expand.cpp\n",
            "  * compiling src/fastaFromBed/fastaFromBedMain.cpp\n",
            "  * compiling src/fastaFromBed/fastaFromBed.cpp\n",
            "  * compiling src/flankBed/flankBedMain.cpp\n",
            "  * compiling src/flankBed/flankBed.cpp\n",
            "  * compiling src/genomeCoverageBed/genomeCoverageMain.cpp\n",
            "  * compiling src/genomeCoverageBed/genomeCoverageBed.cpp\n",
            "In file included from \u001b[01m\u001b[Ksrc/utils/BamTools/include/api/BamReader.h:1\u001b[m\u001b[K,\n",
            "                 from \u001b[01m\u001b[Ksrc/utils/GenomeFile/NewGenomeFile.h:20\u001b[m\u001b[K,\n",
            "                 from \u001b[01m\u001b[Ksrc/genomeCoverageBed/genomeCoverageBed.h:13\u001b[m\u001b[K,\n",
            "                 from \u001b[01m\u001b[Ksrc/genomeCoverageBed/genomeCoverageBed.cpp:13\u001b[m\u001b[K:\n",
            "\u001b[01m\u001b[Ksrc/utils/BamTools/include/SamHeader.hpp:\u001b[m\u001b[K In function ‘\u001b[01m\u001b[Kint htslib_future::sam_hdr_rebuild(bam_hdr_t*)\u001b[m\u001b[K’:\n",
            "\u001b[01m\u001b[Ksrc/utils/BamTools/include/SamHeader.hpp:32:30:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kignoring return value of ‘\u001b[01m\u001b[Kint sam_hdr_write(samFile*, const bam_hdr_t*)\u001b[m\u001b[K’ declared with attribute ‘\u001b[01m\u001b[Kwarn_unused_result\u001b[m\u001b[K’ [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wunused-result\u0007-Wunused-result\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "   32 |                 \u001b[01;35m\u001b[Ksam_hdr_write(fp, hdr)\u001b[m\u001b[K;\n",
            "      |                 \u001b[01;35m\u001b[K~~~~~~~~~~~~~^~~~~~~~~\u001b[m\u001b[K\n",
            "  * compiling src/getOverlap/getOverlap.cpp\n",
            "\u001b[01m\u001b[Ksrc/getOverlap/getOverlap.cpp:\u001b[m\u001b[K In function ‘\u001b[01m\u001b[Kint getoverlap_main(int, char**)\u001b[m\u001b[K’:\n",
            "\u001b[01m\u001b[Ksrc/getOverlap/getOverlap.cpp:44:10:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kvariable ‘\u001b[01m\u001b[KhaveColumns\u001b[m\u001b[K’ set but not used [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wunused-but-set-variable\u0007-Wunused-but-set-variable\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "   44 |     bool \u001b[01;35m\u001b[KhaveColumns\u001b[m\u001b[K = false;\n",
            "      |          \u001b[01;35m\u001b[K^~~~~~~~~~~\u001b[m\u001b[K\n",
            "  * compiling src/groupBy/groupByHelp.cpp\n",
            "  * compiling src/groupBy/groupBy.cpp\n",
            "  * compiling src/intersectFile/intersectHelp.cpp\n",
            "  * compiling src/intersectFile/intersectFile.cpp\n",
            "  * compiling src/fisher/fisherHelp.cpp\n",
            "  * compiling src/fisher/fisher.cpp\n",
            "  * compiling src/fisher/kfunc.cpp\n",
            "  * compiling src/jaccard/jaccardHelp.cpp\n",
            "  * compiling src/jaccard/jaccard.cpp\n",
            "  * compiling src/linksBed/linksMain.cpp\n",
            "  * compiling src/linksBed/linksBed.cpp\n",
            "  * compiling src/maskFastaFromBed/maskFastaFromBedMain.cpp\n",
            "  * compiling src/maskFastaFromBed/maskFastaFromBed.cpp\n",
            "  * compiling src/mapFile/mapHelp.cpp\n",
            "  * compiling src/mapFile/mapFile.cpp\n",
            "  * compiling src/mergeFile/mergeHelp.cpp\n",
            "  * compiling src/mergeFile/mergeFile.cpp\n",
            "  * compiling src/multiBamCov/multiBamCovMain.cpp\n",
            "\u001b[01m\u001b[Ksrc/multiBamCov/multiBamCovMain.cpp:\u001b[m\u001b[K In function ‘\u001b[01m\u001b[Kint multibamcov_main(int, char**)\u001b[m\u001b[K’:\n",
            "\u001b[01m\u001b[Ksrc/multiBamCov/multiBamCovMain.cpp:38:10:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kvariable ‘\u001b[01m\u001b[KhaveBed\u001b[m\u001b[K’ set but not used [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wunused-but-set-variable\u0007-Wunused-but-set-variable\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "   38 |     bool \u001b[01;35m\u001b[KhaveBed\u001b[m\u001b[K           = false;\n",
            "      |          \u001b[01;35m\u001b[K^~~~~~~\u001b[m\u001b[K\n",
            "\u001b[01m\u001b[Ksrc/multiBamCov/multiBamCovMain.cpp:39:10:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kvariable ‘\u001b[01m\u001b[KhaveBams\u001b[m\u001b[K’ set but not used [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wunused-but-set-variable\u0007-Wunused-but-set-variable\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "   39 |     bool \u001b[01;35m\u001b[KhaveBams\u001b[m\u001b[K          = false;\n",
            "      |          \u001b[01;35m\u001b[K^~~~~~~~\u001b[m\u001b[K\n",
            "\u001b[01m\u001b[Ksrc/multiBamCov/multiBamCovMain.cpp:47:10:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kvariable ‘\u001b[01m\u001b[KhaveFraction\u001b[m\u001b[K’ set but not used [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wunused-but-set-variable\u0007-Wunused-but-set-variable\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "   47 |     bool \u001b[01;35m\u001b[KhaveFraction\u001b[m\u001b[K       = false;\n",
            "      |          \u001b[01;35m\u001b[K^~~~~~~~~~~~\u001b[m\u001b[K\n",
            "  * compiling src/multiBamCov/multiBamCov.cpp\n",
            "In file included from \u001b[01m\u001b[Ksrc/utils/BamTools/include/BamReader.hpp:4\u001b[m\u001b[K,\n",
            "                 from \u001b[01m\u001b[Ksrc/utils/BamTools/include/api/BamMultiReader.h:4\u001b[m\u001b[K,\n",
            "                 from \u001b[01m\u001b[Ksrc/multiBamCov/multiBamCov.h:16\u001b[m\u001b[K,\n",
            "                 from \u001b[01m\u001b[Ksrc/multiBamCov/multiBamCov.cpp:13\u001b[m\u001b[K:\n",
            "\u001b[01m\u001b[Ksrc/utils/BamTools/include/SamHeader.hpp:\u001b[m\u001b[K In function ‘\u001b[01m\u001b[Kint htslib_future::sam_hdr_rebuild(bam_hdr_t*)\u001b[m\u001b[K’:\n",
            "\u001b[01m\u001b[Ksrc/utils/BamTools/include/SamHeader.hpp:32:30:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kignoring return value of ‘\u001b[01m\u001b[Kint sam_hdr_write(samFile*, const bam_hdr_t*)\u001b[m\u001b[K’ declared with attribute ‘\u001b[01m\u001b[Kwarn_unused_result\u001b[m\u001b[K’ [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wunused-result\u0007-Wunused-result\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "   32 |                 \u001b[01;35m\u001b[Ksam_hdr_write(fp, hdr)\u001b[m\u001b[K;\n",
            "      |                 \u001b[01;35m\u001b[K~~~~~~~~~~~~~^~~~~~~~~\u001b[m\u001b[K\n",
            "  * compiling src/multiIntersectBed/multiIntersectBedMain.cpp\n",
            "\u001b[01m\u001b[Ksrc/multiIntersectBed/multiIntersectBedMain.cpp:\u001b[m\u001b[K In function ‘\u001b[01m\u001b[Kint multiintersect_main(int, char**)\u001b[m\u001b[K’:\n",
            "\u001b[01m\u001b[Ksrc/multiIntersectBed/multiIntersectBedMain.cpp:45:10:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kvariable ‘\u001b[01m\u001b[KhaveFiles\u001b[m\u001b[K’ set but not used [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wunused-but-set-variable\u0007-Wunused-but-set-variable\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "   45 |     bool \u001b[01;35m\u001b[KhaveFiles\u001b[m\u001b[K         = false;\n",
            "      |          \u001b[01;35m\u001b[K^~~~~~~~~\u001b[m\u001b[K\n",
            "\u001b[01m\u001b[Ksrc/multiIntersectBed/multiIntersectBedMain.cpp:47:10:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kvariable ‘\u001b[01m\u001b[KhaveGenome\u001b[m\u001b[K’ set but not used [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wunused-but-set-variable\u0007-Wunused-but-set-variable\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "   47 |     bool \u001b[01;35m\u001b[KhaveGenome\u001b[m\u001b[K        = false;\n",
            "      |          \u001b[01;35m\u001b[K^~~~~~~~~~\u001b[m\u001b[K\n",
            "\u001b[01m\u001b[Ksrc/multiIntersectBed/multiIntersectBedMain.cpp:48:10:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kvariable ‘\u001b[01m\u001b[KhaveFiller\u001b[m\u001b[K’ set but not used [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wunused-but-set-variable\u0007-Wunused-but-set-variable\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "   48 |     bool \u001b[01;35m\u001b[KhaveFiller\u001b[m\u001b[K        = true;\n",
            "      |          \u001b[01;35m\u001b[K^~~~~~~~~~\u001b[m\u001b[K\n",
            "  * compiling src/multiIntersectBed/multiIntersectBed.cpp\n",
            "  * compiling src/nucBed/nucBedMain.cpp\n",
            "  * compiling src/nucBed/nucBed.cpp\n",
            "  * compiling src/pairToBed/pairToBedMain.cpp\n",
            "\u001b[01m\u001b[Ksrc/pairToBed/pairToBedMain.cpp:\u001b[m\u001b[K In function ‘\u001b[01m\u001b[Kint pairtobed_main(int, char**)\u001b[m\u001b[K’:\n",
            "\u001b[01m\u001b[Ksrc/pairToBed/pairToBedMain.cpp:43:10:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kvariable ‘\u001b[01m\u001b[KhaveFraction\u001b[m\u001b[K’ set but not used [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wunused-but-set-variable\u0007-Wunused-but-set-variable\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "   43 |     bool \u001b[01;35m\u001b[KhaveFraction\u001b[m\u001b[K       = false;\n",
            "      |          \u001b[01;35m\u001b[K^~~~~~~~~~~~\u001b[m\u001b[K\n",
            "  * compiling src/pairToBed/pairToBed.cpp\n",
            "\u001b[01m\u001b[Ksrc/pairToBed/pairToBed.cpp:\u001b[m\u001b[K In member function ‘\u001b[01m\u001b[Kvoid BedIntersectPE::FindSpanningOverlaps(const BEDPE&, std::vector<BED>&, const string&)\u001b[m\u001b[K’:\n",
            "\u001b[01m\u001b[Ksrc/pairToBed/pairToBed.cpp:257:12:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kvariable ‘\u001b[01m\u001b[KspanLength\u001b[m\u001b[K’ set but not used [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wunused-but-set-variable\u0007-Wunused-but-set-variable\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "  257 |     CHRPOS \u001b[01;35m\u001b[KspanLength\u001b[m\u001b[K = 0;\n",
            "      |            \u001b[01;35m\u001b[K^~~~~~~~~~\u001b[m\u001b[K\n",
            "\u001b[01m\u001b[Ksrc/pairToBed/pairToBed.cpp:\u001b[m\u001b[K In member function ‘\u001b[01m\u001b[Kbool BedIntersectPE::FindOneOrMoreSpanningOverlaps(const BEDPE&, const string&)\u001b[m\u001b[K’:\n",
            "\u001b[01m\u001b[Ksrc/pairToBed/pairToBed.cpp:310:12:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kvariable ‘\u001b[01m\u001b[KspanLength\u001b[m\u001b[K’ set but not used [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wunused-but-set-variable\u0007-Wunused-but-set-variable\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "  310 |     CHRPOS \u001b[01;35m\u001b[KspanLength\u001b[m\u001b[K = 0;\n",
            "      |            \u001b[01;35m\u001b[K^~~~~~~~~~\u001b[m\u001b[K\n",
            "In file included from \u001b[01m\u001b[Ksrc/utils/BamTools/include/api/BamReader.h:1\u001b[m\u001b[K,\n",
            "                 from \u001b[01m\u001b[Ksrc/pairToBed/pairToBed.h:15\u001b[m\u001b[K,\n",
            "                 from \u001b[01m\u001b[Ksrc/pairToBed/pairToBed.cpp:13\u001b[m\u001b[K:\n",
            "\u001b[01m\u001b[Ksrc/utils/BamTools/include/SamHeader.hpp:\u001b[m\u001b[K In function ‘\u001b[01m\u001b[Kint htslib_future::sam_hdr_rebuild(bam_hdr_t*)\u001b[m\u001b[K’:\n",
            "\u001b[01m\u001b[Ksrc/utils/BamTools/include/SamHeader.hpp:32:30:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kignoring return value of ‘\u001b[01m\u001b[Kint sam_hdr_write(samFile*, const bam_hdr_t*)\u001b[m\u001b[K’ declared with attribute ‘\u001b[01m\u001b[Kwarn_unused_result\u001b[m\u001b[K’ [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wunused-result\u0007-Wunused-result\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "   32 |                 \u001b[01;35m\u001b[Ksam_hdr_write(fp, hdr)\u001b[m\u001b[K;\n",
            "      |                 \u001b[01;35m\u001b[K~~~~~~~~~~~~~^~~~~~~~~\u001b[m\u001b[K\n",
            "  * compiling src/pairToPair/pairToPairMain.cpp\n",
            "\u001b[01m\u001b[Ksrc/pairToPair/pairToPairMain.cpp:\u001b[m\u001b[K In function ‘\u001b[01m\u001b[Kint pairtopair_main(int, char**)\u001b[m\u001b[K’:\n",
            "\u001b[01m\u001b[Ksrc/pairToPair/pairToPairMain.cpp:44:10:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kvariable ‘\u001b[01m\u001b[KhaveFraction\u001b[m\u001b[K’ set but not used [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wunused-but-set-variable\u0007-Wunused-but-set-variable\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "   44 |     bool \u001b[01;35m\u001b[KhaveFraction\u001b[m\u001b[K = false;\n",
            "      |          \u001b[01;35m\u001b[K^~~~~~~~~~~~\u001b[m\u001b[K\n",
            "  * compiling src/pairToPair/pairToPair.cpp\n",
            "\u001b[01m\u001b[Ksrc/pairToPair/pairToPair.cpp:\u001b[m\u001b[K In member function ‘\u001b[01m\u001b[Kvoid PairToPair::FindOverlaps(const BEDPE&)\u001b[m\u001b[K’:\n",
            "\u001b[01m\u001b[Ksrc/pairToPair/pairToPair.cpp:112:14:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kvariable ‘\u001b[01m\u001b[Kfound1\u001b[m\u001b[K’ set but not used [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wunused-but-set-variable\u0007-Wunused-but-set-variable\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "  112 |         bool \u001b[01;35m\u001b[Kfound1\u001b[m\u001b[K = false;\n",
            "      |              \u001b[01;35m\u001b[K^~~~~~\u001b[m\u001b[K\n",
            "\u001b[01m\u001b[Ksrc/pairToPair/pairToPair.cpp:113:14:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kvariable ‘\u001b[01m\u001b[Kfound2\u001b[m\u001b[K’ set but not used [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wunused-but-set-variable\u0007-Wunused-but-set-variable\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "  113 |         bool \u001b[01;35m\u001b[Kfound2\u001b[m\u001b[K = false;\n",
            "      |              \u001b[01;35m\u001b[K^~~~~~\u001b[m\u001b[K\n",
            "  * compiling src/randomBed/randomBedMain.cpp\n",
            "  * compiling src/randomBed/randomBed.cpp\n",
            "  * compiling src/regressTest/regressTestMain.cpp\n",
            "  * compiling src/regressTest/RegressTest.cpp\n",
            "\u001b[01m\u001b[Ksrc/regressTest/RegressTest.cpp:\u001b[m\u001b[K In member function ‘\u001b[01m\u001b[Kbool RegressTest::executeAndCompareCorrectness(const fileListType&)\u001b[m\u001b[K’:\n",
            "\u001b[01m\u001b[Ksrc/regressTest/RegressTest.cpp:431:15:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kignoring return value of ‘\u001b[01m\u001b[Kint system(const char*)\u001b[m\u001b[K’ declared with attribute ‘\u001b[01m\u001b[Kwarn_unused_result\u001b[m\u001b[K’ [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wunused-result\u0007-Wunused-result\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "  431 |         \u001b[01;35m\u001b[Ksystem(diffCmd.c_str())\u001b[m\u001b[K;\n",
            "      |         \u001b[01;35m\u001b[K~~~~~~^~~~~~~~~~~~~~~~~\u001b[m\u001b[K\n",
            "\u001b[01m\u001b[Ksrc/regressTest/RegressTest.cpp:\u001b[m\u001b[K In member function ‘\u001b[01m\u001b[Kbool RegressTest::startMemoryProfile(bool)\u001b[m\u001b[K’:\n",
            "\u001b[01m\u001b[Ksrc/regressTest/RegressTest.cpp:573:14:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kignoring return value of ‘\u001b[01m\u001b[Kchar* fgets(char*, int, FILE*)\u001b[m\u001b[K’ declared with attribute ‘\u001b[01m\u001b[Kwarn_unused_result\u001b[m\u001b[K’ [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wunused-result\u0007-Wunused-result\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "  573 |         \u001b[01;35m\u001b[Kfgets(sLine, 4192, fp)\u001b[m\u001b[K;\n",
            "      |         \u001b[01;35m\u001b[K~~~~~^~~~~~~~~~~~~~~~~\u001b[m\u001b[K\n",
            "\u001b[01m\u001b[Ksrc/regressTest/RegressTest.cpp:\u001b[m\u001b[K In member function ‘\u001b[01m\u001b[Kbool RegressTest::calcMemoryStats()\u001b[m\u001b[K’:\n",
            "\u001b[01m\u001b[Ksrc/regressTest/RegressTest.cpp:622:22:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kignoring return value of ‘\u001b[01m\u001b[Kchar* fgets(char*, int, FILE*)\u001b[m\u001b[K’ declared with attribute ‘\u001b[01m\u001b[Kwarn_unused_result\u001b[m\u001b[K’ [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wunused-result\u0007-Wunused-result\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "  622 |                 \u001b[01;35m\u001b[Kfgets(sLine, 4192, fp)\u001b[m\u001b[K;\n",
            "      |                 \u001b[01;35m\u001b[K~~~~~^~~~~~~~~~~~~~~~~\u001b[m\u001b[K\n",
            "  * compiling src/reldist/reldistMain.cpp\n",
            "  * compiling src/reldist/reldist.cpp\n",
            "\u001b[01m\u001b[Ksrc/reldist/reldist.cpp:\u001b[m\u001b[K In member function ‘\u001b[01m\u001b[Kvoid RelativeDistance::CalculateRelativeDistance()\u001b[m\u001b[K’:\n",
            "\u001b[01m\u001b[Ksrc/reldist/reldist.cpp:129:25:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kcomparison of integer expressions of different signedness: ‘\u001b[01m\u001b[KCHRPOS\u001b[m\u001b[K’ {aka ‘\u001b[01m\u001b[Klong int\u001b[m\u001b[K’} and ‘\u001b[01m\u001b[Kstd::vector<long int>::size_type\u001b[m\u001b[K’ {aka ‘\u001b[01m\u001b[Klong unsigned int\u001b[m\u001b[K’} [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wsign-compare\u0007-Wsign-compare\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "  129 |             if (\u001b[01;35m\u001b[Klow_idx != chrom_mids->size() - 1\u001b[m\u001b[K)\n",
            "      |                 \u001b[01;35m\u001b[K~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~\u001b[m\u001b[K\n",
            "  * compiling src/sampleFile/sampleHelp.cpp\n",
            "  * compiling src/sampleFile/sampleFile.cpp\n",
            "  * compiling src/shiftBed/shiftBedMain.cpp\n",
            "  * compiling src/shiftBed/shiftBed.cpp\n",
            "  * compiling src/shuffleBed/shuffleBedMain.cpp\n",
            "  * compiling src/shuffleBed/shuffleBed.cpp\n",
            "  * compiling src/slopBed/slopBedMain.cpp\n",
            "  * compiling src/slopBed/slopBed.cpp\n",
            "  * compiling src/sortBed/sortMain.cpp\n",
            "  * compiling src/sortBed/sortBed.cpp\n",
            "  * compiling src/spacingFile/spacingHelp.cpp\n",
            "  * compiling src/spacingFile/spacingFile.cpp\n",
            "  * compiling src/split/splitBedMain.cpp\n",
            "  * compiling src/split/splitBed.cpp\n",
            "  * compiling src/subtractFile/subtractHelp.cpp\n",
            "  * compiling src/subtractFile/subtractFile.cpp\n",
            "  * compiling src/summaryFile/summaryHelp.cpp\n",
            "  * compiling src/summaryFile/summaryFile.cpp\n",
            "  * compiling src/tagBam/tagBamMain.cpp\n",
            "  * compiling src/tagBam/tagBam.cpp\n",
            "In file included from \u001b[01m\u001b[Ksrc/utils/BamTools/include/api/BamReader.h:1\u001b[m\u001b[K,\n",
            "                 from \u001b[01m\u001b[Ksrc/tagBam/tagBam.h:17\u001b[m\u001b[K,\n",
            "                 from \u001b[01m\u001b[Ksrc/tagBam/tagBam.cpp:13\u001b[m\u001b[K:\n",
            "\u001b[01m\u001b[Ksrc/utils/BamTools/include/SamHeader.hpp:\u001b[m\u001b[K In function ‘\u001b[01m\u001b[Kint htslib_future::sam_hdr_rebuild(bam_hdr_t*)\u001b[m\u001b[K’:\n",
            "\u001b[01m\u001b[Ksrc/utils/BamTools/include/SamHeader.hpp:32:30:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kignoring return value of ‘\u001b[01m\u001b[Kint sam_hdr_write(samFile*, const bam_hdr_t*)\u001b[m\u001b[K’ declared with attribute ‘\u001b[01m\u001b[Kwarn_unused_result\u001b[m\u001b[K’ [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wunused-result\u0007-Wunused-result\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "   32 |                 \u001b[01;35m\u001b[Ksam_hdr_write(fp, hdr)\u001b[m\u001b[K;\n",
            "      |                 \u001b[01;35m\u001b[K~~~~~~~~~~~~~^~~~~~~~~\u001b[m\u001b[K\n",
            "  * compiling src/unionBedGraphs/unionBedGraphsMain.cpp\n",
            "\u001b[01m\u001b[Ksrc/unionBedGraphs/unionBedGraphsMain.cpp:\u001b[m\u001b[K In function ‘\u001b[01m\u001b[Kint unionbedgraphs_main(int, char**)\u001b[m\u001b[K’:\n",
            "\u001b[01m\u001b[Ksrc/unionBedGraphs/unionBedGraphsMain.cpp:46:10:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kvariable ‘\u001b[01m\u001b[KhaveFiles\u001b[m\u001b[K’ set but not used [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wunused-but-set-variable\u0007-Wunused-but-set-variable\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "   46 |     bool \u001b[01;35m\u001b[KhaveFiles\u001b[m\u001b[K         = false;\n",
            "      |          \u001b[01;35m\u001b[K^~~~~~~~~\u001b[m\u001b[K\n",
            "\u001b[01m\u001b[Ksrc/unionBedGraphs/unionBedGraphsMain.cpp:48:10:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kvariable ‘\u001b[01m\u001b[KhaveGenome\u001b[m\u001b[K’ set but not used [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wunused-but-set-variable\u0007-Wunused-but-set-variable\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "   48 |     bool \u001b[01;35m\u001b[KhaveGenome\u001b[m\u001b[K        = false;\n",
            "      |          \u001b[01;35m\u001b[K^~~~~~~~~~\u001b[m\u001b[K\n",
            "\u001b[01m\u001b[Ksrc/unionBedGraphs/unionBedGraphsMain.cpp:49:10:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kvariable ‘\u001b[01m\u001b[KhaveFiller\u001b[m\u001b[K’ set but not used [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wunused-but-set-variable\u0007-Wunused-but-set-variable\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "   49 |     bool \u001b[01;35m\u001b[KhaveFiller\u001b[m\u001b[K        = true;\n",
            "      |          \u001b[01;35m\u001b[K^~~~~~~~~~\u001b[m\u001b[K\n",
            "  * compiling src/unionBedGraphs/unionBedGraphs.cpp\n",
            "  * compiling src/windowBed/windowMain.cpp\n",
            "  * compiling src/windowBed/windowBed.cpp\n",
            "In file included from \u001b[01m\u001b[Ksrc/utils/BamTools/include/api/BamReader.h:1\u001b[m\u001b[K,\n",
            "                 from \u001b[01m\u001b[Ksrc/windowBed/windowBed.h:15\u001b[m\u001b[K,\n",
            "                 from \u001b[01m\u001b[Ksrc/windowBed/windowBed.cpp:13\u001b[m\u001b[K:\n",
            "\u001b[01m\u001b[Ksrc/utils/BamTools/include/SamHeader.hpp:\u001b[m\u001b[K In function ‘\u001b[01m\u001b[Kint htslib_future::sam_hdr_rebuild(bam_hdr_t*)\u001b[m\u001b[K’:\n",
            "\u001b[01m\u001b[Ksrc/utils/BamTools/include/SamHeader.hpp:32:30:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kignoring return value of ‘\u001b[01m\u001b[Kint sam_hdr_write(samFile*, const bam_hdr_t*)\u001b[m\u001b[K’ declared with attribute ‘\u001b[01m\u001b[Kwarn_unused_result\u001b[m\u001b[K’ [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wunused-result\u0007-Wunused-result\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "   32 |                 \u001b[01;35m\u001b[Ksam_hdr_write(fp, hdr)\u001b[m\u001b[K;\n",
            "      |                 \u001b[01;35m\u001b[K~~~~~~~~~~~~~^~~~~~~~~\u001b[m\u001b[K\n",
            "  * compiling src/windowMaker/windowMakerMain.cpp\n",
            "  * compiling src/windowMaker/windowMaker.cpp\n",
            "  * compiling src/utils/FileRecordTools/FileRecordMgr.cpp\n",
            "  * compiling src/utils/FileRecordTools/FileRecordMergeMgr.cpp\n",
            "  * compiling src/utils/FileRecordTools/FileReaders/FileReader.cpp\n",
            "  * compiling src/utils/FileRecordTools/FileReaders/SingleLineDelimTextFileReader.cpp\n",
            "  * compiling src/utils/FileRecordTools/FileReaders/BamFileReader.cpp\n",
            "  * compiling src/utils/FileRecordTools/FileReaders/BufferedStreamMgr.cpp\n",
            "\u001b[01m\u001b[Ksrc/utils/FileRecordTools/FileReaders/BufferedStreamMgr.cpp:\u001b[m\u001b[K In member function ‘\u001b[01m\u001b[Kbool BufferedStreamMgr::readFileChunk()\u001b[m\u001b[K’:\n",
            "\u001b[01m\u001b[Ksrc/utils/FileRecordTools/FileReaders/BufferedStreamMgr.cpp:148:37:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kcomparison of integer expressions of different signedness: ‘\u001b[01m\u001b[Kint\u001b[m\u001b[K’ and ‘\u001b[01m\u001b[Ksize_t\u001b[m\u001b[K’ {aka ‘\u001b[01m\u001b[Klong unsigned int\u001b[m\u001b[K’} [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wsign-compare\u0007-Wsign-compare\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "  148 |                 if (\u001b[01;35m\u001b[K_mainBufCurrLen < _useBufSize\u001b[m\u001b[K) {\n",
            "      |                     \u001b[01;35m\u001b[K~~~~~~~~~~~~~~~~^~~~~~~~~~~~~\u001b[m\u001b[K\n",
            "  * compiling src/utils/FileRecordTools/FileReaders/InputStreamMgr.cpp\n",
            "In file included from \u001b[01m\u001b[Ksrc/utils/BamTools/include/api/BamReader.h:1\u001b[m\u001b[K,\n",
            "                 from \u001b[01m\u001b[Ksrc/utils/FileRecordTools/FileReaders/InputStreamMgr.h:14\u001b[m\u001b[K,\n",
            "                 from \u001b[01m\u001b[Ksrc/utils/FileRecordTools/FileReaders/InputStreamMgr.cpp:8\u001b[m\u001b[K:\n",
            "\u001b[01m\u001b[Ksrc/utils/BamTools/include/SamHeader.hpp:\u001b[m\u001b[K In function ‘\u001b[01m\u001b[Kint htslib_future::sam_hdr_rebuild(bam_hdr_t*)\u001b[m\u001b[K’:\n",
            "\u001b[01m\u001b[Ksrc/utils/BamTools/include/SamHeader.hpp:32:30:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kignoring return value of ‘\u001b[01m\u001b[Kint sam_hdr_write(samFile*, const bam_hdr_t*)\u001b[m\u001b[K’ declared with attribute ‘\u001b[01m\u001b[Kwarn_unused_result\u001b[m\u001b[K’ [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wunused-result\u0007-Wunused-result\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "   32 |                 \u001b[01;35m\u001b[Ksam_hdr_write(fp, hdr)\u001b[m\u001b[K;\n",
            "      |                 \u001b[01;35m\u001b[K~~~~~~~~~~~~~^~~~~~~~~\u001b[m\u001b[K\n",
            "  * compiling src/utils/FileRecordTools/Records/Record.cpp\n",
            "  * compiling src/utils/FileRecordTools/Records/EmptyRecord.cpp\n",
            "  * compiling src/utils/FileRecordTools/Records/Bed3Interval.cpp\n",
            "  * compiling src/utils/FileRecordTools/Records/Bed4Interval.cpp\n",
            "  * compiling src/utils/FileRecordTools/Records/BedGraphInterval.cpp\n",
            "  * compiling src/utils/FileRecordTools/Records/Bed5Interval.cpp\n",
            "  * compiling src/utils/FileRecordTools/Records/Bed6Interval.cpp\n",
            "  * compiling src/utils/FileRecordTools/Records/PlusFields.cpp\n",
            "  * compiling src/utils/FileRecordTools/Records/GffRecord.cpp\n",
            "  * compiling src/utils/FileRecordTools/Records/GffPlusRecord.cpp\n",
            "  * compiling src/utils/FileRecordTools/Records/NoPosPlusRecord.cpp\n",
            "  * compiling src/utils/FileRecordTools/Records/BedPlusInterval.cpp\n",
            "  * compiling src/utils/FileRecordTools/Records/Bed12Interval.cpp\n",
            "  * compiling src/utils/FileRecordTools/Records/BamRecord.cpp\n",
            "  * compiling src/utils/FileRecordTools/Records/VcfRecord.cpp\n",
            "  * compiling src/utils/FileRecordTools/Records/BlockMgr.cpp\n",
            "\u001b[01m\u001b[Ksrc/utils/FileRecordTools/Records/BlockMgr.cpp:\u001b[m\u001b[K In member function ‘\u001b[01m\u001b[Kint BlockMgr::getNonRedundantOverlap()\u001b[m\u001b[K’:\n",
            "\u001b[01m\u001b[Ksrc/utils/FileRecordTools/Records/BlockMgr.cpp:194:35:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kcomparison of integer expressions of different signedness: ‘\u001b[01m\u001b[Kint\u001b[m\u001b[K’ and ‘\u001b[01m\u001b[Kstd::vector<long int>::size_type\u001b[m\u001b[K’ {aka ‘\u001b[01m\u001b[Klong unsigned int\u001b[m\u001b[K’} [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wsign-compare\u0007-Wsign-compare\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "  194 |                 for (int i = 1; \u001b[01;35m\u001b[Ki < _overlapStarts.size()\u001b[m\u001b[K; ++i)\n",
            "      |                                 \u001b[01;35m\u001b[K~~^~~~~~~~~~~~~~~~~~~~~~~\u001b[m\u001b[K\n",
            "  * compiling src/utils/FileRecordTools/Records/StrandQueue.cpp\n",
            "  * compiling src/utils/FileRecordTools/Records/RecordMgr.cpp\n",
            "  * compiling src/utils/FileRecordTools/Records/RecordList.cpp\n",
            "  * compiling src/utils/FileRecordTools/Records/RecordKeyList.cpp\n",
            "  * compiling src/utils/FileRecordTools/Records/RecordKeyVector.cpp\n",
            "  * compiling src/utils/bedFile/bedFile.cpp\n",
            "  * compiling src/utils/BinTree/BinTree.cpp\n",
            "  * compiling src/utils/version/version.cpp\n",
            "  * compiling src/utils/bedGraphFile/bedGraphFile.cpp\n",
            "  * compiling src/utils/chromsweep/chromsweep.cpp\n",
            "  * compiling src/utils/Contexts/ContextBase.cpp\n",
            "  * compiling src/utils/Contexts/ContextIntersect.cpp\n",
            "  * compiling src/utils/Contexts/ContextFisher.cpp\n",
            "  * compiling src/utils/Contexts/ContextMap.cpp\n",
            "  * compiling src/utils/Contexts/ContextSample.cpp\n",
            "  * compiling src/utils/Contexts/ContextSpacing.cpp\n",
            "  * compiling src/utils/Contexts/ContextMerge.cpp\n",
            "  * compiling src/utils/Contexts/ContextJaccard.cpp\n",
            "  * compiling src/utils/Contexts/ContextClosest.cpp\n",
            "  * compiling src/utils/Contexts/ContextSubtract.cpp\n",
            "  * compiling src/utils/Contexts/ContextCoverage.cpp\n",
            "  * compiling src/utils/Contexts/ContextComplement.cpp\n",
            "  * compiling src/utils/Contexts/ContextGroupBy.cpp\n",
            "  * compiling src/utils/Contexts/ContextSummary.cpp\n",
            "  * compiling src/utils/general/ParseTools.cpp\n",
            "  * compiling src/utils/general/PushBackStreamBuf.cpp\n",
            "  * compiling src/utils/general/CompressionTools.cpp\n",
            "  * compiling src/utils/general/Tokenizer.cpp\n",
            "  * compiling src/utils/general/CommonHelp.cpp\n",
            "  * compiling src/utils/general/ErrorMsg.cpp\n",
            "  * compiling src/utils/general/Random.cpp\n",
            "g++ -g -Wall -O2 -std=c++11  -Isrc/utils/gzstream  -c -o obj/gzstream.o src/utils/gzstream/gzstream.C\n",
            "  * compiling src/utils/fileType/fileType.cpp\n",
            "  * compiling src/utils/fileType/FileRecordTypeChecker.cpp\n",
            "\u001b[01m\u001b[Ksrc/utils/fileType/FileRecordTypeChecker.cpp:\u001b[m\u001b[K In member function ‘\u001b[01m\u001b[Kbool FileRecordTypeChecker::isTextDelimtedFormat(const char*, size_t)\u001b[m\u001b[K’:\n",
            "\u001b[01m\u001b[Ksrc/utils/fileType/FileRecordTypeChecker.cpp:366:33:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kcomparison of integer expressions of different signedness: ‘\u001b[01m\u001b[Kint\u001b[m\u001b[K’ and ‘\u001b[01m\u001b[Ksize_t\u001b[m\u001b[K’ {aka ‘\u001b[01m\u001b[Klong unsigned int\u001b[m\u001b[K’} [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wsign-compare\u0007-Wsign-compare\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "  366 |                 for (int j=0; \u001b[01;35m\u001b[Kj < len\u001b[m\u001b[K; j++) {\n",
            "      |                               \u001b[01;35m\u001b[K~~^~~~~\u001b[m\u001b[K\n",
            "  * compiling src/utils/bedFilePE/bedFilePE.cpp\n",
            "  * compiling src/utils/KeyListOps/KeyListOps.cpp\n",
            "  * compiling src/utils/KeyListOps/KeyListOpsMethods.cpp\n",
            "  * compiling src/utils/NewChromsweep/NewChromsweep.cpp\n",
            "  * compiling src/utils/NewChromsweep/CloseSweep.cpp\n",
            "  * compiling src/utils/sequenceUtilities/sequenceUtils.cpp\n",
            "  * compiling src/utils/tabFile/tabFile.cpp\n",
            "  * compiling src/utils/BamTools-Ancillary/BamAncillary.cpp\n",
            "  * compiling src/utils/BlockedIntervals/BlockedIntervals.cpp\n",
            "  * compiling src/utils/Fasta/Fasta.cpp\n",
            "\u001b[01m\u001b[Ksrc/utils/Fasta/Fasta.cpp:\u001b[m\u001b[K In member function ‘\u001b[01m\u001b[Kstd::string FastaReference::getSequence(std::string)\u001b[m\u001b[K’:\n",
            "\u001b[01m\u001b[Ksrc/utils/Fasta/Fasta.cpp:292:14:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kignoring return value of ‘\u001b[01m\u001b[Ksize_t fread(void*, size_t, size_t, FILE*)\u001b[m\u001b[K’ declared with attribute ‘\u001b[01m\u001b[Kwarn_unused_result\u001b[m\u001b[K’ [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wunused-result\u0007-Wunused-result\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "  292 |         \u001b[01;35m\u001b[Kfread(seq, sizeof(char), seqlen, file)\u001b[m\u001b[K;\n",
            "      |         \u001b[01;35m\u001b[K~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\u001b[m\u001b[K\n",
            "\u001b[01m\u001b[Ksrc/utils/Fasta/Fasta.cpp:\u001b[m\u001b[K In member function ‘\u001b[01m\u001b[Kstd::string FastaReference::getSubSequence(std::string, CHRPOS, CHRPOS)\u001b[m\u001b[K’:\n",
            "\u001b[01m\u001b[Ksrc/utils/Fasta/Fasta.cpp:334:14:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kignoring return value of ‘\u001b[01m\u001b[Ksize_t fread(void*, size_t, size_t, FILE*)\u001b[m\u001b[K’ declared with attribute ‘\u001b[01m\u001b[Kwarn_unused_result\u001b[m\u001b[K’ [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wunused-result\u0007-Wunused-result\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "  334 |         \u001b[01;35m\u001b[Kfread(seq, sizeof(char), (off_t) seqlen, file)\u001b[m\u001b[K;\n",
            "      |         \u001b[01;35m\u001b[K~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\u001b[m\u001b[K\n",
            "  * compiling src/utils/Fasta/split.cpp\n",
            "  * compiling src/utils/VectorOps/VectorOps.cpp\n",
            "  * compiling src/utils/GenomeFile/NewGenomeFile.cpp\n",
            "  * compiling src/utils/GenomeFile/GenomeFile.cpp\n",
            "\u001b[01m\u001b[Ksrc/utils/GenomeFile/GenomeFile.cpp:\u001b[m\u001b[K In member function ‘\u001b[01m\u001b[Kvoid GenomeFile::loadGenomeFileIntoMap()\u001b[m\u001b[K’:\n",
            "\u001b[01m\u001b[Ksrc/utils/GenomeFile/GenomeFile.cpp:60:26:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kvariable ‘\u001b[01m\u001b[Kc2\u001b[m\u001b[K’ set but not used [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wunused-but-set-variable\u0007-Wunused-but-set-variable\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "   60 |                     long \u001b[01;35m\u001b[Kc2\u001b[m\u001b[K;\n",
            "      |                          \u001b[01;35m\u001b[K^~\u001b[m\u001b[K\n",
            "  * compiling src/utils/RecordOutputMgr/RecordOutputMgr.cpp\n",
            "\u001b[01m\u001b[Ksrc/utils/RecordOutputMgr/RecordOutputMgr.cpp:\u001b[m\u001b[K In member function ‘\u001b[01m\u001b[Kvoid RecordOutputMgr::printClosest(RecordKeyVector&, const std::vector<long int>*)\u001b[m\u001b[K’:\n",
            "\u001b[01m\u001b[Ksrc/utils/RecordOutputMgr/RecordOutputMgr.cpp:170:56:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kcomparison of integer expressions of different signedness: ‘\u001b[01m\u001b[Kint\u001b[m\u001b[K’ and ‘\u001b[01m\u001b[Ksize_t\u001b[m\u001b[K’ {aka ‘\u001b[01m\u001b[Klong unsigned int\u001b[m\u001b[K’} [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wsign-compare\u0007-Wsign-compare\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "  170 |                 if (\u001b[01;35m\u001b[Kcontext->getNumClosestHitsWanted() > keyList.size()\u001b[m\u001b[K)\n",
            "      |                     \u001b[01;35m\u001b[K~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~\u001b[m\u001b[K\n",
            "\u001b[01m\u001b[Ksrc/utils/RecordOutputMgr/RecordOutputMgr.cpp:\u001b[m\u001b[K In member function ‘\u001b[01m\u001b[Kvoid RecordOutputMgr::reportOverlapSummary(RecordKeyVector&)\u001b[m\u001b[K’:\n",
            "\u001b[01m\u001b[Ksrc/utils/RecordOutputMgr/RecordOutputMgr.cpp:455:38:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Kcomparison of integer expressions of different signedness: ‘\u001b[01m\u001b[Ksize_t\u001b[m\u001b[K’ {aka ‘\u001b[01m\u001b[Klong unsigned int\u001b[m\u001b[K’} and ‘\u001b[01m\u001b[Kint\u001b[m\u001b[K’ [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wsign-compare\u0007-Wsign-compare\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "  455 |                 for (size_t i = 1; \u001b[01;35m\u001b[Ki < _context->getNumInputFiles()\u001b[m\u001b[K; i++)\n",
            "      |                                    \u001b[01;35m\u001b[K~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\u001b[m\u001b[K\n",
            "  * compiling src/utils/ToolBase/ToolBase.cpp\n",
            "  * compiling src/utils/driver/BedtoolsDriver.cpp\n",
            "cd src/utils/htslib && make config.h\n",
            "make[1]: Entering directory '/content/bedtools2/src/utils/htslib'\n",
            "echo '/* Default config.h generated by Makefile */' > config.h\n",
            "echo '#define HAVE_LIBBZ2 1' >> config.h\n",
            "echo '#define HAVE_LIBLZMA 1' >> config.h\n",
            "echo '#ifndef __APPLE__' >> config.h\n",
            "echo '#define HAVE_LZMA_H 1' >> config.h\n",
            "echo '#endif' >> config.h\n",
            "echo '#define HAVE_FSEEKO 1' >> config.h\n",
            "echo '#define HAVE_DRAND48 1' >> config.h\n",
            "make[1]: Leaving directory '/content/bedtools2/src/utils/htslib'\n",
            "cd src/utils/htslib && make lib-static\n",
            "make[1]: Entering directory '/content/bedtools2/src/utils/htslib'\n",
            "gcc -g -Wall -O2 -I.  -c -o kfunc.o kfunc.c\n",
            "gcc -g -Wall -O2 -I.  -c -o knetfile.o knetfile.c\n",
            "gcc -g -Wall -O2 -I.  -c -o kstring.o kstring.c\n",
            "gcc -g -Wall -O2 -I.  -c -o bcf_sr_sort.o bcf_sr_sort.c\n",
            "gcc -g -Wall -O2 -I.  -c -o bgzf.o bgzf.c\n",
            "gcc -g -Wall -O2 -I.  -c -o errmod.o errmod.c\n",
            "gcc -g -Wall -O2 -I.  -c -o faidx.o faidx.c\n",
            "gcc -g -Wall -O2 -I.  -c -o hfile.o hfile.c\n",
            "gcc -g -Wall -O2 -I.  -c -o hfile_net.o hfile_net.c\n",
            "echo '#define HTS_VERSION \"1.9\"' > version.h\n",
            "gcc -g -Wall -O2 -I.  -c -o hts.o hts.c\n",
            "\u001b[01m\u001b[Khts.c:\u001b[m\u001b[K In function ‘\u001b[01m\u001b[Khts_idx_init\u001b[m\u001b[K’:\n",
            "\u001b[01m\u001b[Khts.c:1397:59:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[Koverflow in conversion from ‘\u001b[01m\u001b[Kuint32_t\u001b[m\u001b[K’ {aka ‘\u001b[01m\u001b[Kunsigned int\u001b[m\u001b[K’} to ‘\u001b[01m\u001b[Kint\u001b[m\u001b[K’ changes value from ‘\u001b[01m\u001b[Kidx->z.last_bin = 4294967295\u001b[m\u001b[K’ to ‘\u001b[01m\u001b[K-1\u001b[m\u001b[K’ [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Woverflow\u0007-Woverflow\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            " 1397 |     idx->z.save_bin = idx->z.save_tid = idx->z.last_tid = \u001b[01;35m\u001b[Kidx\u001b[m\u001b[K->z.last_bin = 0xffffffffu;\n",
            "      |                                                           \u001b[01;35m\u001b[K^~~\u001b[m\u001b[K\n",
            "gcc -g -Wall -O2 -I.  -c -o hts_os.o hts_os.c\n",
            "gcc -g -Wall -O2 -I.  -c -o md5.o md5.c\n",
            "gcc -g -Wall -O2 -I.  -c -o multipart.o multipart.c\n",
            "gcc -g -Wall -O2 -I.  -c -o probaln.o probaln.c\n",
            "gcc -g -Wall -O2 -I.  -c -o realn.o realn.c\n",
            "gcc -g -Wall -O2 -I.  -c -o regidx.o regidx.c\n",
            "gcc -g -Wall -O2 -I.  -c -o sam.o sam.c\n",
            "gcc -g -Wall -O2 -I.  -c -o synced_bcf_reader.o synced_bcf_reader.c\n",
            "gcc -g -Wall -O2 -I.  -c -o vcf_sweep.o vcf_sweep.c\n",
            "gcc -g -Wall -O2 -I.  -c -o tbx.o tbx.c\n",
            "gcc -g -Wall -O2 -I.  -c -o textutils.o textutils.c\n",
            "gcc -g -Wall -O2 -I.  -c -o thread_pool.o thread_pool.c\n",
            "gcc -g -Wall -O2 -I.  -c -o vcf.o vcf.c\n",
            "gcc -g -Wall -O2 -I.  -c -o vcfutils.o vcfutils.c\n",
            "gcc -g -Wall -O2 -I.  -c -o cram/cram_codecs.o cram/cram_codecs.c\n",
            "gcc -g -Wall -O2 -I.  -c -o cram/cram_decode.o cram/cram_decode.c\n",
            "gcc -g -Wall -O2 -I.  -c -o cram/cram_encode.o cram/cram_encode.c\n",
            "gcc -g -Wall -O2 -I.  -c -o cram/cram_external.o cram/cram_external.c\n",
            "gcc -g -Wall -O2 -I.  -c -o cram/cram_index.o cram/cram_index.c\n",
            "gcc -g -Wall -O2 -I.  -c -o cram/cram_io.o cram/cram_io.c\n",
            "In file included from \u001b[01m\u001b[K/usr/include/string.h:535\u001b[m\u001b[K,\n",
            "                 from \u001b[01m\u001b[Kcram/cram_io.c:53\u001b[m\u001b[K:\n",
            "In function ‘\u001b[01m\u001b[Kstrncpy\u001b[m\u001b[K’,\n",
            "    inlined from ‘\u001b[01m\u001b[Kcram_populate_ref\u001b[m\u001b[K’ at \u001b[01m\u001b[Kcram/cram_io.c:2095:9\u001b[m\u001b[K:\n",
            "\u001b[01m\u001b[K/usr/include/x86_64-linux-gnu/bits/string_fortified.h:95:10:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[K‘\u001b[01m\u001b[K__builtin_strncpy\u001b[m\u001b[K’ specified bound 4096 equals destination size [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wstringop-truncation\u0007-Wstringop-truncation\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "   95 |   return \u001b[01;35m\u001b[K__builtin___strncpy_chk (__dest, __src, __len,\u001b[m\u001b[K\n",
            "      |          \u001b[01;35m\u001b[K^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\u001b[m\u001b[K\n",
            "   96 | \u001b[01;35m\u001b[K                                  __glibc_objsize (__dest))\u001b[m\u001b[K;\n",
            "      |                                   \u001b[01;35m\u001b[K~~~~~~~~~~~~~~~~~~~~~~~~~\u001b[m\u001b[K\n",
            "In function ‘\u001b[01m\u001b[Kstrncpy\u001b[m\u001b[K’,\n",
            "    inlined from ‘\u001b[01m\u001b[Kfull_path\u001b[m\u001b[K’ at \u001b[01m\u001b[Kcram/cram_io.c:3748:13\u001b[m\u001b[K,\n",
            "    inlined from ‘\u001b[01m\u001b[Kcram_write_SAM_hdr\u001b[m\u001b[K’ at \u001b[01m\u001b[Kcram/cram_io.c:3823:17\u001b[m\u001b[K:\n",
            "\u001b[01m\u001b[K/usr/include/x86_64-linux-gnu/bits/string_fortified.h:95:10:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[K‘\u001b[01m\u001b[K__builtin_strncpy\u001b[m\u001b[K’ specified bound 4096 equals destination size [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wstringop-truncation\u0007-Wstringop-truncation\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "   95 |   return \u001b[01;35m\u001b[K__builtin___strncpy_chk (__dest, __src, __len,\u001b[m\u001b[K\n",
            "      |          \u001b[01;35m\u001b[K^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\u001b[m\u001b[K\n",
            "   96 | \u001b[01;35m\u001b[K                                  __glibc_objsize (__dest))\u001b[m\u001b[K;\n",
            "      |                                   \u001b[01;35m\u001b[K~~~~~~~~~~~~~~~~~~~~~~~~~\u001b[m\u001b[K\n",
            "In function ‘\u001b[01m\u001b[Kstrncpy\u001b[m\u001b[K’,\n",
            "    inlined from ‘\u001b[01m\u001b[Kcram_dopen\u001b[m\u001b[K’ at \u001b[01m\u001b[Kcram/cram_io.c:4126:9\u001b[m\u001b[K:\n",
            "\u001b[01m\u001b[K/usr/include/x86_64-linux-gnu/bits/string_fortified.h:95:10:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[K‘\u001b[01m\u001b[K__builtin_strncpy\u001b[m\u001b[K’ specified bound 20 equals destination size [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wstringop-truncation\u0007-Wstringop-truncation\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "   95 |   return \u001b[01;35m\u001b[K__builtin___strncpy_chk (__dest, __src, __len,\u001b[m\u001b[K\n",
            "      |          \u001b[01;35m\u001b[K^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\u001b[m\u001b[K\n",
            "   96 | \u001b[01;35m\u001b[K                                  __glibc_objsize (__dest))\u001b[m\u001b[K;\n",
            "      |                                   \u001b[01;35m\u001b[K~~~~~~~~~~~~~~~~~~~~~~~~~\u001b[m\u001b[K\n",
            "gcc -g -Wall -O2 -I.  -c -o cram/cram_samtools.o cram/cram_samtools.c\n",
            "gcc -g -Wall -O2 -I.  -c -o cram/cram_stats.o cram/cram_stats.c\n",
            "gcc -g -Wall -O2 -I.  -c -o cram/files.o cram/files.c\n",
            "gcc -g -Wall -O2 -I.  -c -o cram/mFILE.o cram/mFILE.c\n",
            "gcc -g -Wall -O2 -I.  -c -o cram/open_trace_file.o cram/open_trace_file.c\n",
            "gcc -g -Wall -O2 -I.  -c -o cram/pooled_alloc.o cram/pooled_alloc.c\n",
            "gcc -g -Wall -O2 -I.  -c -o cram/rANS_static.o cram/rANS_static.c\n",
            "gcc -g -Wall -O2 -I.  -c -o cram/sam_header.o cram/sam_header.c\n",
            "gcc -g -Wall -O2 -I.  -c -o cram/string_alloc.o cram/string_alloc.c\n",
            "In file included from \u001b[01m\u001b[K/usr/include/string.h:535\u001b[m\u001b[K,\n",
            "                 from \u001b[01m\u001b[Kcram/string_alloc.c:41\u001b[m\u001b[K:\n",
            "In function ‘\u001b[01m\u001b[Kstrncpy\u001b[m\u001b[K’,\n",
            "    inlined from ‘\u001b[01m\u001b[Kstring_ndup\u001b[m\u001b[K’ at \u001b[01m\u001b[Kcram/string_alloc.c:151:5\u001b[m\u001b[K,\n",
            "    inlined from ‘\u001b[01m\u001b[Kstring_dup\u001b[m\u001b[K’ at \u001b[01m\u001b[Kcram/string_alloc.c:143:12\u001b[m\u001b[K:\n",
            "\u001b[01m\u001b[K/usr/include/x86_64-linux-gnu/bits/string_fortified.h:95:10:\u001b[m\u001b[K \u001b[01;35m\u001b[Kwarning: \u001b[m\u001b[K‘\u001b[01m\u001b[K__builtin_strncpy\u001b[m\u001b[K’ output truncated before terminating nul copying as many bytes from a string as its length [\u001b[01;35m\u001b[K\u001b]8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wstringop-truncation\u0007-Wstringop-truncation\u001b]8;;\u0007\u001b[m\u001b[K]\n",
            "   95 |   return \u001b[01;35m\u001b[K__builtin___strncpy_chk (__dest, __src, __len,\u001b[m\u001b[K\n",
            "      |          \u001b[01;35m\u001b[K^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\u001b[m\u001b[K\n",
            "   96 | \u001b[01;35m\u001b[K                                  __glibc_objsize (__dest))\u001b[m\u001b[K;\n",
            "      |                                   \u001b[01;35m\u001b[K~~~~~~~~~~~~~~~~~~~~~~~~~\u001b[m\u001b[K\n",
            "\u001b[01m\u001b[Kcram/string_alloc.c:\u001b[m\u001b[K In function ‘\u001b[01m\u001b[Kstring_dup\u001b[m\u001b[K’:\n",
            "\u001b[01m\u001b[Kcram/string_alloc.c:143:12:\u001b[m\u001b[K \u001b[01;36m\u001b[Knote: \u001b[m\u001b[Klength computed here\n",
            "  143 |     return \u001b[01;36m\u001b[Kstring_ndup(a_str, instr, strlen(instr))\u001b[m\u001b[K;\n",
            "      |            \u001b[01;36m\u001b[K^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\u001b[m\u001b[K\n",
            "ar -rc libhts.a kfunc.o knetfile.o kstring.o bcf_sr_sort.o bgzf.o errmod.o faidx.o hfile.o hfile_net.o hts.o hts_os.o md5.o multipart.o probaln.o realn.o regidx.o sam.o synced_bcf_reader.o vcf_sweep.o tbx.o textutils.o thread_pool.o vcf.o vcfutils.o cram/cram_codecs.o cram/cram_decode.o cram/cram_encode.o cram/cram_external.o cram/cram_index.o cram/cram_io.o cram/cram_samtools.o cram/cram_stats.o cram/files.o cram/mFILE.o cram/open_trace_file.o cram/pooled_alloc.o cram/rANS_static.o cram/sam_header.o cram/string_alloc.o\n",
            "ranlib libhts.a\n",
            "make[1]: Leaving directory '/content/bedtools2/src/utils/htslib'\n",
            "- Building main bedtools binary.\n",
            "done.\n",
            "- Creating executables for old CLI.\n",
            "done.\n",
            "  * compiling src/htsutil.cpp\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!cp /content/bedtools2/bin/* /usr/local/bin/"
      ],
      "metadata": {
        "id": "k-lztZHWaoee"
      },
      "execution_count": 57,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "#### make genome file"
      ],
      "metadata": {
        "id": "59fN6cd7aeNU"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "genomic_file = 'GCF_000151665.1_ASM15166v1_genomic.fna.gz'\n",
        "\n",
        "!wget {base_path}/{genomic_file} -q && gunzip -d {genomic_file} -q && mv {genomic_file[:-3]} genome.fna"
      ],
      "metadata": {
        "id": "-IAd1L3md2Nq"
      },
      "execution_count": 58,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "cnt = 1\n",
        "dictionary = {}\n",
        "lst = []\n",
        "len_lst = []\n",
        "lengths = [0]\n",
        "with open('genome.fna', 'r') as f:\n",
        "    for line in f:\n",
        "        if line.startswith('>'):\n",
        "            dictionary[line.split()[0][1:]] = f'scaffold_{cnt}'\n",
        "            line = f'>scaffold_{cnt}\\n'\n",
        "            if lengths != [0]:\n",
        "                len_lst.append(f'scaffold_{cnt - 1}\\t{lengths[-1]}\\n')\n",
        "                lengths.append(0)\n",
        "            cnt += 1\n",
        "        else:\n",
        "            lengths[-1] += len(line)\n",
        "        lst.append(line)\n",
        "with open('ind_genome.fna', 'w') as f:\n",
        "    f.write(''.join(lst))\n",
        "with open('lengths.txt', 'w') as f:\n",
        "    f.write(''.join(len_lst))"
      ],
      "metadata": {
        "id": "UuQcdOusRm2v"
      },
      "execution_count": 59,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!head lengths.txt"
      ],
      "metadata": {
        "id": "6_7P5Upf9_yP",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "f196b4af-9c79-4bc9-819c-25135ac2b7cf"
      },
      "execution_count": 61,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "scaffold_1\t1990940\n",
            "scaffold_2\t1889918\n",
            "scaffold_3\t648327\n",
            "scaffold_4\t598918\n",
            "scaffold_5\t497173\n",
            "scaffold_6\t492884\n",
            "scaffold_7\t491985\n",
            "scaffold_8\t446119\n",
            "scaffold_9\t439394\n",
            "scaffold_10\t419422\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!head ind_genome.fna"
      ],
      "metadata": {
        "id": "ueZ_cYUISj-O",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "4c314668-aa2c-424b-ed7a-4f6e07175e35"
      },
      "execution_count": 62,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            ">scaffold_1\n",
            "CTGACTAAGCATCCCTCTTAAAAGTCGAGgctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccct\n",
            "aaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaa\n",
            "ccctacccatgaaccctgaccctagccctaaccctggccctcaactcgaaccctagccctaaccctggccctcaactcga\n",
            "accctagccctaaccctgaccctaacgGGTCAGGATAGGGTAATGAGAACCTGAAccgaagcatgaacccgaaccctagc\n",
            "cctaaccctggccctcgACTCGGATCcgaaccctgaccctagccctaaccctggccctcaactcgaatccgaaccctgac\n",
            "cctagcccgaacccgaaccctcgTGGGTTAGGAATggattagggttagggttaggaagaaacaagggttacgaaaagaat\n",
            "tagggttagggttaagaagaaagaagggttacgaaaagaattagggttagggttaggaagaagggttacgaaaagaatca\n",
            "gggttagggttaagaagaaagaagggttatACTAACCTGAAccgaagcatgaacccgaaccctagccctaaccctggccc\n",
            "tcaactcgaacccgaACCTGAAGCACGgaccctcaactcgaatccgaaccctgaccctagcCCTAGCCCGAACCCTCGAG\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#### find intersections"
      ],
      "metadata": {
        "id": "v2D8TWznah8S"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "gff_file = 'GCF_000151665.1_ASM15166v1_genomic.gff.gz'\n",
        "\n",
        "!wget {base_path}/{gff_file} -q && gunzip -d {gff_file} -q && mv {gff_file[:-3]} genomic.gff\n",
        "!head -n 20 genomic.gff"
      ],
      "metadata": {
        "id": "zIQMak9tdWq6",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "56abfc8a-d276-4eca-b850-48493155a685"
      },
      "execution_count": 64,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "##gff-version 3\n",
            "#!gff-spec-version 1.21\n",
            "#!processor NCBI annotwriter\n",
            "#!genome-build ASM15166v1\n",
            "#!genome-build-accession NCBI_Assembly:GCF_000151665.1\n",
            "##sequence-region NW_013171815.1 1 1966360\n",
            "##species https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?id=12968\n",
            "NW_013171815.1\tRefSeq\tregion\t1\t1966360\t.\t+\t.\tID=NW_013171815.1:1..1966360;Dbxref=taxon:12968;gbkey=Src;genome=genomic;isolate=Singapore isolate B (sub-type 7);mol_type=genomic DNA;note=scaffold_0\n",
            "NW_013171815.1\tRefSeq\tgene\t28397\t29875\t.\t-\t.\tID=gene-GSBLH_T00000003001;Dbxref=GeneID:24917328;Name=GSBLH_T00000003001;end_range=29875,.;gbkey=Gene;gene_biotype=protein_coding;locus_tag=GSBLH_T00000003001;partial=true;start_range=.,28397\n",
            "NW_013171815.1\tRefSeq\tmRNA\t28397\t29875\t.\t-\t.\tID=rna-XM_013038130.1;Parent=gene-GSBLH_T00000003001;Dbxref=GeneID:24917328,GenBank:XM_013038130.1;Name=XM_013038130.1;end_range=29875,.;gbkey=mRNA;locus_tag=GSBLH_T00000003001;partial=true;start_range=.,28397;transcript_id=XM_013038130.1\n",
            "NW_013171815.1\tRefSeq\texon\t29398\t29875\t.\t-\t.\tID=exon-XM_013038130.1-1;Parent=rna-XM_013038130.1;Dbxref=GeneID:24917328,GenBank:XM_013038130.1;end_range=29875,.;gbkey=mRNA;locus_tag=GSBLH_T00000003001;partial=true;transcript_id=XM_013038130.1\n",
            "NW_013171815.1\tRefSeq\texon\t29259\t29364\t.\t-\t.\tID=exon-XM_013038130.1-2;Parent=rna-XM_013038130.1;Dbxref=GeneID:24917328,GenBank:XM_013038130.1;gbkey=mRNA;locus_tag=GSBLH_T00000003001;partial=true;transcript_id=XM_013038130.1\n",
            "NW_013171815.1\tRefSeq\texon\t29003\t29226\t.\t-\t.\tID=exon-XM_013038130.1-3;Parent=rna-XM_013038130.1;Dbxref=GeneID:24917328,GenBank:XM_013038130.1;gbkey=mRNA;locus_tag=GSBLH_T00000003001;partial=true;transcript_id=XM_013038130.1\n",
            "NW_013171815.1\tRefSeq\texon\t28831\t28969\t.\t-\t.\tID=exon-XM_013038130.1-4;Parent=rna-XM_013038130.1;Dbxref=GeneID:24917328,GenBank:XM_013038130.1;gbkey=mRNA;locus_tag=GSBLH_T00000003001;partial=true;transcript_id=XM_013038130.1\n",
            "NW_013171815.1\tRefSeq\texon\t28708\t28801\t.\t-\t.\tID=exon-XM_013038130.1-5;Parent=rna-XM_013038130.1;Dbxref=GeneID:24917328,GenBank:XM_013038130.1;gbkey=mRNA;locus_tag=GSBLH_T00000003001;partial=true;transcript_id=XM_013038130.1\n",
            "NW_013171815.1\tRefSeq\texon\t28593\t28678\t.\t-\t.\tID=exon-XM_013038130.1-6;Parent=rna-XM_013038130.1;Dbxref=GeneID:24917328,GenBank:XM_013038130.1;gbkey=mRNA;locus_tag=GSBLH_T00000003001;partial=true;transcript_id=XM_013038130.1\n",
            "NW_013171815.1\tRefSeq\texon\t28397\t28559\t.\t-\t.\tID=exon-XM_013038130.1-7;Parent=rna-XM_013038130.1;Dbxref=GeneID:24917328,GenBank:XM_013038130.1;gbkey=mRNA;locus_tag=GSBLH_T00000003001;partial=true;start_range=.,28397;transcript_id=XM_013038130.1\n",
            "NW_013171815.1\tRefSeq\tCDS\t29398\t29875\t.\t-\t0\tID=cds-XP_012893584.1;Parent=rna-XM_013038130.1;Dbxref=UniProtKB/TrEMBL:D8LUP7,GeneID:24917328,GenBank:XP_012893584.1;Name=XP_012893584.1;gbkey=CDS;locus_tag=GSBLH_T00000003001;product=uncharacterized protein;protein_id=XP_012893584.1\n",
            "NW_013171815.1\tRefSeq\tCDS\t29259\t29364\t.\t-\t2\tID=cds-XP_012893584.1;Parent=rna-XM_013038130.1;Dbxref=UniProtKB/TrEMBL:D8LUP7,GeneID:24917328,GenBank:XP_012893584.1;Name=XP_012893584.1;gbkey=CDS;locus_tag=GSBLH_T00000003001;product=uncharacterized protein;protein_id=XP_012893584.1\n",
            "NW_013171815.1\tRefSeq\tCDS\t29003\t29226\t.\t-\t1\tID=cds-XP_012893584.1;Parent=rna-XM_013038130.1;Dbxref=UniProtKB/TrEMBL:D8LUP7,GeneID:24917328,GenBank:XP_012893584.1;Name=XP_012893584.1;gbkey=CDS;locus_tag=GSBLH_T00000003001;product=uncharacterized protein;protein_id=XP_012893584.1\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "def get_bed_file(feature, d):\n",
        "    res = []\n",
        "    with open('genomic.gff', 'r') as f :\n",
        "        for line in f.readlines()[7:]:\n",
        "            s = line.split()\n",
        "            if not s[0].startswith('#') and s[2] == feature:\n",
        "                res.append('\\t'.join([d[s[0]], s[3], s[4], s[6]+'\\n']))\n",
        "        with open(feature + '.bed', 'w') as f:\n",
        "            f.write(''.join(res))"
      ],
      "metadata": {
        "id": "jiLKSy_4eYBT"
      },
      "execution_count": 86,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "get_bed_file('exon', dictionary)\n",
        "get_bed_file('gene', dictionary)\n",
        "get_bed_file('mRNA', dictionary)\n",
        "get_bed_file('CDS', dictionary)"
      ],
      "metadata": {
        "id": "FQ5QU5nDgX3Q"
      },
      "execution_count": 87,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!head exon.bed"
      ],
      "metadata": {
        "id": "AP8hnalAlznJ",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "0212733a-9fe2-4ba3-c564-f4ca348f5e88"
      },
      "execution_count": 88,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "scaffold_1\t29398\t29875\t-\n",
            "scaffold_1\t29259\t29364\t-\n",
            "scaffold_1\t29003\t29226\t-\n",
            "scaffold_1\t28831\t28969\t-\n",
            "scaffold_1\t28708\t28801\t-\n",
            "scaffold_1\t28593\t28678\t-\n",
            "scaffold_1\t28397\t28559\t-\n",
            "scaffold_1\t33987\t34089\t-\n",
            "scaffold_1\t32865\t33937\t-\n",
            "scaffold_1\t37905\t38130\t-\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!head gene.bed"
      ],
      "metadata": {
        "id": "wikuGVT5n8jL",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "817b0844-6693-40c0-9bed-e3a6fa493178"
      },
      "execution_count": 89,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "scaffold_1\t28397\t29875\t-\n",
            "scaffold_1\t32865\t34089\t-\n",
            "scaffold_1\t36801\t38130\t-\n",
            "scaffold_1\t40343\t40664\t-\n",
            "scaffold_1\t41319\t41423\t-\n",
            "scaffold_1\t41664\t42924\t-\n",
            "scaffold_1\t45721\t46508\t+\n",
            "scaffold_1\t48901\t50065\t-\n",
            "scaffold_1\t67197\t68557\t-\n",
            "scaffold_1\t72838\t74089\t-\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# get introns\n",
        "!bedtools subtract -a gene.bed -b exon.bed > intron.bed"
      ],
      "metadata": {
        "id": "Zx6SSZZdahvS"
      },
      "execution_count": 90,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!head -n 10 intron.bed"
      ],
      "metadata": {
        "id": "PmmypUvITrOO",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "707b8c14-3a07-4b0a-bce1-144479a417cd"
      },
      "execution_count": 91,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "scaffold_1\t28559\t28593\t-\n",
            "scaffold_1\t28678\t28708\t-\n",
            "scaffold_1\t28801\t28831\t-\n",
            "scaffold_1\t28969\t29003\t-\n",
            "scaffold_1\t29226\t29259\t-\n",
            "scaffold_1\t29364\t29398\t-\n",
            "scaffold_1\t33937\t33987\t-\n",
            "scaffold_1\t37873\t37905\t-\n",
            "scaffold_1\t42636\t42668\t-\n",
            "scaffold_1\t49439\t49477\t-\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!bedtools flank -l 1000 -r 0 -i gene.bed -g lengths.txt > upstream.bed"
      ],
      "metadata": {
        "id": "M9ti3mtopVU2"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!bedtools flank -r 200 -l 0 -i gene.bed -g lengths.txt > downstream.bed"
      ],
      "metadata": {
        "id": "_Z-yDfbH7cd8"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!bedtools complement -i gene.bed -g lengths.txt > intergenic.bed"
      ],
      "metadata": {
        "id": "dU-zUaZw_s8f"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!bedtools annotate -i exon.bed -files GQ.bed zhunt.bed zdnabert_scaffold_1.bed > frac_exon.bed\n",
        "!bedtools annotate -i intron.bed -files GQ.bed zhunt.bed zdnabert_scaffold_1.bed > frac_intron.bed\n",
        "!bedtools annotate -i upstream.bed -files GQ.bed zhunt.bed zdnabert_scaffold_1.bed > frac_upstream.bed\n",
        "!bedtools annotate -i downstream.bed -files GQ.bed zhunt.bed zdnabert_scaffold_1.bed > frac_downstream.bed\n",
        "!bedtools annotate -i intergenic.bed -files GQ.bed zhunt.bed zdnabert_scaffold_1.bed > frac_intergenic.bed\n",
        "\n",
        "!bedtools annotate -counts -i exon.bed -files GQ.bed zhunt.bed zdnabert_scaffold_1.bed > cnt_exon.bed\n",
        "!bedtools annotate -counts -i intron.bed -files GQ.bed zhunt.bed zdnabert_scaffold_1.bed > cnt_intron.bed\n",
        "!bedtools annotate -counts -i upstream.bed -files GQ.bed zhunt.bed zdnabert_scaffold_1.bed > cnt_upstream.bed\n",
        "!bedtools annotate -counts -i downstream.bed -files GQ.bed zhunt.bed zdnabert_scaffold_1.bed > cnt_downstream.bed\n",
        "!bedtools annotate -counts -i intergenic.bed -files GQ.bed zhunt.bed zdnabert_scaffold_1.bed > cnt_intergenic.bed"
      ],
      "metadata": {
        "id": "BAer6pW7-VZp"
      },
      "execution_count": 94,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "import numpy as np\n",
        "\n",
        "def get_cnt(myfile):\n",
        "    gq, zhunt, bert = 105, 13855, 1178\n",
        "    df = pd.read_csv(myfile, sep='\\t', header=None)\n",
        "    counts = list(df[[4, 5, 6]].sum()) if 6 in df.columns else list(df[[3, 4, 5]].sum())\n",
        "    print(f'| {counts[0]} | {round(counts[0]/gq, 2)} | {counts[1]} | {round(counts[1]/zhunt, 2)} | {counts[2]} | {round(counts[2]/bert, 2)} |')\n",
        "\n",
        "def get_frac(myfile):\n",
        "    df = pd.read_csv(myfile, sep='\\t', header=None)\n",
        "    length = df[df[0]=='scaffold_1'][2].sum() - df[df[0]=='scaffold_1'][1].sum()\n",
        "    counts = (np.array(df[[4, 5, 6]]) * np.array(df[2] - df[1])[:, None]).sum(axis=0) if 6 in df.columns \\\n",
        "        else (np.array(df[[3, 4, 5]]) * np.array(df[2] - df[1])[:, None]).sum(axis=0)\n",
        "    print(f'| {round(counts[0]/length, 3)} | {round(counts[1]/length, 3)} | {round(counts[2]/length, 3)} |')\n",
        "\n",
        "regions = ['exon', 'intron', 'upstream', 'downstream', 'intergenic']\n",
        "\n",
        "print('counts:')\n",
        "print('name', 'gq', 'zhunt', 'zdnabert')\n",
        "for region in regions:\n",
        "    filename = 'cnt_' + region + '.bed'\n",
        "    print('|', region, end=' ')\n",
        "    get_cnt(filename)\n",
        "\n",
        "print('fraction:')\n",
        "print('name', 'gq', 'zhunt', 'zdnabert')\n",
        "for region in regions:\n",
        "    filename = 'frac_' + region + '.bed'\n",
        "    print('|', region, end=' ')\n",
        "    get_frac(filename)"
      ],
      "metadata": {
        "id": "gc1XfJ61Ab2_",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "22fc7a7a-6282-47b4-e40d-1012963acb36"
      },
      "execution_count": 96,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "counts:\n",
            "name gq zhunt zdnabert\n",
            "| exon | 42 | 0.4 | 9731 | 0.7 | 836 | 0.71 |\n",
            "| intron | 8 | 0.08 | 675 | 0.05 | 49 | 0.04 |\n",
            "| upstream | 34 | 0.32 | 4040 | 0.29 | 336 | 0.29 |\n",
            "| downstream | 9 | 0.09 | 682 | 0.05 | 48 | 0.04 |\n",
            "| intergenic | 59 | 0.56 | 3949 | 0.29 | 328 | 0.28 |\n",
            "fraction:\n",
            "name gq zhunt zdnabert\n",
            "| exon | 0.001 | 0.032 | 0.014 |\n",
            "| intron | 0.002 | 0.013 | 0.005 |\n",
            "| upstream | 0.001 | 0.016 | 0.007 |\n",
            "| downstream | 0.002 | 0.012 | 0.005 |\n",
            "| intergenic | 0.002 | 0.013 | 0.005 |\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!head frac_exon.bed"
      ],
      "metadata": {
        "id": "hBnURk-A_Eq6",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "df2387ed-32e9-49a0-faaf-e83fdbc7c814"
      },
      "execution_count": 97,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "scaffold_1\t1048443\t1050102\t+\t0.000000\t0.081977\t0.020494\n",
            "scaffold_1\t128263\t131309\t-\t0.000000\t0.045305\t0.024622\n",
            "scaffold_1\t261862\t262266\t-\t0.000000\t0.000000\t0.000000\n",
            "scaffold_1\t392631\t393322\t+\t0.000000\t0.039074\t0.021708\n",
            "scaffold_1\t917451\t917556\t-\t0.000000\t0.000000\t0.000000\n",
            "scaffold_1\t1441787\t1441991\t+\t0.000000\t0.000000\t0.000000\n",
            "scaffold_1\t1702971\t1704929\t-\t0.000000\t0.000000\t0.000000\n",
            "scaffold_1\t1940419\t1966329\t-\t0.000000\t0.045890\t0.024354\n",
            "scaffold_1\t48901\t49439\t-\t0.000000\t0.000000\t0.000000\n",
            "scaffold_1\t114155\t114841\t+\t0.000000\t0.000000\t0.000000\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "DsyFnORPXqdD"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}